{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1106b087",
   "metadata": {},
   "source": [
    "# VARify"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c87d4722",
   "metadata": {},
   "source": [
    "\n",
    "## Workflow\n",
    "\n",
    "\n",
    "<img src=\"wf.png\" width=\"600\">"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e45e888",
   "metadata": {},
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ac6d6771",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Utility File\n",
    "\n",
    "# libraries\n",
    "import warnings\n",
    "import pandas as pd\n",
    "\n",
    "# For every row in here, we're going to apply this uniqueSNP function\n",
    "\n",
    "def mergeTab(mp, so):\n",
    "    sub_mp = mp[['snp_pos', 'pileup']]\n",
    "    mdf = pd.merge(so, sub_mp, on='snp_pos')\n",
    "    return mdf\n",
    "\n",
    "\n",
    "def varOnly(a):\n",
    "    '''\n",
    "    purpose: remove non-nt values in pileup\n",
    "    input: string of mpileup \n",
    "    output: nt string\n",
    "    '''\n",
    "    x = [i for i in a if i.upper() in ['A','C','T','G']]\n",
    "    \n",
    "    return x\n",
    "\n",
    "def isMono(ref, temp):\n",
    "    # Check if two snps are monomorphic\n",
    "    c=0\n",
    "    count = [c+1 for snp in temp if snp == ref]\n",
    "    if sum(count) >= 2: \n",
    "        warnings.warn(\"More than 2 monomorphic alleles at position.\")\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "    \n",
    "def multiError(x, pos):\n",
    "    if len(x) > 2: \n",
    "        warnings.warn(\"There are more than 2 SNPS at position\"); print(pos)\n",
    "    \n",
    "def uniqueSNP(x):\n",
    "    '''\n",
    "    Takes mPileup\n",
    "    Confirm Alt\n",
    "    '''\n",
    "    l = str(x['pileup'])\n",
    "    # make upper\n",
    "    l = l.upper() # just to make simple\n",
    "    l = varOnly(l)\n",
    "\n",
    "    # Get list of unique\n",
    "    a = [i for i in l]\n",
    "    a_unique = list(set(a)) \n",
    "    \n",
    "    # create a dictionary\n",
    "    df = pd.DataFrame()\n",
    "    temp = [i for i in x['ref_codon']]\n",
    "    snp_dict = {}\n",
    "    \n",
    "#     flag = False\n",
    "    \n",
    "#     multiError(a_unique, x['snp_pos'])\n",
    "\n",
    "    test = []\n",
    "    '''\n",
    "    Creates a combination of SNPs\n",
    "    Replaces Alt at each position\n",
    "    '''\n",
    "    for snp in a_unique:\n",
    "        snp_dict['snp_pos'] = int(x['snp_pos'])\n",
    "        snp_dict['varify_alt'] = snp\n",
    "\n",
    "#         if flag:\n",
    "#             snp_dict['flag'] = '*'\n",
    "#         else:\n",
    "#             snp_dict['flag'] = ''\n",
    "        \n",
    "#         for i in range(len(temp)):\n",
    "#             temp[i] = snp\n",
    "#             test.append(''.join(temp))\n",
    "#             snp_dict[f\"varify_codon{i+1}\"] = ''.join(temp)\n",
    "#             temp = [i for i in ref]\n",
    "            \n",
    "#         flag = isMono(ref, test)  \n",
    "        \n",
    "#         if flag: \n",
    "#             snp_dict['flag'] = '*'\n",
    "#         else:\n",
    "#             snp_dict['flag'] = ''\n",
    "            \n",
    "        df = df.append(snp_dict, ignore_index=True)\n",
    "\n",
    "\n",
    "        snp_dict = {}\n",
    "        test = []\n",
    "\n",
    "    df = pd.merge(cf, df, on='snp_pos')\n",
    "    return df\n",
    "\n",
    "def altValidate(x):\n",
    "    # Varify alt == Alt? T or F\n",
    "    # Add flag to capture\n",
    "    \n",
    "    x['VARified'] = 'Yes' if x['alt_allele'] == x['varify_alt'] else 'No'\n",
    "    \n",
    "    return x\n",
    "\n",
    "def varify(x):\n",
    "    df = x.apply(axis=1, func=uniqueSNP)\n",
    "    df = pd.concat(list(df))\n",
    "    x = df.apply(axis=1, func=altValidate)\n",
    "\n",
    "    df = x.apply(axis = 1, func=altCombo)\n",
    "    df = pd.concat(list(df))\n",
    "    df = pd.merge(x, df, on='snp_pos')\n",
    "    \n",
    "    reorg = ['chr_id', 'snp_pos', 'ref_allele', 'alt_allele', \n",
    "             'varify_alt', 'VARified', \n",
    "             'codon1_genome_pos','codon2_genome_pos',\n",
    "             'codon3_genome_pos', 'ref_codon', \n",
    "             'varify_codon1', 'varify_codon2', 'varify_codon3',#'flag'\n",
    "            ]\n",
    "    \n",
    "    return df[reorg]\n",
    "\n",
    "def getAlt(x):\n",
    "    \n",
    "    '''\n",
    "    Takes mPileup\n",
    "    Out: list of uniqueSNP\n",
    "    '''\n",
    "    l = x['pileup'].astype(str)[0]\n",
    "    # make upper\n",
    "    l = [i for i in l if i.upper() in ['A','C','T','G']]\n",
    "\n",
    "    # Get list of unique\n",
    "    a = [i for i in l]\n",
    "    a_unique = list(set(a)) \n",
    "    a_unique\n",
    "    \n",
    "    return(a_unique)\n",
    "\n",
    "\n",
    "def translate(x):\n",
    "    codon = x['varify_codon'][0]\n",
    "    \n",
    "    table = {\n",
    "    'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "    'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "    'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "    'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',                \n",
    "    'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "    'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "    'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "    'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "    'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "    'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "    'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "    'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "    'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "    'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "    'TAC':'Y', 'TAT':'Y', 'TAA':'*', 'TAG':'*',\n",
    "    'TGC':'C', 'TGT':'C', 'TGA':'*', 'TGG':'W',}\n",
    "    \n",
    "    x['varify_aa'] =table[codon]\n",
    "    \n",
    "    return(x)\n",
    "\n",
    "def effect(x):\n",
    "    # missense\n",
    "    # nonsense\n",
    "    x['effect'] = '_'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b3680f4",
   "metadata": {},
   "source": [
    "## Exploring Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "70705ac3",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "# mpileup \n",
    "mp = pd.read_table('../data/test/mpiletest-out.txt', header=None)\n",
    "mp = mp.rename(columns={1:\"snp_pos\", 4:\"pileup\"})\n",
    "\n",
    "# custom out file\n",
    "cf = pd.read_table('../data/test/snp_out.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ccd793ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>codon1_genome_pos</th>\n",
       "      <th>codon2_genome_pos</th>\n",
       "      <th>codon3_genome_pos</th>\n",
       "      <th>ref_codon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>521</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>519</td>\n",
       "      <td>520</td>\n",
       "      <td>521</td>\n",
       "      <td>AAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>523</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>521</td>\n",
       "      <td>522</td>\n",
       "      <td>523</td>\n",
       "      <td>GTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>524</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>522</td>\n",
       "      <td>523</td>\n",
       "      <td>524</td>\n",
       "      <td>TAC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele  codon1_genome_pos  codon2_genome_pos  \\\n",
       "0   chr1      521          G          C                519                520   \n",
       "1   chr1      523          A          T                521                522   \n",
       "2   chr1      524          C          G                522                523   \n",
       "\n",
       "   codon3_genome_pos ref_codon  \n",
       "0                521       AAG  \n",
       "1                523       GTA  \n",
       "2                524       TAC  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# column names\n",
    "# cols = ['chr_id', 'snp_pos', 'ref_allele', \n",
    "#         'alt_allele', 'gene_id', 'mrna_id',\n",
    "#         'prot_id', 'strand', 'effect', \n",
    "#         'snp_cds_pos', 'codon1_genome_pos',\n",
    "#         'codon2_genome_pos', 'codon3_genome_pos', \n",
    "#         'snp_aa_pos', 'ref_codon','alt_codon', \n",
    "#         'ref_aa', 'alt_aa']\n",
    "\n",
    "sub_cols = ['chr_id', 'snp_pos', 'ref_allele', \n",
    "            'alt_allele', 'codon1_genome_pos',\n",
    "            'codon2_genome_pos', 'codon3_genome_pos', \n",
    "            'ref_codon']\n",
    "\n",
    "# subset of custom file, cf\n",
    "sub_cf = cf[sub_cols]\n",
    "sub_cf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "509754eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>pileup</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>seq1</td>\n",
       "      <td>521</td>\n",
       "      <td>G</td>\n",
       "      <td>26</td>\n",
       "      <td>.,......A.a.A,..,..,.,.,,^].</td>\n",
       "      <td>DJJJJJJJJJJJJJJJJJJJJJJJA7</td>\n",
       "      <td>]]]]]]]]]]]]]]]]]]]]]]]]]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>seq1</td>\n",
       "      <td>522</td>\n",
       "      <td>T</td>\n",
       "      <td>25</td>\n",
       "      <td>,...........,..,..,.,.,,.</td>\n",
       "      <td>JJJJJJJJJJJJJJJJJJJJJJJA7</td>\n",
       "      <td>]]]]]]]]]]]]]]]]]]]]]]]]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>seq1</td>\n",
       "      <td>523</td>\n",
       "      <td>A</td>\n",
       "      <td>25</td>\n",
       "      <td>tTTTTTTTTTTTtTTtTTtTtTttT</td>\n",
       "      <td>JJJJJJJJJJJJJJJJJJJJJJJA7</td>\n",
       "      <td>]]]]]]]]]]]]]]]]]]]]]]]]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>seq1</td>\n",
       "      <td>524</td>\n",
       "      <td>C</td>\n",
       "      <td>27</td>\n",
       "      <td>,...G.......,.G,..,.,.,,G..</td>\n",
       "      <td>JJJJJJJJJJJJJJJJJJJJJJJJG88</td>\n",
       "      <td>]]]]]]]]]]]]]]]]]]]]]]]]]]]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0  snp_pos  2   3                        pileup  \\\n",
       "0  seq1      521  G  26  .,......A.a.A,..,..,.,.,,^].   \n",
       "1  seq1      522  T  25     ,...........,..,..,.,.,,.   \n",
       "2  seq1      523  A  25     tTTTTTTTTTTTtTTtTTtTtTttT   \n",
       "3  seq1      524  C  27   ,...G.......,.G,..,.,.,,G..   \n",
       "\n",
       "                             5                            6  \n",
       "0   DJJJJJJJJJJJJJJJJJJJJJJJA7   ]]]]]]]]]]]]]]]]]]]]]]]]]]  \n",
       "1    JJJJJJJJJJJJJJJJJJJJJJJA7    ]]]]]]]]]]]]]]]]]]]]]]]]]  \n",
       "2    JJJJJJJJJJJJJJJJJJJJJJJA7    ]]]]]]]]]]]]]]]]]]]]]]]]]  \n",
       "3  JJJJJJJJJJJJJJJJJJJJJJJJG88  ]]]]]]]]]]]]]]]]]]]]]]]]]]]  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mp"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26ef5f9d",
   "metadata": {},
   "source": [
    "## Merging the two dataframes together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "98f81b15",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mrna_id</th>\n",
       "      <th>prot_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>effect</th>\n",
       "      <th>snp_cds_pos</th>\n",
       "      <th>codon1_genome_pos</th>\n",
       "      <th>codon2_genome_pos</th>\n",
       "      <th>codon3_genome_pos</th>\n",
       "      <th>snp_aa_pos</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>alt_codon</th>\n",
       "      <th>ref_aa</th>\n",
       "      <th>alt_aa</th>\n",
       "      <th>pileup</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>521</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>seq1</td>\n",
       "      <td>seq1.1</td>\n",
       "      <td>seq1.1</td>\n",
       "      <td>+</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>265</td>\n",
       "      <td>519</td>\n",
       "      <td>520</td>\n",
       "      <td>521</td>\n",
       "      <td>89</td>\n",
       "      <td>AAG</td>\n",
       "      <td>AAC</td>\n",
       "      <td>K</td>\n",
       "      <td>N</td>\n",
       "      <td>.,......A.a.A,..,..,.,.,,^].</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>523</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>seq1</td>\n",
       "      <td>seq1.1</td>\n",
       "      <td>seq1.1</td>\n",
       "      <td>+</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>267</td>\n",
       "      <td>521</td>\n",
       "      <td>522</td>\n",
       "      <td>523</td>\n",
       "      <td>89</td>\n",
       "      <td>GTA</td>\n",
       "      <td>GTT</td>\n",
       "      <td>E</td>\n",
       "      <td>E</td>\n",
       "      <td>tTTTTTTTTTTTtTTtTTtTtTttT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>524</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>seq1</td>\n",
       "      <td>seq1.1</td>\n",
       "      <td>seq1.1</td>\n",
       "      <td>+</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>268</td>\n",
       "      <td>522</td>\n",
       "      <td>523</td>\n",
       "      <td>524</td>\n",
       "      <td>89</td>\n",
       "      <td>TAC</td>\n",
       "      <td>TAG</td>\n",
       "      <td>Y</td>\n",
       "      <td>X</td>\n",
       "      <td>,...G.......,.G,..,.,.,,G..</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele gene_id mrna_id prot_id strand  \\\n",
       "0   chr1      521          G          C    seq1  seq1.1  seq1.1      +   \n",
       "1   chr1      523          A          T    seq1  seq1.1  seq1.1      +   \n",
       "2   chr1      524          C          G    seq1  seq1.1  seq1.1      +   \n",
       "\n",
       "       effect  snp_cds_pos  codon1_genome_pos  codon2_genome_pos  \\\n",
       "0  synonymous          265                519                520   \n",
       "1  synonymous          267                521                522   \n",
       "2  synonymous          268                522                523   \n",
       "\n",
       "   codon3_genome_pos  snp_aa_pos ref_codon alt_codon ref_aa alt_aa  \\\n",
       "0                521          89       AAG       AAC      K      N   \n",
       "1                523          89       GTA       GTT      E      E   \n",
       "2                524          89       TAC       TAG      Y      X   \n",
       "\n",
       "                         pileup  \n",
       "0  .,......A.a.A,..,..,.,.,,^].  \n",
       "1     tTTTTTTTTTTTtTTtTTtTtTttT  \n",
       "2   ,...G.......,.G,..,.,.,,G..  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sub_mp = mp[['snp_pos','pileup']]\n",
    "merged_var = pd.merge(cf, sub_mp, on='snp_pos')\n",
    "merged_var"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13186127",
   "metadata": {},
   "source": [
    "## Split Dataframe\n",
    "\n",
    "Splitting the snpeff table by codons, get genome position and alt/ref codon\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7eac0ac4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>varify_alt</th>\n",
       "      <th>VARified</th>\n",
       "      <th>codon1_genome_pos</th>\n",
       "      <th>codon2_genome_pos</th>\n",
       "      <th>codon3_genome_pos</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>varify_codon1</th>\n",
       "      <th>varify_codon2</th>\n",
       "      <th>varify_codon3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>521</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>No</td>\n",
       "      <td>519</td>\n",
       "      <td>520</td>\n",
       "      <td>521</td>\n",
       "      <td>AAG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>523</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>Yes</td>\n",
       "      <td>521</td>\n",
       "      <td>522</td>\n",
       "      <td>523</td>\n",
       "      <td>GTA</td>\n",
       "      <td>C.T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>524</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>Yes</td>\n",
       "      <td>522</td>\n",
       "      <td>523</td>\n",
       "      <td>524</td>\n",
       "      <td>TAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TTG</td>\n",
       "      <td>TAG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele varify_alt VARified  \\\n",
       "0   chr1      521          G          C          A       No   \n",
       "1   chr1      523          A          T          T      Yes   \n",
       "2   chr1      524          C          G          G      Yes   \n",
       "\n",
       "   codon1_genome_pos  codon2_genome_pos  codon3_genome_pos ref_codon  \\\n",
       "0                519                520                521       AAG   \n",
       "1                521                522                523       GTA   \n",
       "2                522                523                524       TAC   \n",
       "\n",
       "  varify_codon1 varify_codon2 varify_codon3  \n",
       "0           NaN           NaN           AAC  \n",
       "1           C.T           NaN           GTT  \n",
       "2           NaN           TTG           TAG  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdf = mergeTab(mp, cf)\n",
    "varify(merged_var)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c37f09f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>pileup</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>seq1</td>\n",
       "      <td>521</td>\n",
       "      <td>G</td>\n",
       "      <td>26</td>\n",
       "      <td>.,......A.a.A,..,..,.,.,,^].</td>\n",
       "      <td>DJJJJJJJJJJJJJJJJJJJJJJJA7</td>\n",
       "      <td>]]]]]]]]]]]]]]]]]]]]]]]]]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>seq1</td>\n",
       "      <td>522</td>\n",
       "      <td>T</td>\n",
       "      <td>25</td>\n",
       "      <td>,...........,..,..,.,.,,.</td>\n",
       "      <td>JJJJJJJJJJJJJJJJJJJJJJJA7</td>\n",
       "      <td>]]]]]]]]]]]]]]]]]]]]]]]]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>seq1</td>\n",
       "      <td>523</td>\n",
       "      <td>A</td>\n",
       "      <td>25</td>\n",
       "      <td>tTTTTTTTTTTTtTTtTTtTtTttT</td>\n",
       "      <td>JJJJJJJJJJJJJJJJJJJJJJJA7</td>\n",
       "      <td>]]]]]]]]]]]]]]]]]]]]]]]]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>seq1</td>\n",
       "      <td>524</td>\n",
       "      <td>C</td>\n",
       "      <td>27</td>\n",
       "      <td>,...G.......,.G,..,.,.,,G..</td>\n",
       "      <td>JJJJJJJJJJJJJJJJJJJJJJJJG88</td>\n",
       "      <td>]]]]]]]]]]]]]]]]]]]]]]]]]]]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0  snp_pos  2   3                        pileup  \\\n",
       "0  seq1      521  G  26  .,......A.a.A,..,..,.,.,,^].   \n",
       "1  seq1      522  T  25     ,...........,..,..,.,.,,.   \n",
       "2  seq1      523  A  25     tTTTTTTTTTTTtTTtTTtTtTttT   \n",
       "3  seq1      524  C  27   ,...G.......,.G,..,.,.,,G..   \n",
       "\n",
       "                             5                            6  \n",
       "0   DJJJJJJJJJJJJJJJJJJJJJJJA7   ]]]]]]]]]]]]]]]]]]]]]]]]]]  \n",
       "1    JJJJJJJJJJJJJJJJJJJJJJJA7    ]]]]]]]]]]]]]]]]]]]]]]]]]  \n",
       "2    JJJJJJJJJJJJJJJJJJJJJJJA7    ]]]]]]]]]]]]]]]]]]]]]]]]]  \n",
       "3  JJJJJJJJJJJJJJJJJJJJJJJJG88  ]]]]]]]]]]]]]]]]]]]]]]]]]]]  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# pile up \n",
    "mp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "38ae94f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mrna_id</th>\n",
       "      <th>prot_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>effect</th>\n",
       "      <th>snp_cds_pos</th>\n",
       "      <th>codon1_genome_pos</th>\n",
       "      <th>codon2_genome_pos</th>\n",
       "      <th>codon3_genome_pos</th>\n",
       "      <th>snp_aa_pos</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>alt_codon</th>\n",
       "      <th>ref_aa</th>\n",
       "      <th>alt_aa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>521</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>seq1</td>\n",
       "      <td>seq1.1</td>\n",
       "      <td>seq1.1</td>\n",
       "      <td>+</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>265</td>\n",
       "      <td>519</td>\n",
       "      <td>520</td>\n",
       "      <td>521</td>\n",
       "      <td>89</td>\n",
       "      <td>AAG</td>\n",
       "      <td>AAC</td>\n",
       "      <td>K</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>523</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>seq1</td>\n",
       "      <td>seq1.1</td>\n",
       "      <td>seq1.1</td>\n",
       "      <td>+</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>267</td>\n",
       "      <td>521</td>\n",
       "      <td>522</td>\n",
       "      <td>523</td>\n",
       "      <td>89</td>\n",
       "      <td>GTA</td>\n",
       "      <td>GTT</td>\n",
       "      <td>E</td>\n",
       "      <td>E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>524</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>seq1</td>\n",
       "      <td>seq1.1</td>\n",
       "      <td>seq1.1</td>\n",
       "      <td>+</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>268</td>\n",
       "      <td>522</td>\n",
       "      <td>523</td>\n",
       "      <td>524</td>\n",
       "      <td>89</td>\n",
       "      <td>TAC</td>\n",
       "      <td>TAG</td>\n",
       "      <td>Y</td>\n",
       "      <td>X</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele gene_id mrna_id prot_id strand  \\\n",
       "0   chr1      521          G          C    seq1  seq1.1  seq1.1      +   \n",
       "1   chr1      523          A          T    seq1  seq1.1  seq1.1      +   \n",
       "2   chr1      524          C          G    seq1  seq1.1  seq1.1      +   \n",
       "\n",
       "       effect  snp_cds_pos  codon1_genome_pos  codon2_genome_pos  \\\n",
       "0  synonymous          265                519                520   \n",
       "1  synonymous          267                521                522   \n",
       "2  synonymous          268                522                523   \n",
       "\n",
       "   codon3_genome_pos  snp_aa_pos ref_codon alt_codon ref_aa alt_aa  \n",
       "0                521          89       AAG       AAC      K      N  \n",
       "1                523          89       GTA       GTT      E      E  \n",
       "2                524          89       TAC       TAG      Y      X  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# snp-eff like table\n",
    "cf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32ede3ea",
   "metadata": {},
   "source": [
    "# VARify for same ALT as prediction\n",
    "\n",
    "| Pos     | Ref     | Alt |Varify_Alt|Varify|\n",
    "| ----------- | ----------- | ----------- | ----------- | ----------- |\n",
    "|202| A      | T       |T|Yes|\n",
    "|204| C   | G        |.|No|\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "efd4acb4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>varify_alt</th>\n",
       "      <th>VARified</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>521</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>523</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>524</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele varify_alt VARified\n",
       "0   chr1      521          G          C          A       No\n",
       "1   chr1      523          A          T          T      Yes\n",
       "2   chr1      524          C          G          G      Yes"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "varify_out = varify(merged_var)\n",
    "\n",
    "varify_out.iloc[:,:6]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c7dcd34",
   "metadata": {},
   "source": [
    "# Capture codon combinations from different positions\n",
    "\n",
    "Logic: \n",
    "If position 1,2,3 and 1 and 3 have variances - we want to maintain these variances. \n",
    "\n",
    "We can place variances of 1 and 3 into dictionaries. \n",
    "If at position 3 and position 1 has variances --> combine and showcase AA\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "380eaf3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Codon Combos\n",
    "import numpy as np\n",
    "import math \n",
    "\n",
    "def altDict(x):\n",
    "    '''\n",
    "    \n",
    "    Creates a combination of SNPs\n",
    "    if snps have existing alts\n",
    "    \n",
    "    Outputs into a dictionary\n",
    "    \n",
    "    '''\n",
    "    # At each line, takes reference codon\n",
    "    # sub_x = x.loc[x['VARified'] == 'Yes']\n",
    "    \n",
    "    # create a dictionary of Alts\n",
    "    alt_dict = dict(zip(x['snp_pos'], x['alt_allele']))\n",
    "    clean_dict = {k: alt_dict[k] for k in alt_dict if pd.isna(alt_dict[k]) == False}\n",
    "    return clean_dict\n",
    "\n",
    "# alt_dict = altDict(merged_var)\n",
    "\n",
    "def altCombo(x, alt_dict=alt_dict):\n",
    "    '''\n",
    "    Creates combos based on\n",
    "    Current position and alt dictionary\n",
    "    Can be used with apply\n",
    "    '''\n",
    "    \n",
    "    p3 = int(x['snp_pos'])\n",
    "    p2 = p3-1\n",
    "    p1 = p2-1\n",
    "    \n",
    "    subDict = {p1 : '.', p2 : '.', p3 : '.'}\n",
    "    \n",
    "    # create a dictionary\n",
    "    ref = x['ref_codon']\n",
    "    df = pd.DataFrame()\n",
    "    temp = [i for i in ref]\n",
    "    snp_dict = {}\n",
    "    \n",
    "    # Add values from full dictionary if keys are in sub\n",
    "    res = {key: alt_dict.get(key, subDict[key]) for key in subDict}\n",
    "    res = ''.join(res.values())\n",
    "        \n",
    "    # Create combination of alts\n",
    "    # at pos3, replace 1; at pos2, replace 2\n",
    "    for n in range(3):\n",
    "        snp_dict['snp_pos'] = int(x['snp_pos'])\n",
    "        if res[n] != '.':\n",
    "            # replace res into temp\n",
    "            temp[n:] = res[n:]\n",
    "            snp_dict[f\"varify_codon{n+1}\"] = ''.join(temp)\n",
    "            \n",
    "        else:\n",
    "            snp_dict[f\"varify_codon{n+1}\"] = np.nan\n",
    "\n",
    "        temp = [i for i in ref]\n",
    "    \n",
    "    df = df.append(snp_dict, ignore_index=True)\n",
    "        \n",
    "    #df = pd.concat(list(df))\n",
    "    #print(df)     \n",
    "    #df = pd.merge(x, df, on='snp_pos')\n",
    "    \n",
    "    return(df)\n",
    "\n",
    "def flatten(d, parent_key='', sep='_'):\n",
    "    import collections\n",
    "    items = []\n",
    "    for k, v in d.items():\n",
    "        new_key = parent_key + sep + k if parent_key else k\n",
    "        if isinstance(v, collections.MutableMapping):\n",
    "            items.extend(flatten(v, new_key, sep=sep).items())\n",
    "        else:\n",
    "            items.append((new_key, v))\n",
    "    return dict(items)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "6d6dfacc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>varify_alt</th>\n",
       "      <th>VARified</th>\n",
       "      <th>codon1_genome_pos</th>\n",
       "      <th>codon2_genome_pos</th>\n",
       "      <th>codon3_genome_pos</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>varify_codon1_x</th>\n",
       "      <th>varify_codon2_x</th>\n",
       "      <th>varify_codon3_x</th>\n",
       "      <th>varify_codon1_y</th>\n",
       "      <th>varify_codon2_y</th>\n",
       "      <th>varify_codon3_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>521</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>No</td>\n",
       "      <td>519</td>\n",
       "      <td>520</td>\n",
       "      <td>521</td>\n",
       "      <td>AAG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>523</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>Yes</td>\n",
       "      <td>521</td>\n",
       "      <td>522</td>\n",
       "      <td>523</td>\n",
       "      <td>GTA</td>\n",
       "      <td>C.T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GTT</td>\n",
       "      <td>C.T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>524</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>Yes</td>\n",
       "      <td>522</td>\n",
       "      <td>523</td>\n",
       "      <td>524</td>\n",
       "      <td>TAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TTG</td>\n",
       "      <td>TAG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TTG</td>\n",
       "      <td>TAG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele varify_alt VARified  \\\n",
       "0   chr1      521          G          C          A       No   \n",
       "1   chr1      523          A          T          T      Yes   \n",
       "2   chr1      524          C          G          G      Yes   \n",
       "\n",
       "   codon1_genome_pos  codon2_genome_pos  codon3_genome_pos ref_codon  \\\n",
       "0                519                520                521       AAG   \n",
       "1                521                522                523       GTA   \n",
       "2                522                523                524       TAC   \n",
       "\n",
       "  varify_codon1_x varify_codon2_x varify_codon3_x varify_codon1_y  \\\n",
       "0             NaN             NaN             AAC             NaN   \n",
       "1             C.T             NaN             GTT             C.T   \n",
       "2             NaN             TTG             TAG             NaN   \n",
       "\n",
       "  varify_codon2_y varify_codon3_y  \n",
       "0             NaN             AAC  \n",
       "1             NaN             GTT  \n",
       "2             TTG             TAG  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 3/31/23 - Do not remove Non-VARify Alts\n",
    "# sub_x = varify_out.loc[varify_out['VARified'] == 'Yes'] \n",
    "# alt_dict = dict(zip(sub_x['snp_pos'], sub_x['alt_allele']))\n",
    "df = varify_out.apply(axis = 1, func=altCombo)\n",
    "df = pd.concat(list(df))\n",
    "df = pd.merge(varify_out, df, on='snp_pos')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "3ee16fbc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>varify_alt</th>\n",
       "      <th>VARified</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>varify_codon1</th>\n",
       "      <th>varify_codon2</th>\n",
       "      <th>varify_codon3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>521</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>No</td>\n",
       "      <td>AAG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>523</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>Yes</td>\n",
       "      <td>GTA</td>\n",
       "      <td>C.T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>524</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>Yes</td>\n",
       "      <td>TAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TTG</td>\n",
       "      <td>TAG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele varify_alt VARified ref_codon  \\\n",
       "0   chr1      521          G          C          A       No       AAG   \n",
       "1   chr1      523          A          T          T      Yes       GTA   \n",
       "2   chr1      524          C          G          G      Yes       TAC   \n",
       "\n",
       "  varify_codon1 varify_codon2 varify_codon3  \n",
       "0           NaN           NaN           AAC  \n",
       "1           C.T           NaN           GTT  \n",
       "2           NaN           TTG           TAG  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out = varify(merged_var)\n",
    "out[['chr_id', 'snp_pos', 'ref_allele', 'alt_allele', 'varify_alt',\n",
    "       'VARified','ref_codon', 'varify_codon1', 'varify_codon2',\n",
    "       'varify_codon3']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f708429",
   "metadata": {},
   "source": [
    "# Running Script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "37bbd614",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  chr_id  snp_pos ref_allele alt_allele varify_alt  codon1_genome_pos  \\\n",
      "0   chr1      521          G          C          A                519   \n",
      "0   chr1      523          A          T          T                521   \n",
      "0   chr1      524          C          G          G                522   \n",
      "\n",
      "   codon2_genome_pos  codon3_genome_pos ref_codon varify_codon1 varify_codon2  \\\n",
      "0                520                521       AAG           AAG           AAG   \n",
      "0                522                523       GTA           TTA           GTA   \n",
      "0                523                524       TAC           GAC           TGC   \n",
      "\n",
      "  varify_codon3 flag  \n",
      "0           AAA       \n",
      "0           GTT       \n",
      "0           TAG       \n"
     ]
    }
   ],
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "%run '../script/main.py' --in_mp '../data/test/mpiletest-out.txt' --in_so '../data/test/snp_out.txt'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "302e41ef",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Case 1\n",
    "\n",
    "Consider the following mutation that introduces a G->A mutation.\n",
    " \n",
    "VCF:\n",
    "Chr1\t1100023      \t.       \tG     \tA      \t60    \tPASS\t.       \tGT    \t0/1\n",
    " \n",
    "* Genotype: G/A\n",
    "* Reference base: G\n",
    "* Alternate base: A\n",
    "* Predicted reference codon: TGG (tryptophan)\n",
    "* Predicted alternate codon: TGA (premature termination codon)\n",
    " \n",
    "A tool like snpEff can be used to predict the effect of the mutation using a reference genome annotation (e.g., GFF). In this case the mutation occurs in a coding exon at third position of a codon that codes for tryptophan in the reference proteome and a premature termination codon in the case of the alternate allele. snpEff will determine the reference and alternate codon and return the two codons and their translation.\n",
    " \n",
    "The predicted effect is therefore a nonsense mutation.\n",
    "\n",
    "## Goal Example\n",
    "| chr_id|snp_pos     | Ref     | Pred_Alt |Varify_Alt|VARified|ref_codon|varify_codon|aa|effect|\n",
    "| -----------| ----------- | ----------- | ----------- | ----------- | ----------- |----------- |----------- |----------- |----------- |\n",
    "|Chr1|1100023| G      | A       |A|Yes| TGG | TGA |*|stop_gain|"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d279c23d",
   "metadata": {},
   "source": [
    "## Read in the Toy Mpileup File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "76adbb9d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>3</th>\n",
       "      <th>pileup</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100021</td>\n",
       "      <td>T</td>\n",
       "      <td>12</td>\n",
       "      <td>.............</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100022</td>\n",
       "      <td>G</td>\n",
       "      <td>13</td>\n",
       "      <td>.............</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>G</td>\n",
       "      <td>13</td>\n",
       "      <td>...A...A..A..</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0  snp_pos ref   3         pileup\n",
       "0  Chr1  1100021   T  12  .............\n",
       "1  Chr1  1100022   G  13  .............\n",
       "2  Chr1  1100023   G  13  ...A...A..A.."
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def readMP(x):\n",
    "    import pandas as pd\n",
    "    mp = pd.read_table(x, header=None)\n",
    "    mp = mp.rename(columns={1:\"snp_pos\", 2:\"ref\", 4:\"pileup\"})\n",
    "    \n",
    "    return(mp)\n",
    "    \n",
    "# mpileup \n",
    "mp1 = readMP('../data/test/case1_pileup.txt')\n",
    "\n",
    "mp1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9225db78",
   "metadata": {},
   "source": [
    "## Read in VCF File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "39f98440",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>NA00001</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>60</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312000</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>60</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312001</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>55</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  #CHROM      POS ID REF ALT  QUAL FILTER INFO FORMAT NA00001\n",
       "0   Chr1  1100023  .   G   A    60   PASS    .     GT     0/1\n",
       "1   Chr2   312000  .   G   A    60   PASS    .     GT     0/1\n",
       "2   Chr2   312001  .   T   C    55   PASS    .     GT     0/1"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The toy VCF\n",
    "import pandas as pd\n",
    "toy_vcf = pd.read_table('../data/toy/toy.vcf')\n",
    "\n",
    "toy_vcf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ba2b86d",
   "metadata": {},
   "source": [
    "## Toy SNPeff Table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "6a05677f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mrna_id</th>\n",
       "      <th>prot_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>effect</th>\n",
       "      <th>snp_cds_pos</th>\n",
       "      <th>codon1_genome_pos</th>\n",
       "      <th>codon2_genome_pos</th>\n",
       "      <th>codon3_genome_pos</th>\n",
       "      <th>snp_aa_pos</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>alt_codon</th>\n",
       "      <th>ref_aa</th>\n",
       "      <th>alt_aa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gene1</td>\n",
       "      <td>mrna1</td>\n",
       "      <td>prot1</td>\n",
       "      <td>+</td>\n",
       "      <td>stop_gain</td>\n",
       "      <td>21</td>\n",
       "      <td>1100021</td>\n",
       "      <td>1100022</td>\n",
       "      <td>1100023</td>\n",
       "      <td>7</td>\n",
       "      <td>TGG</td>\n",
       "      <td>TGA</td>\n",
       "      <td>W</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312000</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gene2</td>\n",
       "      <td>mrna2</td>\n",
       "      <td>prot2</td>\n",
       "      <td>+</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>4</td>\n",
       "      <td>312000</td>\n",
       "      <td>312001</td>\n",
       "      <td>312002</td>\n",
       "      <td>2</td>\n",
       "      <td>GTA</td>\n",
       "      <td>ATA</td>\n",
       "      <td>V</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312001</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>gene2</td>\n",
       "      <td>mrna2</td>\n",
       "      <td>prot2</td>\n",
       "      <td>+</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>5</td>\n",
       "      <td>312000</td>\n",
       "      <td>312001</td>\n",
       "      <td>312002</td>\n",
       "      <td>2</td>\n",
       "      <td>GTA</td>\n",
       "      <td>GCA</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele gene_id mrna_id prot_id strand  \\\n",
       "0   Chr1  1100023          G          A   gene1   mrna1   prot1      +   \n",
       "1   Chr2   312000          G          A   gene2   mrna2   prot2      +   \n",
       "2   Chr2   312001          T          C   gene2   mrna2   prot2      +   \n",
       "\n",
       "          effect  snp_cds_pos  codon1_genome_pos  codon2_genome_pos  \\\n",
       "0      stop_gain           21            1100021            1100022   \n",
       "1  nonsynonymous            4             312000             312001   \n",
       "2  nonsynonymous            5             312000             312001   \n",
       "\n",
       "   codon3_genome_pos  snp_aa_pos ref_codon alt_codon ref_aa alt_aa  \n",
       "0            1100023           7       TGG       TGA      W      *  \n",
       "1             312002           2       GTA       ATA      V      I  \n",
       "2             312002           2       GTA       GCA      V      A  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Toy Table\n",
    "toy_snpeff = pd.read_table('../data/toy/toy.snp_effects.tsv')\n",
    "toy_snpeff"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10b4c427",
   "metadata": {},
   "source": [
    "## Merge the mpileup & snpeff table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "2e778eab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# if pileup information is . at the previous positions, output the reference, else variant\n",
    "def mergeTab(mp, so):\n",
    "    sub_mp = mp[['snp_pos', 'ref', 'pileup']]\n",
    "    mdf = pd.merge(so, sub_mp, on='snp_pos')\n",
    "    return mdf\n",
    "\n",
    "# Merging the two tables\n",
    "x = mergeTab(mp1, toy_snpeff)\n",
    "\n",
    "# dictionary of alternate codons from pileup\n",
    "mp_dict = altDict(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "3e99460c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mrna_id</th>\n",
       "      <th>prot_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>effect</th>\n",
       "      <th>snp_cds_pos</th>\n",
       "      <th>codon1_genome_pos</th>\n",
       "      <th>codon2_genome_pos</th>\n",
       "      <th>codon3_genome_pos</th>\n",
       "      <th>snp_aa_pos</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>alt_codon</th>\n",
       "      <th>ref_aa</th>\n",
       "      <th>alt_aa</th>\n",
       "      <th>ref</th>\n",
       "      <th>pileup</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gene1</td>\n",
       "      <td>mrna1</td>\n",
       "      <td>prot1</td>\n",
       "      <td>+</td>\n",
       "      <td>stop_gain</td>\n",
       "      <td>21</td>\n",
       "      <td>1100021</td>\n",
       "      <td>1100022</td>\n",
       "      <td>1100023</td>\n",
       "      <td>7</td>\n",
       "      <td>TGG</td>\n",
       "      <td>TGA</td>\n",
       "      <td>W</td>\n",
       "      <td>*</td>\n",
       "      <td>G</td>\n",
       "      <td>...A...A..A..</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele gene_id mrna_id prot_id strand  \\\n",
       "0   Chr1  1100023          G          A   gene1   mrna1   prot1      +   \n",
       "\n",
       "      effect  snp_cds_pos  codon1_genome_pos  codon2_genome_pos  \\\n",
       "0  stop_gain           21            1100021            1100022   \n",
       "\n",
       "   codon3_genome_pos  snp_aa_pos ref_codon alt_codon ref_aa alt_aa ref  \\\n",
       "0            1100023           7       TGG       TGA      W      *   G   \n",
       "\n",
       "          pileup  \n",
       "0  ...A...A..A..  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merged tables\n",
    "x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "f62d90c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1100023: 'A'}"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# dictionary of alts from merged tables\n",
    "mp_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "4dc97477",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Using this information, \n",
    "we're going to create our varify codon based on any\n",
    "existing alts on all positions. \n",
    "\n",
    "If the position does not exist, we take from the referenc\n",
    "'''\n",
    "alt_dict = mp_dict\n",
    "# a_unique = getAlt(x) # Replaced with alt_dict[3]\n",
    "\n",
    "# Takes the current position of SNP\n",
    "p3 = int(x['snp_pos'])\n",
    "p2 = p3-1 \n",
    "p1 = p2-1\n",
    "\n",
    "subDict = {p1 : '.', p2 : '.', p3 : '.'}\n",
    "\n",
    "# create a dictionary\n",
    "ref = str(x['ref_codon'][0]) # starts from reference\n",
    "\n",
    "# set up to create varify_alt\n",
    "df = pd.DataFrame()\n",
    "temp = [i for i in ref]\n",
    "snp_dict = {}\n",
    "\n",
    "# Add values from full dictionary if keys are in sub\n",
    "res = {key: alt_dict.get(key, subDict[key]) for key in subDict}\n",
    "res = ''.join(res.values())\n",
    "\n",
    "for n in range(3):\n",
    "    snp_dict['snp_pos'] = int(x['snp_pos'][0])\n",
    "    snp_dict['varify_alt'] = alt_dict[p3] # alt of 'current' position\n",
    "    if res[n] != '.':\n",
    "        temp[n] = res[n]\n",
    "        \n",
    "    snp_dict[\"varify_codon\"] = ''.join(temp)\n",
    "    \n",
    "# snp_dict output: {'snp_pos': 1100023, 'varify_alt': 'A', 'varify_codon': 'TGA'}\n",
    "\n",
    "# Add the dictionary \n",
    "df = df.append(snp_dict, ignore_index=True)\n",
    "df = pd.merge(x, df, on='snp_pos')\n",
    "df = df.apply(axis=1, func=altValidate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "5741b160",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>varify_alt</th>\n",
       "      <th>VARified</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>varify_codon</th>\n",
       "      <th>varify_aa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>Yes</td>\n",
       "      <td>TGG</td>\n",
       "      <td>TGA</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele varify_alt VARified ref_codon  \\\n",
       "0   Chr1  1100023          G          A          A      Yes       TGG   \n",
       "\n",
       "  varify_codon varify_aa  \n",
       "0          TGA         *  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = translate(df)\n",
    "#df = effect(df)\n",
    "\n",
    "reorg = ['chr_id', 'snp_pos', 'ref_allele', 'alt_allele', \n",
    "         'varify_alt', 'VARified', 'ref_codon', 'varify_codon', 'varify_aa', #'effect'\n",
    "        ]\n",
    "\n",
    "df = df[reorg]\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfd7e130",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Case 2: Multi-substitution case\n",
    "\n",
    "In the case of multi substitution, we have alts on pos1 and pos2 of the codon. \n",
    "We want to capture this information, although pos3 has no alt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a670b8dc",
   "metadata": {},
   "source": [
    "## Case2 Mpileup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "26b47f42",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>3</th>\n",
       "      <th>pileup</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312000</td>\n",
       "      <td>G</td>\n",
       "      <td>12</td>\n",
       "      <td>.A.A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312001</td>\n",
       "      <td>T</td>\n",
       "      <td>13</td>\n",
       "      <td>..C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312002</td>\n",
       "      <td>A</td>\n",
       "      <td>13</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0  snp_pos ref   3 pileup\n",
       "0  Chr2   312000   G  12   .A.A\n",
       "1  Chr2   312001   T  13    ..C\n",
       "2  Chr2   312002   A  13    ..."
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mp2 = readMP('../data/toy/toy_mpileup.txt')\n",
    "mp2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b33be7ea",
   "metadata": {},
   "source": [
    "## Case2 Snpeff-like table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "174b97ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mrna_id</th>\n",
       "      <th>prot_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>effect</th>\n",
       "      <th>snp_cds_pos</th>\n",
       "      <th>codon1_genome_pos</th>\n",
       "      <th>codon2_genome_pos</th>\n",
       "      <th>codon3_genome_pos</th>\n",
       "      <th>snp_aa_pos</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>alt_codon</th>\n",
       "      <th>ref_aa</th>\n",
       "      <th>alt_aa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gene1</td>\n",
       "      <td>mrna1</td>\n",
       "      <td>prot1</td>\n",
       "      <td>+</td>\n",
       "      <td>stop_gain</td>\n",
       "      <td>21.0</td>\n",
       "      <td>1100021.0</td>\n",
       "      <td>1100022.0</td>\n",
       "      <td>1100023.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>TGG</td>\n",
       "      <td>TGA</td>\n",
       "      <td>W</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312000</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gene2</td>\n",
       "      <td>mrna2</td>\n",
       "      <td>prot2</td>\n",
       "      <td>+</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>4.0</td>\n",
       "      <td>312000.0</td>\n",
       "      <td>312001.0</td>\n",
       "      <td>312002.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>GTA</td>\n",
       "      <td>ATA</td>\n",
       "      <td>V</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312001</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>gene2</td>\n",
       "      <td>mrna2</td>\n",
       "      <td>prot2</td>\n",
       "      <td>+</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>5.0</td>\n",
       "      <td>312000.0</td>\n",
       "      <td>312001.0</td>\n",
       "      <td>312002.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>GTA</td>\n",
       "      <td>GCA</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312002</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele gene_id mrna_id prot_id strand  \\\n",
       "0   Chr1  1100023          G          A   gene1   mrna1   prot1      +   \n",
       "1   Chr2   312000          G          A   gene2   mrna2   prot2      +   \n",
       "2   Chr2   312001          T          C   gene2   mrna2   prot2      +   \n",
       "3   Chr2   312002          A        NaN     NaN     NaN     NaN    NaN   \n",
       "\n",
       "          effect  snp_cds_pos  codon1_genome_pos  codon2_genome_pos  \\\n",
       "0      stop_gain         21.0          1100021.0          1100022.0   \n",
       "1  nonsynonymous          4.0           312000.0           312001.0   \n",
       "2  nonsynonymous          5.0           312000.0           312001.0   \n",
       "3            NaN          NaN                NaN                NaN   \n",
       "\n",
       "   codon3_genome_pos  snp_aa_pos ref_codon alt_codon ref_aa alt_aa  \n",
       "0          1100023.0         7.0       TGG       TGA      W      *  \n",
       "1           312002.0         2.0       GTA       ATA      V      I  \n",
       "2           312002.0         2.0       GTA       GCA      V      A  \n",
       "3                NaN         NaN       NaN       NaN    NaN    NaN  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "toy_snpeff2 = pd.read_table('../data/toy/toy.snp_effects.tsv') # toy_snpeff\n",
    "toy_snpeff2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe13a88f",
   "metadata": {},
   "source": [
    "## Goal Example\n",
    "| chr_id|snp_pos     | Ref     | Pred_Alt |Varify_Alt|VARified|ref_codon|varify_codon|varify_aa|effect|alt_codons|\n",
    "| -----------| ----------- | ----------- | ----------- | ----------- | ----------- |----------- |----------- |----------- |----------- |---|\n",
    "|Chr2|312002| G      | A       |A|Yes| GTA | ACA |T|*|GTA;ATA;GCA;ACA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef12a42d",
   "metadata": {},
   "source": [
    "## Merging mpileup and snpeff together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "e4f6ee47",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{312000: 'A', 312001: 'C'}"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merging the two tables\n",
    "x2 = mergeTab(mp2, toy_snpeff2)\n",
    "\n",
    "# dictionary of alternate codons from pileup\n",
    "mp_dict2 = altDict(x2)\n",
    "\n",
    "mp_dict2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "id": "9ca5f03a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('A', 'T', 'A'), ('A', 'C', 'A'), ('G', 'T', 'A'), ('G', 'C', 'A')]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['ATA', 'ACA', 'GTA', 'GCA']"
      ]
     },
     "execution_count": 226,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import itertools\n",
    "\n",
    "permutations = list(itertools.product([\"A\",\"G\"], [\"T\",\"C\"], [\"A\"]))\n",
    "\n",
    "print(permutations)\n",
    "\n",
    "[\"\".join(x) for x in permutations]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "932d0c46",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import math\n",
    "\n",
    "math.isnan(mp_dict2[312002])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 291,
   "id": "937c824b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ATA', 'ACA', 'GTA', 'GCA']\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "Using this information, \n",
    "we're going to create our varify codon based on any\n",
    "existing alts on all positions. \n",
    "\n",
    "starting from last position 312002\n",
    "'''\n",
    "import itertools\n",
    "    \n",
    "# Merging the two tables\n",
    "x = x2[x2.snp_pos == 312002]\n",
    "\n",
    "# dictionary of alternate codons from pileup\n",
    "mp_dict2 = altDict(x2)\n",
    "\n",
    "\n",
    "alt_dict = mp_dict2\n",
    "# a_unique = getAlt(x) # Replaced with alt_dict[3]\n",
    "\n",
    "# Takes the current position of SNP\n",
    "p3 = int(x['snp_pos'])\n",
    "p2 = p3-1 \n",
    "p1 = p2-1\n",
    "\n",
    "subDict = {p1 : '.', p2 : '.', p3 : '.'}\n",
    "nt1=[]\n",
    "nt2=[]\n",
    "nt3=[]\n",
    "\n",
    "# set up to create varify_alt\n",
    "ref = str(x2['ref_codon'][0]) # starts from reference\n",
    "df = pd.DataFrame()\n",
    "temp = [i for i in ref] # temp from reference\n",
    "\n",
    "# create a reference dictionary\n",
    "snp_dict = {}\n",
    "# altCombo =\n",
    "\n",
    "# Add values from full dictionary if keys are in sub\n",
    "res = {key: alt_dict.get(key, subDict[key]) for key in subDict}\n",
    "res = ''.join(res.values())\n",
    "\n",
    "# if there are alts in any of the positions, create varify codon\n",
    "for p in [p1, p2, p3]:\n",
    "    if p in list(alt_dict.keys()):\n",
    "        pos_check = True\n",
    "        break\n",
    "\n",
    "# If the position exists in the dictionary, then create varify_alt \n",
    "if pos_check == True: \n",
    "    for n in range(3):\n",
    "        snp_dict['snp_pos'] = int(p3)\n",
    "\n",
    "        # alt of 'current' position if exists\n",
    "        snp_dict['varify_alt'] = alt_dict[p3] if p3 in alt_dict.keys() == True else temp[2]\n",
    "\n",
    "        # replace only values that exist\n",
    "        if res[n] != '.':\n",
    "            temp[n] = res[n]\n",
    "\n",
    "        snp_dict[\"varify_codon\"] = ''.join(temp)\n",
    "        \n",
    "        \n",
    "    nt1.extend([snp_dict[\"varify_codon\"][0], ref[0]])\n",
    "    nt2.extend([snp_dict[\"varify_codon\"][1], ref[1]])\n",
    "    nt3.extend([snp_dict[\"varify_codon\"][2], ref[2]])\n",
    "    \n",
    "    permutations = list(itertools.product(set(nt1), set(nt2), set(nt3)))\n",
    "    snp_dict['alt_codons'] = [\"\".join(x) for x in permutations]\n",
    "    snp_dict['alt_codons'] = ';'.join(snp_dict['alt_codons'])\n",
    "\n",
    "\n",
    "    print([\"\".join(x) for x in permutations])\n",
    "\n",
    "    \n",
    "    \n",
    "# snp_dict output: {'snp_pos': 1100023, 'varify_alt': 'A', 'varify_codon': 'TGA'}\n",
    "\n",
    "# If there are extras, create extra list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 292,
   "id": "e2eec107",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>varify_alt</th>\n",
       "      <th>VARified</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>varify_codon</th>\n",
       "      <th>varify_aa</th>\n",
       "      <th>alt_codons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312002</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACA</td>\n",
       "      <td>T</td>\n",
       "      <td>ATA;ACA;GTA;GCA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele  alt_allele varify_alt VARified  ref_codon  \\\n",
       "0   Chr2   312002          A         NaN          A       No        NaN   \n",
       "\n",
       "  varify_codon varify_aa       alt_codons  \n",
       "0          ACA         T  ATA;ACA;GTA;GCA  "
      ]
     },
     "execution_count": 292,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add the dictionary \n",
    "df = df.append(snp_dict, ignore_index=True)\n",
    "df = pd.merge(x, df, on='snp_pos')\n",
    "df = df.apply(axis=1, func=altValidate)\n",
    "\n",
    "df = translate(df)\n",
    "#df = effect(df)\n",
    "\n",
    "reorg = ['chr_id', 'snp_pos', 'ref_allele', 'alt_allele', \n",
    "         'varify_alt', 'VARified', 'ref_codon', 'varify_codon', 'varify_aa', \"alt_codons\",\n",
    "         #'effect'\n",
    "        ]\n",
    "\n",
    "df = df[reorg]\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9131e49b",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Real World Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 325,
   "id": "a6a3b353",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>sge_Aug_encodev2_2_local</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>820881</td>\n",
       "      <td>0_15</td>\n",
       "      <td>C</td>\n",
       "      <td>CTCTACACACCTGCCTGGCCAGCAGATCCACCCTGTCTACACTACC...</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PRECISE;SVMETHOD=JASMINE;CHR2=chr1;END=820881;...</td>\n",
       "      <td>GT:DR:DV</td>\n",
       "      <td>0|1:18:25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>822462</td>\n",
       "      <td>0_16</td>\n",
       "      <td>GCCTGGCCAGCAGATCCACCCTGTCTATACTACCTG</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PRECISE;SVMETHOD=JASMINE;CHR2=chr1;END=822497;...</td>\n",
       "      <td>GT:DR:DV</td>\n",
       "      <td>1|1:0:37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>839492</td>\n",
       "      <td>0_17</td>\n",
       "      <td>CAAACACACCTGGACACACACACCTAGACACACACACCTGGAC</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PRECISE;SVMETHOD=JASMINE;CHR2=chr1;END=839534;...</td>\n",
       "      <td>GT:DR:DV</td>\n",
       "      <td>1|0:34:49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>876016</td>\n",
       "      <td>0_23</td>\n",
       "      <td>C</td>\n",
       "      <td>CCCCCCACTCTCCTCCCCCACACTCCCCCACACTCCTCCCCCATAC...</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>IMPRECISE;SVMETHOD=JASMINE;CHR2=chr1;END=87601...</td>\n",
       "      <td>GT:DR:DV</td>\n",
       "      <td>0|1:8:19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>876431</td>\n",
       "      <td>0_22</td>\n",
       "      <td>C</td>\n",
       "      <td>CTTCCCCAAACTCCCCCATACTCCCCCATACTCCCCCACATTCCCT...</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PRECISE;SVMETHOD=JASMINE;CHR2=chr1;END=876431;...</td>\n",
       "      <td>GT:DR:DV</td>\n",
       "      <td>0|1:10:25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29843</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155560529</td>\n",
       "      <td>0_219692</td>\n",
       "      <td>TGGTGGTCAGGGTAGGCCTCATTAAGAAAGATGACATTTGATCAAA...</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PRECISE;SVMETHOD=JASMINE;CHR2=chrX;END=1555736...</td>\n",
       "      <td>GT:DR:DV</td>\n",
       "      <td>1|1:0:33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29844</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155638323</td>\n",
       "      <td>0_219696</td>\n",
       "      <td>TATATATATATATATATATATATATATATATATATATATATATATAT</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PRECISE;SVMETHOD=JASMINE;CHR2=chrX;END=1556383...</td>\n",
       "      <td>GT:DR:DV</td>\n",
       "      <td>1|1:0:58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29845</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155685495</td>\n",
       "      <td>0_219699</td>\n",
       "      <td>ATATTATTTCCTCACCCAGGTATTAAGCCTAGTACCCATGAGTTAT...</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PRECISE;SVMETHOD=JASMINE;CHR2=chrX;END=1556855...</td>\n",
       "      <td>GT:DR:DV</td>\n",
       "      <td>1/0:3:28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29846</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155777638</td>\n",
       "      <td>0_219707</td>\n",
       "      <td>T</td>\n",
       "      <td>TATATATAATATATATATATATATATATATATATGTAA</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PRECISE;SVMETHOD=JASMINE;CHR2=chrX;END=1557776...</td>\n",
       "      <td>GT:DR:DV</td>\n",
       "      <td>1|1:3:43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29847</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155873360</td>\n",
       "      <td>0_219715</td>\n",
       "      <td>C</td>\n",
       "      <td>CGCTGTGAAGTCAGCTTGCTGTGAAGTCAGCTTGCTGTGAAGTCAGCTT</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>IMPRECISE;SVMETHOD=JASMINE;CHR2=chrX;END=15587...</td>\n",
       "      <td>GT:DR:DV</td>\n",
       "      <td>1|1:0:20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29848 rows  10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      CHROM        POS        ID  \\\n",
       "0      chr1     820881      0_15   \n",
       "1      chr1     822462      0_16   \n",
       "2      chr1     839492      0_17   \n",
       "3      chr1     876016      0_23   \n",
       "4      chr1     876431      0_22   \n",
       "...     ...        ...       ...   \n",
       "29843  chrX  155560529  0_219692   \n",
       "29844  chrX  155638323  0_219696   \n",
       "29845  chrX  155685495  0_219699   \n",
       "29846  chrX  155777638  0_219707   \n",
       "29847  chrX  155873360  0_219715   \n",
       "\n",
       "                                                     REF  \\\n",
       "0                                                      C   \n",
       "1                   GCCTGGCCAGCAGATCCACCCTGTCTATACTACCTG   \n",
       "2            CAAACACACCTGGACACACACACCTAGACACACACACCTGGAC   \n",
       "3                                                      C   \n",
       "4                                                      C   \n",
       "...                                                  ...   \n",
       "29843  TGGTGGTCAGGGTAGGCCTCATTAAGAAAGATGACATTTGATCAAA...   \n",
       "29844    TATATATATATATATATATATATATATATATATATATATATATATAT   \n",
       "29845  ATATTATTTCCTCACCCAGGTATTAAGCCTAGTACCCATGAGTTAT...   \n",
       "29846                                                  T   \n",
       "29847                                                  C   \n",
       "\n",
       "                                                     ALT QUAL FILTER  \\\n",
       "0      CTCTACACACCTGCCTGGCCAGCAGATCCACCCTGTCTACACTACC...    .   PASS   \n",
       "1                                                      G    .   PASS   \n",
       "2                                                      C    .   PASS   \n",
       "3      CCCCCCACTCTCCTCCCCCACACTCCCCCACACTCCTCCCCCATAC...    .   PASS   \n",
       "4      CTTCCCCAAACTCCCCCATACTCCCCCATACTCCCCCACATTCCCT...    .   PASS   \n",
       "...                                                  ...  ...    ...   \n",
       "29843                                                  T    .   PASS   \n",
       "29844                                                  T    .   PASS   \n",
       "29845                                                  A    .   PASS   \n",
       "29846             TATATATAATATATATATATATATATATATATATGTAA    .   PASS   \n",
       "29847  CGCTGTGAAGTCAGCTTGCTGTGAAGTCAGCTTGCTGTGAAGTCAGCTT    .   PASS   \n",
       "\n",
       "                                                    INFO    FORMAT  \\\n",
       "0      PRECISE;SVMETHOD=JASMINE;CHR2=chr1;END=820881;...  GT:DR:DV   \n",
       "1      PRECISE;SVMETHOD=JASMINE;CHR2=chr1;END=822497;...  GT:DR:DV   \n",
       "2      PRECISE;SVMETHOD=JASMINE;CHR2=chr1;END=839534;...  GT:DR:DV   \n",
       "3      IMPRECISE;SVMETHOD=JASMINE;CHR2=chr1;END=87601...  GT:DR:DV   \n",
       "4      PRECISE;SVMETHOD=JASMINE;CHR2=chr1;END=876431;...  GT:DR:DV   \n",
       "...                                                  ...       ...   \n",
       "29843  PRECISE;SVMETHOD=JASMINE;CHR2=chrX;END=1555736...  GT:DR:DV   \n",
       "29844  PRECISE;SVMETHOD=JASMINE;CHR2=chrX;END=1556383...  GT:DR:DV   \n",
       "29845  PRECISE;SVMETHOD=JASMINE;CHR2=chrX;END=1556855...  GT:DR:DV   \n",
       "29846  PRECISE;SVMETHOD=JASMINE;CHR2=chrX;END=1557776...  GT:DR:DV   \n",
       "29847  IMPRECISE;SVMETHOD=JASMINE;CHR2=chrX;END=15587...  GT:DR:DV   \n",
       "\n",
       "      sge_Aug_encodev2_2_local  \n",
       "0                    0|1:18:25  \n",
       "1                     1|1:0:37  \n",
       "2                    1|0:34:49  \n",
       "3                     0|1:8:19  \n",
       "4                    0|1:10:25  \n",
       "...                        ...  \n",
       "29843                 1|1:0:33  \n",
       "29844                 1|1:0:58  \n",
       "29845                 1/0:3:28  \n",
       "29846                 1|1:3:43  \n",
       "29847                 1|1:0:20  \n",
       "\n",
       "[29848 rows x 10 columns]"
      ]
     },
     "execution_count": 325,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import io\n",
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "def read_vcf(path):\n",
    "    with open(path, 'r') as f:\n",
    "        lines = [l for l in f if not l.startswith('##')]\n",
    "    return pd.read_csv(\n",
    "        io.StringIO(''.join(lines)),\n",
    "        dtype={'#CHROM': str, 'POS': int, 'ID': str, 'REF': str, 'ALT': str,\n",
    "               'QUAL': str, 'FILTER': str, 'INFO': str},\n",
    "        sep='\\t'\n",
    "    ).rename(columns={'#CHROM': 'CHROM'})\n",
    "\n",
    "vcf_file = '../data/ENCFF853AEO.vcf'\n",
    "\n",
    "vcf = read_vcf(vcf_file)\n",
    "\n",
    "vcf['POS'] == 822764"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 329,
   "id": "bf4babf8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>sge_Aug_encodev2_2_local</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>876431</td>\n",
       "      <td>0_22</td>\n",
       "      <td>C</td>\n",
       "      <td>CTTCCCCAAACTCCCCCATACTCCCCCATACTCCCCCACATTCCCT...</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PRECISE;SVMETHOD=JASMINE;CHR2=chr1;END=876431;...</td>\n",
       "      <td>GT:DR:DV</td>\n",
       "      <td>0|1:10:25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>893782</td>\n",
       "      <td>0_25</td>\n",
       "      <td>GAAAAAAAAAAAAAAAAAAAATATATATATATATATAT</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PRECISE;SVMETHOD=JASMINE;CHR2=chr1;END=893819;...</td>\n",
       "      <td>GT:DR:DV</td>\n",
       "      <td>0|1:21:49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>904541</td>\n",
       "      <td>0_26</td>\n",
       "      <td>C</td>\n",
       "      <td>CCCTCCTCCGAACGTGGCCGCCTCCTCCTCCGAACGTGGCCTCCTC...</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>IMPRECISE;SVMETHOD=JASMINE;CHR2=chr1;END=90454...</td>\n",
       "      <td>GT:DR:DV</td>\n",
       "      <td>0|1:26:46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>907838</td>\n",
       "      <td>0_29</td>\n",
       "      <td>G</td>\n",
       "      <td>GTGCCCGGTCCTTCTGACCAGCCGAGAGAGTA</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PRECISE;SVMETHOD=JASMINE;CHR2=chr1;END=907838;...</td>\n",
       "      <td>GT:DR:DV</td>\n",
       "      <td>1|1:0:49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>908918</td>\n",
       "      <td>0_30</td>\n",
       "      <td>ACCGGGTCAGGCCCCCCGGGCGCACCGTTGCTGGTATATGCGGTGA...</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>IMPRECISE;SVMETHOD=JASMINE;CHR2=chr1;END=90936...</td>\n",
       "      <td>GT:DR:DV</td>\n",
       "      <td>1|1:3:50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29843</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155560529</td>\n",
       "      <td>0_219692</td>\n",
       "      <td>TGGTGGTCAGGGTAGGCCTCATTAAGAAAGATGACATTTGATCAAA...</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PRECISE;SVMETHOD=JASMINE;CHR2=chrX;END=1555736...</td>\n",
       "      <td>GT:DR:DV</td>\n",
       "      <td>1|1:0:33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29844</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155638323</td>\n",
       "      <td>0_219696</td>\n",
       "      <td>TATATATATATATATATATATATATATATATATATATATATATATAT</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PRECISE;SVMETHOD=JASMINE;CHR2=chrX;END=1556383...</td>\n",
       "      <td>GT:DR:DV</td>\n",
       "      <td>1|1:0:58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29845</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155685495</td>\n",
       "      <td>0_219699</td>\n",
       "      <td>ATATTATTTCCTCACCCAGGTATTAAGCCTAGTACCCATGAGTTAT...</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PRECISE;SVMETHOD=JASMINE;CHR2=chrX;END=1556855...</td>\n",
       "      <td>GT:DR:DV</td>\n",
       "      <td>1/0:3:28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29846</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155777638</td>\n",
       "      <td>0_219707</td>\n",
       "      <td>T</td>\n",
       "      <td>TATATATAATATATATATATATATATATATATATGTAA</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PRECISE;SVMETHOD=JASMINE;CHR2=chrX;END=1557776...</td>\n",
       "      <td>GT:DR:DV</td>\n",
       "      <td>1|1:3:43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29847</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155873360</td>\n",
       "      <td>0_219715</td>\n",
       "      <td>C</td>\n",
       "      <td>CGCTGTGAAGTCAGCTTGCTGTGAAGTCAGCTTGCTGTGAAGTCAGCTT</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>IMPRECISE;SVMETHOD=JASMINE;CHR2=chrX;END=15587...</td>\n",
       "      <td>GT:DR:DV</td>\n",
       "      <td>1|1:0:20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29073 rows  10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      CHROM        POS        ID  \\\n",
       "4      chr1     876431      0_22   \n",
       "5      chr1     893782      0_25   \n",
       "6      chr1     904541      0_26   \n",
       "7      chr1     907838      0_29   \n",
       "8      chr1     908918      0_30   \n",
       "...     ...        ...       ...   \n",
       "29843  chrX  155560529  0_219692   \n",
       "29844  chrX  155638323  0_219696   \n",
       "29845  chrX  155685495  0_219699   \n",
       "29846  chrX  155777638  0_219707   \n",
       "29847  chrX  155873360  0_219715   \n",
       "\n",
       "                                                     REF  \\\n",
       "4                                                      C   \n",
       "5                 GAAAAAAAAAAAAAAAAAAAATATATATATATATATAT   \n",
       "6                                                      C   \n",
       "7                                                      G   \n",
       "8      ACCGGGTCAGGCCCCCCGGGCGCACCGTTGCTGGTATATGCGGTGA...   \n",
       "...                                                  ...   \n",
       "29843  TGGTGGTCAGGGTAGGCCTCATTAAGAAAGATGACATTTGATCAAA...   \n",
       "29844    TATATATATATATATATATATATATATATATATATATATATATATAT   \n",
       "29845  ATATTATTTCCTCACCCAGGTATTAAGCCTAGTACCCATGAGTTAT...   \n",
       "29846                                                  T   \n",
       "29847                                                  C   \n",
       "\n",
       "                                                     ALT QUAL FILTER  \\\n",
       "4      CTTCCCCAAACTCCCCCATACTCCCCCATACTCCCCCACATTCCCT...    .   PASS   \n",
       "5                                                      G    .   PASS   \n",
       "6      CCCTCCTCCGAACGTGGCCGCCTCCTCCTCCGAACGTGGCCTCCTC...    .   PASS   \n",
       "7                       GTGCCCGGTCCTTCTGACCAGCCGAGAGAGTA    .   PASS   \n",
       "8                                                      A    .   PASS   \n",
       "...                                                  ...  ...    ...   \n",
       "29843                                                  T    .   PASS   \n",
       "29844                                                  T    .   PASS   \n",
       "29845                                                  A    .   PASS   \n",
       "29846             TATATATAATATATATATATATATATATATATATGTAA    .   PASS   \n",
       "29847  CGCTGTGAAGTCAGCTTGCTGTGAAGTCAGCTTGCTGTGAAGTCAGCTT    .   PASS   \n",
       "\n",
       "                                                    INFO    FORMAT  \\\n",
       "4      PRECISE;SVMETHOD=JASMINE;CHR2=chr1;END=876431;...  GT:DR:DV   \n",
       "5      PRECISE;SVMETHOD=JASMINE;CHR2=chr1;END=893819;...  GT:DR:DV   \n",
       "6      IMPRECISE;SVMETHOD=JASMINE;CHR2=chr1;END=90454...  GT:DR:DV   \n",
       "7      PRECISE;SVMETHOD=JASMINE;CHR2=chr1;END=907838;...  GT:DR:DV   \n",
       "8      IMPRECISE;SVMETHOD=JASMINE;CHR2=chr1;END=90936...  GT:DR:DV   \n",
       "...                                                  ...       ...   \n",
       "29843  PRECISE;SVMETHOD=JASMINE;CHR2=chrX;END=1555736...  GT:DR:DV   \n",
       "29844  PRECISE;SVMETHOD=JASMINE;CHR2=chrX;END=1556383...  GT:DR:DV   \n",
       "29845  PRECISE;SVMETHOD=JASMINE;CHR2=chrX;END=1556855...  GT:DR:DV   \n",
       "29846  PRECISE;SVMETHOD=JASMINE;CHR2=chrX;END=1557776...  GT:DR:DV   \n",
       "29847  IMPRECISE;SVMETHOD=JASMINE;CHR2=chrX;END=15587...  GT:DR:DV   \n",
       "\n",
       "      sge_Aug_encodev2_2_local  \n",
       "4                    0|1:10:25  \n",
       "5                    0|1:21:49  \n",
       "6                    0|1:26:46  \n",
       "7                     1|1:0:49  \n",
       "8                     1|1:3:50  \n",
       "...                        ...  \n",
       "29843                 1|1:0:33  \n",
       "29844                 1|1:0:58  \n",
       "29845                 1/0:3:28  \n",
       "29846                 1|1:3:43  \n",
       "29847                 1|1:0:20  \n",
       "\n",
       "[29073 rows x 10 columns]"
      ]
     },
     "execution_count": 329,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf[vcf['POS'] >= 876431]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 332,
   "id": "82cf17f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcffile2 = '../data/ENCFF541HLI.vcf'\n",
    "vcf2 = read_vcf(vcffile2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 335,
   "id": "6c021232",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>sge_Aug_encodev2_2_local</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>779</th>\n",
       "      <td>chr1</td>\n",
       "      <td>876887</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>464.541</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NS=1;DP=15;DPB=15.0;AC=2;AN=2;AF=1.0;RO=0;AO=1...</td>\n",
       "      <td>GT:DP:RO:QR:AO:QA:GL:BX</td>\n",
       "      <td>1/1:15:0:0:15:536:-48.5748,-4.51545,0.0:,GATTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>780</th>\n",
       "      <td>chr1</td>\n",
       "      <td>876904</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>492.048</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NS=1;DP=16;DPB=16.0;AC=2;AN=2;AF=1.0;RO=0;AO=1...</td>\n",
       "      <td>GT:DP:RO:QR:AO:QA:GL:BX</td>\n",
       "      <td>1/1:16:0:0:16:576:-52.175,-4.81648,0.0:,GATTAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>781</th>\n",
       "      <td>chr1</td>\n",
       "      <td>877371</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>275.031</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NS=1;DP=9;DPB=9.0;AC=2;AN=2;AF=1.0;RO=0;AO=13;...</td>\n",
       "      <td>GT:DP:RO:QR:AO:QA:GL:BX</td>\n",
       "      <td>1/1:9:0:0:9:324:-29.5059,-2.70927,0.0:,ACGCCAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>782</th>\n",
       "      <td>chr1</td>\n",
       "      <td>877997</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>87.8305</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NS=1;DP=3;DPB=3.0;AC=1;AN=1;AF=0.5;RO=20;AO=4;...</td>\n",
       "      <td>GT:DP:PS:PQ:PD</td>\n",
       "      <td>1|0:3:591356:100:73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>783</th>\n",
       "      <td>chr1</td>\n",
       "      <td>878005</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>58.4182</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NS=1;DP=31;DPB=31.0;AC=1;AN=2;AF=0.5;RO=22;AO=...</td>\n",
       "      <td>GT:DP:PS:PQ:PD</td>\n",
       "      <td>1|0:31:591356:100:80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4505883</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156030103</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>128.86</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NS=1;DP=19;DPB=19.0;AC=1;AN=2;AF=0.5;RO=15;AO=...</td>\n",
       "      <td>GT:DP:PS:PQ:PD</td>\n",
       "      <td>1|0:19:222469:20:11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4505884</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156030118</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>GT</td>\n",
       "      <td>615.36</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NS=1;DP=48;DPB=86.3333;AC=2;AN=2;AF=1.0;RO=17;...</td>\n",
       "      <td>GT:BX</td>\n",
       "      <td>1/1:TGGGAAGGTGGTAAGC-1_43;CCCGGAAAGCATCATC-1_4...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4505885</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156030138</td>\n",
       "      <td>.</td>\n",
       "      <td>AG</td>\n",
       "      <td>A</td>\n",
       "      <td>497.709</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NS=1;DP=89;DPB=77.0;AC=1;AN=2;AF=0.5;RO=17;AO=...</td>\n",
       "      <td>GT:PS:PQ:PD</td>\n",
       "      <td>0|1:222469:100:49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4505886</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156030145</td>\n",
       "      <td>.</td>\n",
       "      <td>AG</td>\n",
       "      <td>A</td>\n",
       "      <td>72.7247</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NS=1;DP=78;DPB=72.6667;AC=1;AN=2;AF=0.5;RO=44;...</td>\n",
       "      <td>GT:BX:JQ</td>\n",
       "      <td>1/1:CCTAAAGGTTCACGAT-1_43;TCGAACAGTGATGCCC-1_4...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4505887</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156030186</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>424.75</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NS=1;DP=539;DPB=539.0;AC=1;AN=2;AF=0.5;RO=325;...</td>\n",
       "      <td>GT:DP:PS:PQ:PD</td>\n",
       "      <td>1|0:539:222469:100:83</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4475044 rows  10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        CHROM        POS ID REF ALT     QUAL FILTER  \\\n",
       "779      chr1     876887  .   A   G  464.541   PASS   \n",
       "780      chr1     876904  .   C   G  492.048   PASS   \n",
       "781      chr1     877371  .   T   C  275.031   PASS   \n",
       "782      chr1     877997  .   G   T  87.8305   PASS   \n",
       "783      chr1     878005  .   G   T  58.4182   PASS   \n",
       "...       ...        ... ..  ..  ..      ...    ...   \n",
       "4505883  chrX  156030103  .   G   A   128.86   PASS   \n",
       "4505884  chrX  156030118  .   G  GT   615.36   PASS   \n",
       "4505885  chrX  156030138  .  AG   A  497.709   PASS   \n",
       "4505886  chrX  156030145  .  AG   A  72.7247   PASS   \n",
       "4505887  chrX  156030186  .   G   T   424.75   PASS   \n",
       "\n",
       "                                                      INFO  \\\n",
       "779      NS=1;DP=15;DPB=15.0;AC=2;AN=2;AF=1.0;RO=0;AO=1...   \n",
       "780      NS=1;DP=16;DPB=16.0;AC=2;AN=2;AF=1.0;RO=0;AO=1...   \n",
       "781      NS=1;DP=9;DPB=9.0;AC=2;AN=2;AF=1.0;RO=0;AO=13;...   \n",
       "782      NS=1;DP=3;DPB=3.0;AC=1;AN=1;AF=0.5;RO=20;AO=4;...   \n",
       "783      NS=1;DP=31;DPB=31.0;AC=1;AN=2;AF=0.5;RO=22;AO=...   \n",
       "...                                                    ...   \n",
       "4505883  NS=1;DP=19;DPB=19.0;AC=1;AN=2;AF=0.5;RO=15;AO=...   \n",
       "4505884  NS=1;DP=48;DPB=86.3333;AC=2;AN=2;AF=1.0;RO=17;...   \n",
       "4505885  NS=1;DP=89;DPB=77.0;AC=1;AN=2;AF=0.5;RO=17;AO=...   \n",
       "4505886  NS=1;DP=78;DPB=72.6667;AC=1;AN=2;AF=0.5;RO=44;...   \n",
       "4505887  NS=1;DP=539;DPB=539.0;AC=1;AN=2;AF=0.5;RO=325;...   \n",
       "\n",
       "                          FORMAT  \\\n",
       "779      GT:DP:RO:QR:AO:QA:GL:BX   \n",
       "780      GT:DP:RO:QR:AO:QA:GL:BX   \n",
       "781      GT:DP:RO:QR:AO:QA:GL:BX   \n",
       "782               GT:DP:PS:PQ:PD   \n",
       "783               GT:DP:PS:PQ:PD   \n",
       "...                          ...   \n",
       "4505883           GT:DP:PS:PQ:PD   \n",
       "4505884                    GT:BX   \n",
       "4505885              GT:PS:PQ:PD   \n",
       "4505886                 GT:BX:JQ   \n",
       "4505887           GT:DP:PS:PQ:PD   \n",
       "\n",
       "                                  sge_Aug_encodev2_2_local  \n",
       "779      1/1:15:0:0:15:536:-48.5748,-4.51545,0.0:,GATTA...  \n",
       "780      1/1:16:0:0:16:576:-52.175,-4.81648,0.0:,GATTAC...  \n",
       "781      1/1:9:0:0:9:324:-29.5059,-2.70927,0.0:,ACGCCAG...  \n",
       "782                                    1|0:3:591356:100:73  \n",
       "783                                   1|0:31:591356:100:80  \n",
       "...                                                    ...  \n",
       "4505883                                1|0:19:222469:20:11  \n",
       "4505884  1/1:TGGGAAGGTGGTAAGC-1_43;CCCGGAAAGCATCATC-1_4...  \n",
       "4505885                                  0|1:222469:100:49  \n",
       "4505886  1/1:CCTAAAGGTTCACGAT-1_43;TCGAACAGTGATGCCC-1_4...  \n",
       "4505887                              1|0:539:222469:100:83  \n",
       "\n",
       "[4475044 rows x 10 columns]"
      ]
     },
     "execution_count": 335,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf2[vcf2['POS'] >= 876431]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4839a4ec",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'mp2' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mmp2\u001b[49m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'mp2' is not defined"
     ]
    }
   ],
   "source": [
    "mp2"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
