{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1106b087",
   "metadata": {},
   "source": [
    "# VARify"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c87d4722",
   "metadata": {},
   "source": [
    "\n",
    "## Workflow\n",
    "\n",
    "\n",
    "<img src=\"wf.png\" width=\"600\">"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31c60665",
   "metadata": {},
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 479,
   "id": "3b057713",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Utility File\n",
    "\n",
    "# libraries\n",
    "import warnings\n",
    "import pandas as pd\n",
    "\n",
    "asciiDict = {chr(i):(i-33) for i in range(33,74)}\n",
    "\n",
    "def varify(x, mp):\n",
    "    # dictionary of alternate codons from pileup\n",
    "    # alt_dict = altDict(mp)\n",
    "    \n",
    "    # running the script\n",
    "    snp_df = x.apply(axis=1, func=snpDict)\n",
    "    snp_df = pd.concat(list(snp_df))\n",
    "    df = pd.merge(x, snp_df, on='snp_pos')\n",
    "\n",
    "    # Add the dictionary\n",
    "    df = df.apply(axis=1, func=translate)\n",
    "    df = df.apply(axis=1, func=altValidate)\n",
    "    df = df.apply(axis=1, func=codonVarify)\n",
    "    df = df.apply(axis=1, func=aaVarify)\n",
    "    df = df.drop(columns='pileup', axis=1)\n",
    "    \n",
    "    reorg = ['chr_id', 'snp_pos', 'ref_allele', 'alt_allele', 'gene_id', 'mrna_id',\n",
    "             'prot_id', 'strand', 'effect', 'snp_cds_pos', 'codon1_genome_pos',\n",
    "             'codon2_genome_pos', 'codon3_genome_pos', 'snp_aa_pos', 'ref_codon',\n",
    "             'alt_codon', 'ref_aa', 'alt_aa', #'ref', \n",
    "             'pos1_pileup', 'pos2_pileup', 'pos3_pileup',\n",
    "             'varify_codon', 'varify_allele',\n",
    "             'varify_aa', 'nt_VARified', 'codon_VARified', 'aa_VARified','comment']\n",
    "    return df[reorg]\n",
    "\n",
    "\n",
    "\n",
    "def snpDict(x):\n",
    "    # import itertools # for permutations\n",
    "    #print(x.snp_pos)\n",
    "    # Takes the current position of SNP\n",
    "    p3 = int(x.codon3_genome_pos)\n",
    "    p2 = int(x.codon2_genome_pos)\n",
    "    p1 = int(x.codon1_genome_pos)\n",
    "    subDict = {p1 : '.', p2 : '.', p3 : '.'}\n",
    "\n",
    "    \n",
    "    # Fills dictionary with SNP of position from mpileup\n",
    "    res_dict = {key: alt_dict.get(key, subDict[key]) for key in subDict}\n",
    "    res = ''.join(res_dict.values())\n",
    "    \n",
    "    pos_check = False\n",
    "    pos_list = [p1, p2, p3]\n",
    "#     print(pos_list)\n",
    "#     print(subDict)\n",
    "    \n",
    "    # If there are any differences between RNA & DNA, return position\n",
    "    if pos != False: \n",
    "        for p in pos_list:\n",
    "            if p in pos:\n",
    "                # only add p into pos_dict\n",
    "                pos_check = True\n",
    "                pos_dict = {p: alt_dict.get(p)}\n",
    "                #pos_dict = {key:val for key, val in pos_dict.items() if val != '.'}\n",
    "    \n",
    "    # starts from reference \n",
    "    ref = list(x['ref_codon']) \n",
    "    \n",
    "    # create a new dictionary\n",
    "    snp_dict = {}\n",
    "    multi_check = False\n",
    "    comment=''\n",
    "    flag = ''\n",
    "    \n",
    "    # create pileup in df\n",
    "    for i in range(len(pos_list)):\n",
    "        cname = ['pos1_pileup', 'pos2_pileup', 'pos3_pileup']\n",
    "        \n",
    "        try:\n",
    "            snp_dict[cname[i]] = mp[mp.snp_pos == pos_list[i]].pileup.tolist()[0]\n",
    "        except:\n",
    "            # if the pileup does not have info, replace with *\n",
    "            snp_dict[cname[i]] = \"*\"\n",
    "        \n",
    "    \n",
    "    # MultiFlag for complex codons\n",
    "    if (list(res)).count(\".\") < 2:\n",
    "        multi_check = True\n",
    "        if pos_check == False:\n",
    "            # if < 2, then there are multi snps \n",
    "            flag = 'complex_codon'\n",
    "        elif pos_check == True: \n",
    "            flag = f'NA'\n",
    "            comment=f'SNP(s) {pos_dict} only in RNA'\n",
    "        \n",
    "    if multi_check == True: \n",
    "        # if multi_check true, then complex_codon\n",
    "        snp_dict['varify_codon'] = flag\n",
    "        snp_dict['snp_pos'] = x['snp_pos']\n",
    "        current_pos = int(x['snp_pos'])\n",
    "        snp_dict['varify_allele'] = alt_dict[int(current_pos)] # fill with current pos snp\n",
    "        snp_dict['comment'] = comment\n",
    "                                      \n",
    "    else:\n",
    "        current_pos = int(x['snp_pos'])\n",
    "        idx = list(subDict.keys()).index(current_pos)\n",
    "        # get ref\n",
    "        try:\n",
    "            ref[idx] = alt_dict[current_pos]\n",
    "            snp_dict['varify_allele'] = alt_dict[int(current_pos)] # fill with current pos snp\n",
    "        except: \n",
    "            ref[idx] = x.ref_allele\n",
    "            snp_dict['varify_allele'] = x.ref_allele\n",
    "            \n",
    "        \n",
    "        snp_dict['varify_codon'] = ''.join(ref)\n",
    "        snp_dict['snp_pos'] = x['snp_pos']\n",
    "        snp_dict['comment'] = comment\n",
    "\n",
    "        \n",
    "    snp_df = pd.DataFrame([snp_dict])\n",
    "\n",
    "\n",
    "    return snp_df\n",
    "\n",
    "def has_numbers(inputString):\n",
    "    return any(char.isdigit() for char in inputString)\n",
    "\n",
    "def getAlt(x):\n",
    "    indel = has_numbers(x.pileup)\n",
    "    l = list(x['pileup'])\n",
    "    nts = ['A','C','T','G']\n",
    "    \n",
    "    # check if indel, then skip\n",
    "    while indel == True:\n",
    "        for i in l: \n",
    "            if i.isnumeric():\n",
    "                string = ''.join(l)\n",
    "                n = int(i)\n",
    "                idx = l.index(i)\n",
    "                r = range(idx-1,idx+n+1)\n",
    "                pat = ''.join(l[r[0]:r[-1]+1])\n",
    "                string = string.replace(pat, '')\n",
    "        l = [i for i in string]\n",
    "        indel = has_numbers(string)\n",
    "\n",
    "    l = [i for i in l if i.isalpha()]\n",
    "    l = [i for i in l if i.upper() in nts]\n",
    "\n",
    "    if len(l) == 0:\n",
    "        x['varify_allele'] = None\n",
    "    else:   \n",
    "        # Get list of unique\n",
    "        a = [i for i in l]\n",
    "        a_unique = list(set(a)) \n",
    "\n",
    "        # getting max alt\n",
    "        a_unique = pd.DataFrame(a_unique)\n",
    "\n",
    "        # Get the max snp\n",
    "        #nt = a_unique.groupby([0]).apply(lambda x: x.value_counts().index[0])[0]\n",
    "        #nt = nt[0]\n",
    "        nt = pd.value_counts(a_unique[0].values.flatten()).index[0]\n",
    "        x['varify_allele'] = nt.upper()\n",
    "    \n",
    "    return x\n",
    "\n",
    "def read_vcf(path):\n",
    "    with open(path, 'r') as f:\n",
    "        lines = [l for l in f if not l.startswith('##')]\n",
    "    return pd.read_csv(\n",
    "        io.StringIO(''.join(lines)),\n",
    "        dtype={'#CHROM': str, 'POS': int, 'ID': str, 'REF': str, 'ALT': str,\n",
    "               'QUAL': str, 'FILTER': str, 'INFO': str},\n",
    "        sep='\\t'\n",
    "    ).rename(columns={'#CHROM': 'CHROM'})\n",
    "\n",
    "def checkRNADict(x, mp):\n",
    "    x_dict = altDict(x) # for DNA\n",
    "    mp_dict = altDict(mp) # for RNA\n",
    "    \n",
    "    if str(set(mp_dict)-set(x_dict)) == 'set()':\n",
    "        return False, mp_dict\n",
    "    else:\n",
    "        # returns position difference from RNA \n",
    "        return list(set(mp_dict)-set(x_dict)), mp_dict\n",
    "\n",
    "def readMP(x):\n",
    "    import pandas as pd\n",
    "    mp = pd.read_table(x, header=None)\n",
    "    mp = mp.rename(columns={0:\"chr\", 1:\"snp_pos\", 2:\"ref\", 3:\"reads\", 4:\"pileup\"})\n",
    "    mp = mp[mp.reads >= 1]\n",
    "    mp = mp.dropna(axis=0)\n",
    "    return(mp)\n",
    "    \n",
    "\n",
    "def aaVarify(x):\n",
    "    x['aa_VARified'] = 'Yes' if x['varify_aa'] == x['alt_aa'] else \"No\"\n",
    "    return x\n",
    "\n",
    "def codonVarify(x):\n",
    "    x['codon_VARified'] = 'Yes' if x['varify_codon'] == x['alt_codon'] else 'No'\n",
    "    return x\n",
    "\n",
    "def translate(x):\n",
    "    table = {\n",
    "    'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "    'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "    'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "    'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',                \n",
    "    'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "    'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "    'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "    'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "    'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "    'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "    'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "    'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "    'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "    'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "    'TAC':'Y', 'TAT':'Y', 'TAA':'*', 'TAG':'*',\n",
    "    'TGC':'C', 'TGT':'C', 'TGA':'*', 'TGG':'W',}\n",
    "    \n",
    "    codon = x['varify_codon']\n",
    "\n",
    "    if codon not in table.keys():\n",
    "        x['varify_aa'] = \"\"\n",
    "    else:\n",
    "        x['varify_aa'] =table[codon]\n",
    "        \n",
    "    return(x)\n",
    "\n",
    "\n",
    "# For every row in here, we're going to apply this uniqueSNP function\n",
    "\n",
    "def mergeTab(mp, so):\n",
    "    sub_mp = mp[['snp_pos', 'pileup']]\n",
    "    # sub_mp['snp_pos'] = str(sub_mp['snp_pos'])\n",
    "    mdf = pd.merge(so, sub_mp, on='snp_pos')\n",
    "    return mdf\n",
    "\n",
    "\n",
    "def varOnly(a):\n",
    "    '''\n",
    "    purpose: remove non-nt values in pileup\n",
    "    input: string of mpileup \n",
    "    output: nt string\n",
    "    '''\n",
    "    x = [i for i in a if i.upper() in ['A','C','T','G']]\n",
    "    \n",
    "    return x\n",
    "\n",
    "def isMono(ref, temp):\n",
    "    # Check if two snps are monomorphic\n",
    "    c=0\n",
    "    count = [c+1 for snp in temp if snp == ref]\n",
    "    if sum(count) >= 2: \n",
    "        warnings.warn(\"More than 2 monomorphic alleles at position.\")\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "    \n",
    "\n",
    "def altValidate(x):\n",
    "    # Varify alt == Alt? T or F\n",
    "    # Add flag to capture\n",
    "    \n",
    "    x['nt_VARified'] = 'Yes' if x['alt_allele'] == x['varify_allele'] else 'No'\n",
    "    \n",
    "    return x\n",
    "\n",
    "\n",
    "\n",
    "def effect(x):\n",
    "    # missense\n",
    "    # nonsense\n",
    "    x['effect'] = '_'\n",
    "    \n",
    "\n",
    "###################Older, probably outdated############################\n",
    "\n",
    "# Codon Combos\n",
    "import numpy as np\n",
    "import math \n",
    "\n",
    "\n",
    "\n",
    "def altDict(x):\n",
    "    \n",
    "    '''\n",
    "    Takes mPileup\n",
    "    Out: mpileup dictionary with snp_pos and varify_allele\n",
    "    '''\n",
    "    \n",
    "    # dataframe of mpileup goes through each pileup with apply\n",
    "    x = x.apply(axis=1, func=getAlt)\n",
    "    alt_dict = dict(zip((x['snp_pos']), x['varify_allele']))\n",
    "    clean_dict = {k: alt_dict[k] for k in alt_dict if pd.isna(alt_dict[k]) == False}\n",
    "    \n",
    "    \n",
    "    return clean_dict\n",
    "    \n",
    "\n",
    "def flatten(d, parent_key='', sep='_'):\n",
    "    import collections\n",
    "    items = []\n",
    "    for k, v in d.items():\n",
    "        new_key = parent_key + sep + k if parent_key else k\n",
    "        if isinstance(v, collections.MutableMapping):\n",
    "            items.extend(flatten(v, new_key, sep=sep).items())\n",
    "        else:\n",
    "            items.append((new_key, v))\n",
    "    return dict(items)\n",
    "\n",
    "# import itertools\n",
    "\n",
    "# permutations = list(itertools.product([\"A\",\"G\"], [\"T\",\"C\"], [\"A\"]))\n",
    "\n",
    "# print(permutations)\n",
    "\n",
    "# [\"\".join(x) for x in permutations]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "302e41ef",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Case 1\n",
    "\n",
    "Consider the following mutation that introduces a G->A mutation.\n",
    " \n",
    "VCF:\n",
    "Chr1\t1100023      \t.       \tG     \tA      \t60    \tPASS\t.       \tGT    \t0/1\n",
    " \n",
    "* Genotype: G/A\n",
    "* Reference base: G\n",
    "* Alternate base: A\n",
    "* Predicted reference codon: TGG (tryptophan)\n",
    "* Predicted alternate codon: TGA (premature termination codon)\n",
    " \n",
    "A tool like snpEff can be used to predict the effect of the mutation using a reference genome annotation (e.g., GFF). In this case the mutation occurs in a coding exon at third position of a codon that codes for tryptophan in the reference proteome and a premature termination codon in the case of the alternate allele. snpEff will determine the reference and alternate codon and return the two codons and their translation.\n",
    " \n",
    "The predicted effect is therefore a nonsense mutation.\n",
    "\n",
    "## Goal Example\n",
    "| chr_id|snp_pos     | Ref     | Pred_Alt |Varify_Alt|VARified|ref_codon|varify_codon|aa|effect|\n",
    "| -----------| ----------- | ----------- | ----------- | ----------- | ----------- |----------- |----------- |----------- |----------- |\n",
    "|Chr1|1100023| G      | A       |A|Yes| TGG | TGA |*|stop_gain|"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e54cc130",
   "metadata": {},
   "source": [
    "## Read in the Toy Mpileup File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 462,
   "id": "76adbb9d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>reads</th>\n",
       "      <th>pileup</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100021</td>\n",
       "      <td>T</td>\n",
       "      <td>12</td>\n",
       "      <td>.............</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100022</td>\n",
       "      <td>G</td>\n",
       "      <td>13</td>\n",
       "      <td>.............</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>G</td>\n",
       "      <td>13</td>\n",
       "      <td>...A...A..A..</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr  snp_pos ref  reads         pileup\n",
       "0  Chr1  1100021   T     12  .............\n",
       "1  Chr1  1100022   G     13  .............\n",
       "2  Chr1  1100023   G     13  ...A...A..A.."
      ]
     },
     "execution_count": 462,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# mpileup \n",
    "mp1 = readMP('../data/test/case1_pileup.txt')\n",
    "\n",
    "mp1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ddaf25c",
   "metadata": {},
   "source": [
    "## Read in VCF File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 464,
   "id": "39f98440",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>NA00001</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>60</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312000</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>60</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312001</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>55</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  #CHROM      POS ID REF ALT  QUAL FILTER INFO FORMAT NA00001\n",
       "0   Chr1  1100023  .   G   A    60   PASS    .     GT     0/1\n",
       "1   Chr2   312000  .   G   A    60   PASS    .     GT     0/1\n",
       "2   Chr2   312001  .   T   C    55   PASS    .     GT     0/1"
      ]
     },
     "execution_count": 464,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The toy VCF\n",
    "toy_vcf = pd.read_table('../data/toy/toy.vcf')\n",
    "\n",
    "toy_vcf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66da6f0e",
   "metadata": {},
   "source": [
    "## Toy SNPeff Table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 465,
   "id": "6a05677f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mrna_id</th>\n",
       "      <th>prot_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>effect</th>\n",
       "      <th>snp_cds_pos</th>\n",
       "      <th>codon1_genome_pos</th>\n",
       "      <th>codon2_genome_pos</th>\n",
       "      <th>codon3_genome_pos</th>\n",
       "      <th>snp_aa_pos</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>alt_codon</th>\n",
       "      <th>ref_aa</th>\n",
       "      <th>alt_aa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gene1</td>\n",
       "      <td>mrna1</td>\n",
       "      <td>prot1</td>\n",
       "      <td>+</td>\n",
       "      <td>stop_gain</td>\n",
       "      <td>21</td>\n",
       "      <td>1100021</td>\n",
       "      <td>1100022</td>\n",
       "      <td>1100023</td>\n",
       "      <td>7</td>\n",
       "      <td>TGG</td>\n",
       "      <td>TGA</td>\n",
       "      <td>W</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312000</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gene2</td>\n",
       "      <td>mrna2</td>\n",
       "      <td>prot2</td>\n",
       "      <td>+</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>4</td>\n",
       "      <td>312000</td>\n",
       "      <td>312001</td>\n",
       "      <td>312002</td>\n",
       "      <td>2</td>\n",
       "      <td>GTA</td>\n",
       "      <td>ATA</td>\n",
       "      <td>V</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312001</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>gene2</td>\n",
       "      <td>mrna2</td>\n",
       "      <td>prot2</td>\n",
       "      <td>+</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>5</td>\n",
       "      <td>312000</td>\n",
       "      <td>312001</td>\n",
       "      <td>312002</td>\n",
       "      <td>2</td>\n",
       "      <td>GTA</td>\n",
       "      <td>GCA</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele gene_id mrna_id prot_id strand  \\\n",
       "0   Chr1  1100023          G          A   gene1   mrna1   prot1      +   \n",
       "1   Chr2   312000          G          A   gene2   mrna2   prot2      +   \n",
       "2   Chr2   312001          T          C   gene2   mrna2   prot2      +   \n",
       "\n",
       "          effect  snp_cds_pos  codon1_genome_pos  codon2_genome_pos  \\\n",
       "0      stop_gain           21            1100021            1100022   \n",
       "1  nonsynonymous            4             312000             312001   \n",
       "2  nonsynonymous            5             312000             312001   \n",
       "\n",
       "   codon3_genome_pos  snp_aa_pos ref_codon alt_codon ref_aa alt_aa  \n",
       "0            1100023           7       TGG       TGA      W      *  \n",
       "1             312002           2       GTA       ATA      V      I  \n",
       "2             312002           2       GTA       GCA      V      A  "
      ]
     },
     "execution_count": 465,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Toy Table\n",
    "toy_snpeff = pd.read_table('../data/toy/toy.snp_effects.tsv')\n",
    "toy_snpeff"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db940360",
   "metadata": {},
   "source": [
    "## Merge the mpileup & snpeff table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 466,
   "id": "2e778eab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mrna_id</th>\n",
       "      <th>prot_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>effect</th>\n",
       "      <th>snp_cds_pos</th>\n",
       "      <th>...</th>\n",
       "      <th>pos1_pileup</th>\n",
       "      <th>pos2_pileup</th>\n",
       "      <th>pos3_pileup</th>\n",
       "      <th>varify_codon</th>\n",
       "      <th>varify_allele</th>\n",
       "      <th>varify_aa</th>\n",
       "      <th>nt_VARified</th>\n",
       "      <th>codon_VARified</th>\n",
       "      <th>aa_VARified</th>\n",
       "      <th>comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gene1</td>\n",
       "      <td>mrna1</td>\n",
       "      <td>prot1</td>\n",
       "      <td>+</td>\n",
       "      <td>stop_gain</td>\n",
       "      <td>21</td>\n",
       "      <td>...</td>\n",
       "      <td>.............</td>\n",
       "      <td>.............</td>\n",
       "      <td>...A...A..A..</td>\n",
       "      <td>TGA</td>\n",
       "      <td>A</td>\n",
       "      <td>*</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele gene_id mrna_id prot_id strand  \\\n",
       "0   Chr1  1100023          G          A   gene1   mrna1   prot1      +   \n",
       "\n",
       "      effect  snp_cds_pos  ...    pos1_pileup    pos2_pileup    pos3_pileup  \\\n",
       "0  stop_gain           21  ...  .............  .............  ...A...A..A..   \n",
       "\n",
       "   varify_codon varify_allele varify_aa nt_VARified codon_VARified  \\\n",
       "0           TGA             A         *         Yes            Yes   \n",
       "\n",
       "  aa_VARified comment  \n",
       "0         Yes          \n",
       "\n",
       "[1 rows x 28 columns]"
      ]
     },
     "execution_count": 466,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merging the two tables\n",
    "mp = mp1\n",
    "x = mergeTab(mp1, toy_snpeff)\n",
    "\n",
    "# dictionary of alternate codons from pileup\n",
    "pos, alt_dict = checkRNADict(x, mp)\n",
    "varify(x, mp)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0927cefa",
   "metadata": {},
   "source": [
    "# Case 1.b and Case 1.c \n",
    "\n",
    "To determine if the codon is VARified, VARify will look at the two other positions in the same codon (the ‘non-focal positions’) and determine if ONLY the reference nucleotide is observed in RNA. \n",
    "\n",
    "For example, if both non-focal positions are monomorphic for the reference nucleotide in the RNA mpileup (i.e., the reference nucleotide from VCF in snpeff table), then the codon is verified and the “codon_VARified” column will be set to ‘yes’. The Case 1 example above would have the codon_VARified column set to ‘yes’.\n",
    "\n",
    "All other situations will have the “codon_VARified” column set to “no”. These cases could include cases in which one (or both) non-focal positions have two or more nucleotides in the RNA mpileup (note: this is NOT shown in Case 1 example above). As we discussed, this could occur frequently because of sequencing errors. We will case this “Case 1b”. \n",
    "\n",
    "Another situation might be if one (or both) non-focal positions are monomorphic for a non-reference nucleotide. This also would be inconsistent with the codon reported in alt_codon column and the “codon_VARified” column should be reported as “no”. We will call this “Case 1c”\n",
    "\n",
    "Case 1b:\n",
    "\n",
    "Given: VCF same as Case 1.\n",
    "RNA Mpileup:\n",
    "|chromosome_id|snp_pos|ref_allele|pos|pileup||\n",
    "|-----|-----|-----|-----|-----|-----|\n",
    "|Chr1|1100021|T|12|.......|other columns|\n",
    "|Chr1|1100022|G|13|..TT..|other columns|\n",
    "|Chr1|1100023|G|13|..A..A..A..|other columns|\n",
    "\n",
    "\n",
    "Case 1c:\n",
    "\n",
    "Given: VCF same as Case 1\n",
    "RNA Mpileup:\n",
    "|chromosome_id|snp_pos|ref_allele|pos|pileup||\n",
    "|-----|-----|-----|-----|-----|-----|\n",
    "|Chr1|1100021|T|12|CCCCCC|other columns|\n",
    "|Chr1|1100022|G|13|...|other columns|\n",
    "|Chr1|1100023|G|13|..A..A..A..|other columns|\n",
    "\n",
    "*note: in Case 1c, the mpileup does not have the reference nucleotide at non-focal position 1100021. Therefore the alt_codon (TGG) is not verified (even though both non-focal positions are monomorphic in RNA mpileup.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f44e47b1",
   "metadata": {},
   "source": [
    "## Goal Example\n",
    "| chr_id|snp_pos     | Ref     | Pred_Alt |Varify_Alt|VARified|ref_codon|varify_codon|pred_codon|varify_aa|effect|alt_codons|comment|\n",
    "| ---| --- | --- | --- | --- | --- |--- |--- |--- |--- |---|--|--|\n",
    "|Chr2|312000| G      | A       |A|Yes| GTA | ATA |T|*|||Pos1: A; Pos2; C\n",
    "|Chr2|312001| T      | C       |C|Yes| GTA | GCA |T|*|||"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 467,
   "id": "1fcd619b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>reads</th>\n",
       "      <th>pileup</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100021</td>\n",
       "      <td>T</td>\n",
       "      <td>12</td>\n",
       "      <td>......,.,.,.</td>\n",
       "      <td>&lt;&lt;&lt;&lt;&lt;;&lt;&lt;&lt;;&lt;&lt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100022</td>\n",
       "      <td>G</td>\n",
       "      <td>13</td>\n",
       "      <td>.,.,..TT.,...</td>\n",
       "      <td>=&amp;;&lt;&lt;&lt;;&lt;&lt;&lt;&lt;&lt;&lt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>G</td>\n",
       "      <td>13</td>\n",
       "      <td>...A...A..A,.</td>\n",
       "      <td>==&lt;;;;&lt;&lt;&lt;&lt;&lt;&lt;&lt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr  snp_pos ref  reads         pileup              5\n",
       "0  Chr1  1100021   T     12   ......,.,.,.   <<<<<;<<<;<<\n",
       "1  Chr1  1100022   G     13  .,.,..TT.,...  =&;<<<;<<<<<<\n",
       "2  Chr1  1100023   G     13  ...A...A..A,.  ==<;;;<<<<<<<"
      ]
     },
     "execution_count": 467,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mp_1b = readMP('../data/toy/toy.case1b.pileup')\n",
    "mp_1c = readMP('../data/toy/toy.case1c.pileup')\n",
    "\n",
    "mp_1b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 468,
   "id": "7b936536",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>reads</th>\n",
       "      <th>pileup</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100021</td>\n",
       "      <td>T</td>\n",
       "      <td>12</td>\n",
       "      <td>CCCCCCCCCCCC</td>\n",
       "      <td>&lt;&lt;&lt;&lt;&lt;;&lt;&lt;&lt;;&lt;&lt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100022</td>\n",
       "      <td>G</td>\n",
       "      <td>13</td>\n",
       "      <td>.,.,.....,...</td>\n",
       "      <td>=&amp;;&lt;&lt;&lt;;&lt;&lt;&lt;&lt;&lt;&lt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>G</td>\n",
       "      <td>13</td>\n",
       "      <td>...A...A..A,.</td>\n",
       "      <td>==&lt;;;;&lt;&lt;&lt;&lt;&lt;&lt;&lt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr  snp_pos ref  reads         pileup              5\n",
       "0  Chr1  1100021   T     12   CCCCCCCCCCCC   <<<<<;<<<;<<\n",
       "1  Chr1  1100022   G     13  .,.,.....,...  =&;<<<;<<<<<<\n",
       "2  Chr1  1100023   G     13  ...A...A..A,.  ==<;;;<<<<<<<"
      ]
     },
     "execution_count": 468,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mp_1c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 469,
   "id": "88b1bf1e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mrna_id</th>\n",
       "      <th>prot_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>effect</th>\n",
       "      <th>snp_cds_pos</th>\n",
       "      <th>codon1_genome_pos</th>\n",
       "      <th>codon2_genome_pos</th>\n",
       "      <th>codon3_genome_pos</th>\n",
       "      <th>snp_aa_pos</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>alt_codon</th>\n",
       "      <th>ref_aa</th>\n",
       "      <th>alt_aa</th>\n",
       "      <th>pileup</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gene1</td>\n",
       "      <td>mrna1</td>\n",
       "      <td>prot1</td>\n",
       "      <td>+</td>\n",
       "      <td>stop_gain</td>\n",
       "      <td>21</td>\n",
       "      <td>1100021</td>\n",
       "      <td>1100022</td>\n",
       "      <td>1100023</td>\n",
       "      <td>7</td>\n",
       "      <td>TGG</td>\n",
       "      <td>TGA</td>\n",
       "      <td>W</td>\n",
       "      <td>*</td>\n",
       "      <td>...A...A..A,.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele gene_id mrna_id prot_id strand  \\\n",
       "0   Chr1  1100023          G          A   gene1   mrna1   prot1      +   \n",
       "\n",
       "      effect  snp_cds_pos  codon1_genome_pos  codon2_genome_pos  \\\n",
       "0  stop_gain           21            1100021            1100022   \n",
       "\n",
       "   codon3_genome_pos  snp_aa_pos ref_codon alt_codon ref_aa alt_aa  \\\n",
       "0            1100023           7       TGG       TGA      W      *   \n",
       "\n",
       "          pileup  \n",
       "0  ...A...A..A,.  "
      ]
     },
     "execution_count": 469,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_1c = mergeTab(mp_1c, toy_snpeff)\n",
    "x_1c"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "437e65ab",
   "metadata": {},
   "source": [
    "There are two issues\n",
    "1. When there are two snps in one codon found in both DNA & RNA\n",
    "    * Solution: output complex codon\n",
    "2. When there are two snps found in RNA only \n",
    "    * Solution: Output comment - POS2 in RNA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e99649cc",
   "metadata": {},
   "source": [
    "# Case 1b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 480,
   "id": "5540a4bd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mrna_id</th>\n",
       "      <th>prot_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>effect</th>\n",
       "      <th>snp_cds_pos</th>\n",
       "      <th>...</th>\n",
       "      <th>pos1_pileup</th>\n",
       "      <th>pos2_pileup</th>\n",
       "      <th>pos3_pileup</th>\n",
       "      <th>varify_codon</th>\n",
       "      <th>varify_allele</th>\n",
       "      <th>varify_aa</th>\n",
       "      <th>nt_VARified</th>\n",
       "      <th>codon_VARified</th>\n",
       "      <th>aa_VARified</th>\n",
       "      <th>comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gene1</td>\n",
       "      <td>mrna1</td>\n",
       "      <td>prot1</td>\n",
       "      <td>+</td>\n",
       "      <td>stop_gain</td>\n",
       "      <td>21</td>\n",
       "      <td>...</td>\n",
       "      <td>......,.,.,.</td>\n",
       "      <td>.,.,..TT.,...</td>\n",
       "      <td>...A...A..A,.</td>\n",
       "      <td>NA</td>\n",
       "      <td>A</td>\n",
       "      <td></td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>SNP(s) {1100022: 'T'} only in RNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele gene_id mrna_id prot_id strand  \\\n",
       "0   Chr1  1100023          G          A   gene1   mrna1   prot1      +   \n",
       "\n",
       "      effect  snp_cds_pos  ...   pos1_pileup    pos2_pileup    pos3_pileup  \\\n",
       "0  stop_gain           21  ...  ......,.,.,.  .,.,..TT.,...  ...A...A..A,.   \n",
       "\n",
       "   varify_codon varify_allele varify_aa nt_VARified codon_VARified  \\\n",
       "0            NA             A                   Yes             No   \n",
       "\n",
       "  aa_VARified                            comment  \n",
       "0          No  SNP(s) {1100022: 'T'} only in RNA  \n",
       "\n",
       "[1 rows x 28 columns]"
      ]
     },
     "execution_count": 480,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_1b = mergeTab(mp_1b, toy_snpeff)\n",
    "\n",
    "x = x_1b \n",
    "mp = mp_1b\n",
    "\n",
    "pos, alt_dict = checkRNADict(x, mp)\n",
    "\n",
    "varify(x, mp)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bf424fa",
   "metadata": {},
   "source": [
    "# Case 1c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 481,
   "id": "d438d49b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mrna_id</th>\n",
       "      <th>prot_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>effect</th>\n",
       "      <th>snp_cds_pos</th>\n",
       "      <th>...</th>\n",
       "      <th>pos1_pileup</th>\n",
       "      <th>pos2_pileup</th>\n",
       "      <th>pos3_pileup</th>\n",
       "      <th>varify_codon</th>\n",
       "      <th>varify_allele</th>\n",
       "      <th>varify_aa</th>\n",
       "      <th>nt_VARified</th>\n",
       "      <th>codon_VARified</th>\n",
       "      <th>aa_VARified</th>\n",
       "      <th>comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gene1</td>\n",
       "      <td>mrna1</td>\n",
       "      <td>prot1</td>\n",
       "      <td>+</td>\n",
       "      <td>stop_gain</td>\n",
       "      <td>21</td>\n",
       "      <td>...</td>\n",
       "      <td>CCCCCCCCCCCC</td>\n",
       "      <td>.,.,.....,...</td>\n",
       "      <td>...A...A..A,.</td>\n",
       "      <td>NA</td>\n",
       "      <td>A</td>\n",
       "      <td></td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>SNP(s) {1100021: 'C'} only in RNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele gene_id mrna_id prot_id strand  \\\n",
       "0   Chr1  1100023          G          A   gene1   mrna1   prot1      +   \n",
       "\n",
       "      effect  snp_cds_pos  ...   pos1_pileup    pos2_pileup    pos3_pileup  \\\n",
       "0  stop_gain           21  ...  CCCCCCCCCCCC  .,.,.....,...  ...A...A..A,.   \n",
       "\n",
       "   varify_codon varify_allele varify_aa nt_VARified codon_VARified  \\\n",
       "0            NA             A                   Yes             No   \n",
       "\n",
       "  aa_VARified                            comment  \n",
       "0          No  SNP(s) {1100021: 'C'} only in RNA  \n",
       "\n",
       "[1 rows x 28 columns]"
      ]
     },
     "execution_count": 481,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_1c = mergeTab(mp_1c, toy_snpeff)\n",
    "\n",
    "x = x_1c \n",
    "mp = mp_1c\n",
    "\n",
    "pos, alt_dict = checkRNADict(x, mp)\n",
    "\n",
    "# running the script\n",
    "varify(x, mp)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfd7e130",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Case 2: Multi-substitution case\n",
    "\n",
    "In the case of multi substitution, we have alts on pos1 and pos2 of the codon. \n",
    "We want to capture this information, although pos3 has no alt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a87d112e",
   "metadata": {},
   "source": [
    "## Case2 Mpileup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 472,
   "id": "26b47f42",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>reads</th>\n",
       "      <th>pileup</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312000</td>\n",
       "      <td>G</td>\n",
       "      <td>4</td>\n",
       "      <td>.A.A</td>\n",
       "      <td>&lt;&lt;&lt;;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312001</td>\n",
       "      <td>T</td>\n",
       "      <td>3</td>\n",
       "      <td>..C</td>\n",
       "      <td>&lt;;&lt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312002</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>&lt;&lt;&lt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr  snp_pos ref  reads pileup     5\n",
       "0  Chr2   312000   G      4   .A.A  <<<;\n",
       "1  Chr2   312001   T      3    ..C   <;<\n",
       "2  Chr2   312002   A      3    ...   <<<"
      ]
     },
     "execution_count": 472,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mp2 = readMP('../data/toy/toy.case2.pileup')\n",
    "mp2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be421564",
   "metadata": {},
   "source": [
    "## Case2 Snpeff-like table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 473,
   "id": "174b97ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mrna_id</th>\n",
       "      <th>prot_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>effect</th>\n",
       "      <th>snp_cds_pos</th>\n",
       "      <th>codon1_genome_pos</th>\n",
       "      <th>codon2_genome_pos</th>\n",
       "      <th>codon3_genome_pos</th>\n",
       "      <th>snp_aa_pos</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>alt_codon</th>\n",
       "      <th>ref_aa</th>\n",
       "      <th>alt_aa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gene1</td>\n",
       "      <td>mrna1</td>\n",
       "      <td>prot1</td>\n",
       "      <td>+</td>\n",
       "      <td>stop_gain</td>\n",
       "      <td>21</td>\n",
       "      <td>1100021</td>\n",
       "      <td>1100022</td>\n",
       "      <td>1100023</td>\n",
       "      <td>7</td>\n",
       "      <td>TGG</td>\n",
       "      <td>TGA</td>\n",
       "      <td>W</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312000</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gene2</td>\n",
       "      <td>mrna2</td>\n",
       "      <td>prot2</td>\n",
       "      <td>+</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>4</td>\n",
       "      <td>312000</td>\n",
       "      <td>312001</td>\n",
       "      <td>312002</td>\n",
       "      <td>2</td>\n",
       "      <td>GTA</td>\n",
       "      <td>ATA</td>\n",
       "      <td>V</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312001</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>gene2</td>\n",
       "      <td>mrna2</td>\n",
       "      <td>prot2</td>\n",
       "      <td>+</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>5</td>\n",
       "      <td>312000</td>\n",
       "      <td>312001</td>\n",
       "      <td>312002</td>\n",
       "      <td>2</td>\n",
       "      <td>GTA</td>\n",
       "      <td>GCA</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele gene_id mrna_id prot_id strand  \\\n",
       "0   Chr1  1100023          G          A   gene1   mrna1   prot1      +   \n",
       "1   Chr2   312000          G          A   gene2   mrna2   prot2      +   \n",
       "2   Chr2   312001          T          C   gene2   mrna2   prot2      +   \n",
       "\n",
       "          effect  snp_cds_pos  codon1_genome_pos  codon2_genome_pos  \\\n",
       "0      stop_gain           21            1100021            1100022   \n",
       "1  nonsynonymous            4             312000             312001   \n",
       "2  nonsynonymous            5             312000             312001   \n",
       "\n",
       "   codon3_genome_pos  snp_aa_pos ref_codon alt_codon ref_aa alt_aa  \n",
       "0            1100023           7       TGG       TGA      W      *  \n",
       "1             312002           2       GTA       ATA      V      I  \n",
       "2             312002           2       GTA       GCA      V      A  "
      ]
     },
     "execution_count": 473,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "toy_snpeff2 = pd.read_table('../data/toy/toy.snp_effects.tsv') # toy_snpeff\n",
    "toy_snpeff2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe13a88f",
   "metadata": {},
   "source": [
    "## Goal Example\n",
    "| chr_id|snp_pos     | Ref     | Pred_Alt |Varify_Alt|VARified|ref_codon|varify_codon|pred_codon|varify_aa|effect|alt_codons|flag|\n",
    "| ---| --- | --- | --- | --- | --- |--- |--- |--- |--- |---|--|--|\n",
    "|Chr2|312000| G      | A       |A|Yes| GTA | ATA |T|*|||\n",
    "|Chr2|312001| T      | C       |C|Yes| GTA | GCA |T|*|||"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f0bdfb5",
   "metadata": {},
   "source": [
    "## Merging mpileup and snpeff together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 474,
   "id": "bf819b82",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merging the two tables\n",
    "x2 = mergeTab(mp2, toy_snpeff2)\n",
    "pos, alt_dict = checkRNADict(x2, mp2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "327cfdb8",
   "metadata": {},
   "source": [
    "## Running Case 2\n",
    "\n",
    "Things to Add: \n",
    "    * A control if nothing exists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 475,
   "id": "696b23c0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mrna_id</th>\n",
       "      <th>prot_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>effect</th>\n",
       "      <th>snp_cds_pos</th>\n",
       "      <th>...</th>\n",
       "      <th>pos1_pileup</th>\n",
       "      <th>pos2_pileup</th>\n",
       "      <th>pos3_pileup</th>\n",
       "      <th>varify_codon</th>\n",
       "      <th>varify_allele</th>\n",
       "      <th>varify_aa</th>\n",
       "      <th>nt_VARified</th>\n",
       "      <th>codon_VARified</th>\n",
       "      <th>aa_VARified</th>\n",
       "      <th>comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312000</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gene2</td>\n",
       "      <td>mrna2</td>\n",
       "      <td>prot2</td>\n",
       "      <td>+</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>complex_codon</td>\n",
       "      <td>A</td>\n",
       "      <td></td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312001</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>gene2</td>\n",
       "      <td>mrna2</td>\n",
       "      <td>prot2</td>\n",
       "      <td>+</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>complex_codon</td>\n",
       "      <td>C</td>\n",
       "      <td></td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele gene_id mrna_id prot_id strand  \\\n",
       "0   Chr2   312000          G          A   gene2   mrna2   prot2      +   \n",
       "1   Chr2   312001          T          C   gene2   mrna2   prot2      +   \n",
       "\n",
       "          effect  snp_cds_pos  ...  pos1_pileup  pos2_pileup  pos3_pileup  \\\n",
       "0  nonsynonymous            4  ...            *            *            *   \n",
       "1  nonsynonymous            5  ...            *            *            *   \n",
       "\n",
       "    varify_codon varify_allele varify_aa nt_VARified codon_VARified  \\\n",
       "0  complex_codon             A                   Yes             No   \n",
       "1  complex_codon             C                   Yes             No   \n",
       "\n",
       "  aa_VARified comment  \n",
       "0          No          \n",
       "1          No          \n",
       "\n",
       "[2 rows x 28 columns]"
      ]
     },
     "execution_count": 475,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "varify(x2, mp2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c632d698",
   "metadata": {},
   "source": [
    "# Case 2.5 - Both Case 1 and Case 2 on one "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 482,
   "id": "250a0be5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>reads</th>\n",
       "      <th>pileup</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100021</td>\n",
       "      <td>T</td>\n",
       "      <td>12</td>\n",
       "      <td>......,.,.,.</td>\n",
       "      <td>&lt;&lt;&lt;&lt;&lt;;&lt;&lt;&lt;;&lt;&lt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100022</td>\n",
       "      <td>G</td>\n",
       "      <td>13</td>\n",
       "      <td>.,.,..TT.,...</td>\n",
       "      <td>=&amp;;&lt;&lt;&lt;;&lt;&lt;&lt;&lt;&lt;&lt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>G</td>\n",
       "      <td>13</td>\n",
       "      <td>...A...A..A,.</td>\n",
       "      <td>==&lt;;;;&lt;&lt;&lt;&lt;&lt;&lt;&lt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312000</td>\n",
       "      <td>G</td>\n",
       "      <td>4</td>\n",
       "      <td>.A.A</td>\n",
       "      <td>&lt;&lt;&lt;;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312001</td>\n",
       "      <td>T</td>\n",
       "      <td>3</td>\n",
       "      <td>..C</td>\n",
       "      <td>&lt;;&lt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312002</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>&lt;&lt;&lt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr  snp_pos ref  reads         pileup              5\n",
       "0  Chr1  1100021   T     12   ......,.,.,.   <<<<<;<<<;<<\n",
       "1  Chr1  1100022   G     13  .,.,..TT.,...  =&;<<<;<<<<<<\n",
       "2  Chr1  1100023   G     13  ...A...A..A,.  ==<;;;<<<<<<<\n",
       "0  Chr2   312000   G      4           .A.A           <<<;\n",
       "1  Chr2   312001   T      3            ..C            <;<\n",
       "2  Chr2   312002   A      3            ...            <<<"
      ]
     },
     "execution_count": 482,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# need to merge the mpileups\n",
    "mp3 = pd.concat([mp_1b, mp2])\n",
    "\n",
    "# this will allow us to create one dictionary\n",
    "mp3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 483,
   "id": "6857bcaa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mrna_id</th>\n",
       "      <th>prot_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>effect</th>\n",
       "      <th>snp_cds_pos</th>\n",
       "      <th>codon1_genome_pos</th>\n",
       "      <th>codon2_genome_pos</th>\n",
       "      <th>codon3_genome_pos</th>\n",
       "      <th>snp_aa_pos</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>alt_codon</th>\n",
       "      <th>ref_aa</th>\n",
       "      <th>alt_aa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gene1</td>\n",
       "      <td>mrna1</td>\n",
       "      <td>prot1</td>\n",
       "      <td>+</td>\n",
       "      <td>stop_gain</td>\n",
       "      <td>21</td>\n",
       "      <td>1100021</td>\n",
       "      <td>1100022</td>\n",
       "      <td>1100023</td>\n",
       "      <td>7</td>\n",
       "      <td>TGG</td>\n",
       "      <td>TGA</td>\n",
       "      <td>W</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312000</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gene2</td>\n",
       "      <td>mrna2</td>\n",
       "      <td>prot2</td>\n",
       "      <td>+</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>4</td>\n",
       "      <td>312000</td>\n",
       "      <td>312001</td>\n",
       "      <td>312002</td>\n",
       "      <td>2</td>\n",
       "      <td>GTA</td>\n",
       "      <td>ATA</td>\n",
       "      <td>V</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312001</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>gene2</td>\n",
       "      <td>mrna2</td>\n",
       "      <td>prot2</td>\n",
       "      <td>+</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>5</td>\n",
       "      <td>312000</td>\n",
       "      <td>312001</td>\n",
       "      <td>312002</td>\n",
       "      <td>2</td>\n",
       "      <td>GTA</td>\n",
       "      <td>GCA</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele gene_id mrna_id prot_id strand  \\\n",
       "0   Chr1  1100023          G          A   gene1   mrna1   prot1      +   \n",
       "1   Chr2   312000          G          A   gene2   mrna2   prot2      +   \n",
       "2   Chr2   312001          T          C   gene2   mrna2   prot2      +   \n",
       "\n",
       "          effect  snp_cds_pos  codon1_genome_pos  codon2_genome_pos  \\\n",
       "0      stop_gain           21            1100021            1100022   \n",
       "1  nonsynonymous            4             312000             312001   \n",
       "2  nonsynonymous            5             312000             312001   \n",
       "\n",
       "   codon3_genome_pos  snp_aa_pos ref_codon alt_codon ref_aa alt_aa  \n",
       "0            1100023           7       TGG       TGA      W      *  \n",
       "1             312002           2       GTA       ATA      V      I  \n",
       "2             312002           2       GTA       GCA      V      A  "
      ]
     },
     "execution_count": 483,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# toysnpeff2 has the snptable info for everything\n",
    "toy_snpeff2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 484,
   "id": "2e3c6a1f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mrna_id</th>\n",
       "      <th>prot_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>effect</th>\n",
       "      <th>snp_cds_pos</th>\n",
       "      <th>...</th>\n",
       "      <th>pos1_pileup</th>\n",
       "      <th>pos2_pileup</th>\n",
       "      <th>pos3_pileup</th>\n",
       "      <th>varify_codon</th>\n",
       "      <th>varify_allele</th>\n",
       "      <th>varify_aa</th>\n",
       "      <th>nt_VARified</th>\n",
       "      <th>codon_VARified</th>\n",
       "      <th>aa_VARified</th>\n",
       "      <th>comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gene1</td>\n",
       "      <td>mrna1</td>\n",
       "      <td>prot1</td>\n",
       "      <td>+</td>\n",
       "      <td>stop_gain</td>\n",
       "      <td>21</td>\n",
       "      <td>...</td>\n",
       "      <td>......,.,.,.</td>\n",
       "      <td>.,.,..TT.,...</td>\n",
       "      <td>...A...A..A,.</td>\n",
       "      <td>NA</td>\n",
       "      <td>A</td>\n",
       "      <td></td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>SNP(s) {1100022: 'T'} only in RNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312000</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gene2</td>\n",
       "      <td>mrna2</td>\n",
       "      <td>prot2</td>\n",
       "      <td>+</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>.A.A</td>\n",
       "      <td>..C</td>\n",
       "      <td>...</td>\n",
       "      <td>complex_codon</td>\n",
       "      <td>A</td>\n",
       "      <td></td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312001</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>gene2</td>\n",
       "      <td>mrna2</td>\n",
       "      <td>prot2</td>\n",
       "      <td>+</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>.A.A</td>\n",
       "      <td>..C</td>\n",
       "      <td>...</td>\n",
       "      <td>complex_codon</td>\n",
       "      <td>C</td>\n",
       "      <td></td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele gene_id mrna_id prot_id strand  \\\n",
       "0   Chr1  1100023          G          A   gene1   mrna1   prot1      +   \n",
       "1   Chr2   312000          G          A   gene2   mrna2   prot2      +   \n",
       "2   Chr2   312001          T          C   gene2   mrna2   prot2      +   \n",
       "\n",
       "          effect  snp_cds_pos  ...   pos1_pileup    pos2_pileup  \\\n",
       "0      stop_gain           21  ...  ......,.,.,.  .,.,..TT.,...   \n",
       "1  nonsynonymous            4  ...          .A.A            ..C   \n",
       "2  nonsynonymous            5  ...          .A.A            ..C   \n",
       "\n",
       "     pos3_pileup   varify_codon varify_allele varify_aa nt_VARified  \\\n",
       "0  ...A...A..A,.             NA             A                   Yes   \n",
       "1            ...  complex_codon             A                   Yes   \n",
       "2            ...  complex_codon             C                   Yes   \n",
       "\n",
       "  codon_VARified aa_VARified                            comment  \n",
       "0             No          No  SNP(s) {1100022: 'T'} only in RNA  \n",
       "1             No          No                                     \n",
       "2             No          No                                     \n",
       "\n",
       "[3 rows x 28 columns]"
      ]
     },
     "execution_count": 484,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merging the two tables\n",
    "mp = mp3\n",
    "x = mergeTab(mp3, toy_snpeff2)\n",
    "pos, alt_dict = checkRNADict(x, mp3)\n",
    "\n",
    "varify(x, mp)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9131e49b",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Real World Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "add3a452",
   "metadata": {},
   "source": [
    "## Reading in the Mpileup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ceaf96aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "enc_mp = readMP('../data/test/ENCFF283TLK_chr1.mpileup')\n",
    "enc_mp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d14b4608",
   "metadata": {},
   "outputs": [],
   "source": [
    "sub_m3 = enc_mp[enc_mp.snp_pos.isin(range(69265, 4712659))]\n",
    "sub_m3.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "042a5b15",
   "metadata": {},
   "source": [
    "## Reading in the SnpEff-like Table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abd2d08f",
   "metadata": {},
   "outputs": [],
   "source": [
    "snptable3 = pd.read_table('../data/test/ENCFF541HLI.snp_effects.chr1_only_10kgenic_snps.no_intron.no_utr.tsv')\n",
    "snptable3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17170aa0",
   "metadata": {},
   "outputs": [],
   "source": [
    "mp = sub_m3\n",
    "x = mergeTab(sub_m3, snptable3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e63b2f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "pos, alt_dict = checkRNADict(x, mp)\n",
    "df = varify(x[0:100], mp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12ad59aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.sample(n=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9668c45",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('../data/output/varify_output_RWDsubset.csv', header=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
