{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1106b087",
   "metadata": {},
   "source": [
    "# VARify"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c87d4722",
   "metadata": {},
   "source": [
    "\n",
    "## Workflow\n",
    "\n",
    "\n",
    "<img src=\"wf.png\" width=\"600\">"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31c60665",
   "metadata": {},
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 458,
   "id": "3b057713",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Utility File\n",
    "\n",
    "# libraries\n",
    "import warnings\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "\n",
    "def aaVarify(x):\n",
    "    x['aa_VARified'] = 'Yes' if x['varify_aa'] == x['alt_aa'] else \"No\"\n",
    "    return x\n",
    "\n",
    "def codonVarify(x):\n",
    "    x['codon_VARified'] = 'Yes' if x['varify_codon'] == x['alt_codon'] else 'No'\n",
    "    return x\n",
    "\n",
    "def translate(x):\n",
    "    table = {\n",
    "    'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "    'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "    'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "    'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',                \n",
    "    'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "    'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "    'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "    'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "    'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "    'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "    'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "    'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "    'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "    'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "    'TAC':'Y', 'TAT':'Y', 'TAA':'*', 'TAG':'*',\n",
    "    'TGC':'C', 'TGT':'C', 'TGA':'*', 'TGG':'W',}\n",
    "    \n",
    "    codon = x['varify_codon']\n",
    "\n",
    "    if codon not in table.keys():\n",
    "        x['varify_aa'] = \"\"\n",
    "    else:\n",
    "        x['varify_aa'] =table[codon]\n",
    "        \n",
    "    return(x)\n",
    "\n",
    "\n",
    "# For every row in here, we're going to apply this uniqueSNP function\n",
    "\n",
    "def mergeTab(mp, so):\n",
    "    sub_mp = mp[['snp_pos', 'pileup']]\n",
    "    # sub_mp['snp_pos'] = str(sub_mp['snp_pos'])\n",
    "    mdf = pd.merge(so, sub_mp, on='snp_pos')\n",
    "    return mdf\n",
    "\n",
    "\n",
    "def varOnly(a):\n",
    "    '''\n",
    "    purpose: remove non-nt values in pileup\n",
    "    input: string of mpileup \n",
    "    output: nt string\n",
    "    '''\n",
    "    x = [i for i in a if i.upper() in ['A','C','T','G']]\n",
    "    \n",
    "    return x\n",
    "\n",
    "def isMono(ref, temp):\n",
    "    # Check if two snps are monomorphic\n",
    "    c=0\n",
    "    count = [c+1 for snp in temp if snp == ref]\n",
    "    if sum(count) >= 2: \n",
    "        warnings.warn(\"More than 2 monomorphic alleles at position.\")\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "    \n",
    "\n",
    "def altValidate(x):\n",
    "    # Varify alt == Alt? T or F\n",
    "    # Add flag to capture\n",
    "    \n",
    "    x['nt_VARified'] = 'Yes' if x['alt_allele'] == x['varify_allele'] else 'No'\n",
    "    \n",
    "    return x\n",
    "\n",
    "def varify(x, mp):\n",
    "    # dictionary of alternate codons from pileup\n",
    "    alt_dict = altDict(mp)\n",
    "\n",
    "    # running the script\n",
    "    snp_df = x.apply(axis=1, func=snpDict)\n",
    "    snp_df = pd.concat(list(snp_df))\n",
    "    df = pd.merge(x, snp_df, on='snp_pos')\n",
    "\n",
    "    # Add the dictionary\n",
    "    df = df.apply(axis=1, func=translate)\n",
    "    df = df.apply(axis=1, func=altValidate)\n",
    "    df = df.apply(axis=1, func=codonVarify)\n",
    "    df = df.apply(axis=1, func=aaVarify)\n",
    "    df = df.drop(columns='pileup', axis=1)\n",
    "    \n",
    "    return df\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def effect(x):\n",
    "    # missense\n",
    "    # nonsense\n",
    "    x['effect'] = '_'\n",
    "    \n",
    "\n",
    "###################Older, probably outdated############################\n",
    "\n",
    "# Codon Combos\n",
    "import numpy as np\n",
    "import math \n",
    "\n",
    "def getAlt(x):\n",
    "    l = list(x['pileup'])\n",
    "    # Check if alphabet\n",
    "    l = [i for i in l if i.isalpha()]\n",
    "    if len(l) == 0:\n",
    "        x['varify_allele'] = None\n",
    "    else:     \n",
    "        # make upper\n",
    "        l = [i for i in l if i.upper() in ['A','C','T','G']]\n",
    "\n",
    "        # Get list of unique\n",
    "        a = [i for i in l]\n",
    "        a_unique = list(set(a)) \n",
    "\n",
    "        # getting max alt\n",
    "        a_unique = pd.DataFrame(a_unique)\n",
    "        \n",
    "        # Get the max snp\n",
    "        nt = a_unique.groupby([0]).apply(lambda x: x.value_counts().index[0])[0]\n",
    "        nt = nt[0]\n",
    "        x['varify_allele'] = nt\n",
    "    \n",
    "    return x\n",
    "\n",
    "def altDict(x):\n",
    "    \n",
    "    '''\n",
    "    Takes mPileup\n",
    "    Out: mpileup dictionary with snp_pos and varify_allele\n",
    "    '''\n",
    "    \n",
    "    # dataframe of mpileup goes through each pileup with apply\n",
    "    x = x.apply(axis=1, func=getAlt)\n",
    "    alt_dict = dict(zip((x['snp_pos']), x['varify_allele']))\n",
    "    clean_dict = {k: alt_dict[k] for k in alt_dict if pd.isna(alt_dict[k]) == False}\n",
    "    \n",
    "    \n",
    "    return clean_dict\n",
    "    \n",
    "\n",
    "def flatten(d, parent_key='', sep='_'):\n",
    "    import collections\n",
    "    items = []\n",
    "    for k, v in d.items():\n",
    "        new_key = parent_key + sep + k if parent_key else k\n",
    "        if isinstance(v, collections.MutableMapping):\n",
    "            items.extend(flatten(v, new_key, sep=sep).items())\n",
    "        else:\n",
    "            items.append((new_key, v))\n",
    "    return dict(items)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f708429",
   "metadata": {},
   "source": [
    "# Running Script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "37bbd614",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  chr_id  snp_pos ref_allele alt_allele varify_alt  codon1_genome_pos  \\\n",
      "0   chr1      521          G          C          A                519   \n",
      "0   chr1      523          A          T          T                521   \n",
      "0   chr1      524          C          G          G                522   \n",
      "\n",
      "   codon2_genome_pos  codon3_genome_pos ref_codon varify_codon1 varify_codon2  \\\n",
      "0                520                521       AAG           AAG           AAG   \n",
      "0                522                523       GTA           TTA           GTA   \n",
      "0                523                524       TAC           GAC           TGC   \n",
      "\n",
      "  varify_codon3 flag  \n",
      "0           AAA       \n",
      "0           GTT       \n",
      "0           TAG       \n"
     ]
    }
   ],
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "%run '../script/main.py' --in_mp '../data/test/mpiletest-out.txt' --in_so '../data/test/snp_out.txt'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "302e41ef",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Case 1\n",
    "\n",
    "Consider the following mutation that introduces a G->A mutation.\n",
    " \n",
    "VCF:\n",
    "Chr1\t1100023      \t.       \tG     \tA      \t60    \tPASS\t.       \tGT    \t0/1\n",
    " \n",
    "* Genotype: G/A\n",
    "* Reference base: G\n",
    "* Alternate base: A\n",
    "* Predicted reference codon: TGG (tryptophan)\n",
    "* Predicted alternate codon: TGA (premature termination codon)\n",
    " \n",
    "A tool like snpEff can be used to predict the effect of the mutation using a reference genome annotation (e.g., GFF). In this case the mutation occurs in a coding exon at third position of a codon that codes for tryptophan in the reference proteome and a premature termination codon in the case of the alternate allele. snpEff will determine the reference and alternate codon and return the two codons and their translation.\n",
    " \n",
    "The predicted effect is therefore a nonsense mutation.\n",
    "\n",
    "## Goal Example\n",
    "| chr_id|snp_pos     | Ref     | Pred_Alt |Varify_Alt|VARified|ref_codon|varify_codon|aa|effect|\n",
    "| -----------| ----------- | ----------- | ----------- | ----------- | ----------- |----------- |----------- |----------- |----------- |\n",
    "|Chr1|1100023| G      | A       |A|Yes| TGG | TGA |*|stop_gain|"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e54cc130",
   "metadata": {},
   "source": [
    "## Read in the Toy Mpileup File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 459,
   "id": "76adbb9d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>3</th>\n",
       "      <th>pileup</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100021</td>\n",
       "      <td>T</td>\n",
       "      <td>12</td>\n",
       "      <td>.............</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100022</td>\n",
       "      <td>G</td>\n",
       "      <td>13</td>\n",
       "      <td>.............</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>G</td>\n",
       "      <td>13</td>\n",
       "      <td>...A...A..A..</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0  snp_pos ref   3         pileup\n",
       "0  Chr1  1100021   T  12  .............\n",
       "1  Chr1  1100022   G  13  .............\n",
       "2  Chr1  1100023   G  13  ...A...A..A.."
      ]
     },
     "execution_count": 459,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def readMP(x):\n",
    "    import pandas as pd\n",
    "    mp = pd.read_table(x, header=None)\n",
    "    mp = mp.rename(columns={1:\"snp_pos\", 2:\"ref\", 4:\"pileup\"})\n",
    "    \n",
    "    return(mp)\n",
    "    \n",
    "# mpileup \n",
    "mp1 = readMP('../data/test/case1_pileup.txt')\n",
    "\n",
    "mp1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ddaf25c",
   "metadata": {},
   "source": [
    "## Read in VCF File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 460,
   "id": "39f98440",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>NA00001</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>60</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312000</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>60</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312001</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>55</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  #CHROM      POS ID REF ALT  QUAL FILTER INFO FORMAT NA00001\n",
       "0   Chr1  1100023  .   G   A    60   PASS    .     GT     0/1\n",
       "1   Chr2   312000  .   G   A    60   PASS    .     GT     0/1\n",
       "2   Chr2   312001  .   T   C    55   PASS    .     GT     0/1"
      ]
     },
     "execution_count": 460,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The toy VCF\n",
    "import pandas as pd\n",
    "toy_vcf = pd.read_table('../data/toy/toy.vcf')\n",
    "\n",
    "toy_vcf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66da6f0e",
   "metadata": {},
   "source": [
    "## Toy SNPeff Table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 461,
   "id": "6a05677f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mrna_id</th>\n",
       "      <th>prot_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>effect</th>\n",
       "      <th>snp_cds_pos</th>\n",
       "      <th>codon1_genome_pos</th>\n",
       "      <th>codon2_genome_pos</th>\n",
       "      <th>codon3_genome_pos</th>\n",
       "      <th>snp_aa_pos</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>alt_codon</th>\n",
       "      <th>ref_aa</th>\n",
       "      <th>alt_aa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gene1</td>\n",
       "      <td>mrna1</td>\n",
       "      <td>prot1</td>\n",
       "      <td>+</td>\n",
       "      <td>stop_gain</td>\n",
       "      <td>21</td>\n",
       "      <td>1100021</td>\n",
       "      <td>1100022</td>\n",
       "      <td>1100023</td>\n",
       "      <td>7</td>\n",
       "      <td>TGG</td>\n",
       "      <td>TGA</td>\n",
       "      <td>W</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312000</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gene2</td>\n",
       "      <td>mrna2</td>\n",
       "      <td>prot2</td>\n",
       "      <td>+</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>4</td>\n",
       "      <td>312000</td>\n",
       "      <td>312001</td>\n",
       "      <td>312002</td>\n",
       "      <td>2</td>\n",
       "      <td>GTA</td>\n",
       "      <td>ATA</td>\n",
       "      <td>V</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312001</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>gene2</td>\n",
       "      <td>mrna2</td>\n",
       "      <td>prot2</td>\n",
       "      <td>+</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>5</td>\n",
       "      <td>312000</td>\n",
       "      <td>312001</td>\n",
       "      <td>312002</td>\n",
       "      <td>2</td>\n",
       "      <td>GTA</td>\n",
       "      <td>GCA</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele gene_id mrna_id prot_id strand  \\\n",
       "0   Chr1  1100023          G          A   gene1   mrna1   prot1      +   \n",
       "1   Chr2   312000          G          A   gene2   mrna2   prot2      +   \n",
       "2   Chr2   312001          T          C   gene2   mrna2   prot2      +   \n",
       "\n",
       "          effect  snp_cds_pos  codon1_genome_pos  codon2_genome_pos  \\\n",
       "0      stop_gain           21            1100021            1100022   \n",
       "1  nonsynonymous            4             312000             312001   \n",
       "2  nonsynonymous            5             312000             312001   \n",
       "\n",
       "   codon3_genome_pos  snp_aa_pos ref_codon alt_codon ref_aa alt_aa  \n",
       "0            1100023           7       TGG       TGA      W      *  \n",
       "1             312002           2       GTA       ATA      V      I  \n",
       "2             312002           2       GTA       GCA      V      A  "
      ]
     },
     "execution_count": 461,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Toy Table\n",
    "toy_snpeff = pd.read_table('../data/toy/toy.snp_effects.tsv')\n",
    "toy_snpeff"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db940360",
   "metadata": {},
   "source": [
    "## Merge the mpileup & snpeff table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 464,
   "id": "2e778eab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# if pileup information is . at the previous positions, output the reference, else variant\n",
    "def mergeTab(mp, so):\n",
    "    sub_mp = mp[['snp_pos', 'ref', 'pileup']]\n",
    "    mdf = pd.merge(so, sub_mp, on='snp_pos')\n",
    "    return mdf\n",
    "\n",
    "# Merging the two tables\n",
    "x = mergeTab(mp1, toy_snpeff)\n",
    "\n",
    "# dictionary of alternate codons from pileup\n",
    "mp_dict = altDict(mp1) # needs to be based on mp1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 466,
   "id": "d15861cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mrna_id</th>\n",
       "      <th>prot_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>effect</th>\n",
       "      <th>snp_cds_pos</th>\n",
       "      <th>codon1_genome_pos</th>\n",
       "      <th>codon2_genome_pos</th>\n",
       "      <th>codon3_genome_pos</th>\n",
       "      <th>snp_aa_pos</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>alt_codon</th>\n",
       "      <th>ref_aa</th>\n",
       "      <th>alt_aa</th>\n",
       "      <th>ref</th>\n",
       "      <th>pileup</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gene1</td>\n",
       "      <td>mrna1</td>\n",
       "      <td>prot1</td>\n",
       "      <td>+</td>\n",
       "      <td>stop_gain</td>\n",
       "      <td>21</td>\n",
       "      <td>1100021</td>\n",
       "      <td>1100022</td>\n",
       "      <td>1100023</td>\n",
       "      <td>7</td>\n",
       "      <td>TGG</td>\n",
       "      <td>TGA</td>\n",
       "      <td>W</td>\n",
       "      <td>*</td>\n",
       "      <td>G</td>\n",
       "      <td>...A...A..A..</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele gene_id mrna_id prot_id strand  \\\n",
       "0   Chr1  1100023          G          A   gene1   mrna1   prot1      +   \n",
       "\n",
       "      effect  snp_cds_pos  codon1_genome_pos  codon2_genome_pos  \\\n",
       "0  stop_gain           21            1100021            1100022   \n",
       "\n",
       "   codon3_genome_pos  snp_aa_pos ref_codon alt_codon ref_aa alt_aa ref  \\\n",
       "0            1100023           7       TGG       TGA      W      *   G   \n",
       "\n",
       "          pileup  \n",
       "0  ...A...A..A..  "
      ]
     },
     "execution_count": 466,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merged tables\n",
    "x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 467,
   "id": "c0856215",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1100023: 'A'}"
      ]
     },
     "execution_count": 467,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# dictionary of alts from merged tables\n",
    "mp_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 468,
   "id": "38f731df",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Using this information, \n",
    "we're going to create our varify codon based on any\n",
    "existing alts on all positions. \n",
    "\n",
    "If the position does not exist, we take from the referenc\n",
    "'''\n",
    "alt_dict = mp_dict\n",
    "\n",
    "# a_unique = getAlt(x) # Replaced with alt_dict[3]\n",
    "\n",
    "# Takes the current position of SNP\n",
    "p3 = int(x['snp_pos'])\n",
    "p2 = p3-1 \n",
    "p1 = p2-1\n",
    "\n",
    "subDict = {p1 : '.', p2 : '.', p3 : '.'}\n",
    "\n",
    "# create a dictionary\n",
    "ref = str(x['ref_codon'][0]) # starts from reference\n",
    "\n",
    "# set up to create varify_alt\n",
    "df = pd.DataFrame()\n",
    "temp = [i for i in ref]\n",
    "snp_dict = {}\n",
    "\n",
    "# Add values from full dictionary if keys are in sub\n",
    "res = {key: alt_dict.get(key, subDict[key]) for key in subDict}\n",
    "res = ''.join(res.values())\n",
    "\n",
    "for n in range(3):\n",
    "    snp_dict['snp_pos'] = int(x['snp_pos'][0])\n",
    "    snp_dict['varify_allele'] = alt_dict[p3] # alt of 'current' position\n",
    "    if res[n] != '.':\n",
    "        temp[n] = res[n]\n",
    "        \n",
    "    snp_dict[\"varify_codon\"] = ''.join(temp)\n",
    "    \n",
    "# snp_dict output: {'snp_pos': 1100023, 'varify_alt': 'A', 'varify_codon': 'TGA'}\n",
    "\n",
    "# Add the dictionary \n",
    "df = df.append(snp_dict, ignore_index=True)\n",
    "df = pd.merge(x, df, on='snp_pos')\n",
    "df = df.apply(axis=1, func=altValidate)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 470,
   "id": "9b5146b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mrna_id</th>\n",
       "      <th>prot_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>effect</th>\n",
       "      <th>snp_cds_pos</th>\n",
       "      <th>...</th>\n",
       "      <th>snp_aa_pos</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>alt_codon</th>\n",
       "      <th>ref_aa</th>\n",
       "      <th>alt_aa</th>\n",
       "      <th>ref</th>\n",
       "      <th>pileup</th>\n",
       "      <th>varify_allele</th>\n",
       "      <th>varify_codon</th>\n",
       "      <th>nt_VARified</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gene1</td>\n",
       "      <td>mrna1</td>\n",
       "      <td>prot1</td>\n",
       "      <td>+</td>\n",
       "      <td>stop_gain</td>\n",
       "      <td>21</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>TGG</td>\n",
       "      <td>TGA</td>\n",
       "      <td>W</td>\n",
       "      <td>*</td>\n",
       "      <td>G</td>\n",
       "      <td>...A...A..A..</td>\n",
       "      <td>A</td>\n",
       "      <td>TGA</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele gene_id mrna_id prot_id strand  \\\n",
       "0   Chr1  1100023          G          A   gene1   mrna1   prot1      +   \n",
       "\n",
       "      effect  snp_cds_pos  ...  snp_aa_pos  ref_codon  alt_codon  ref_aa  \\\n",
       "0  stop_gain           21  ...           7        TGG        TGA       W   \n",
       "\n",
       "  alt_aa ref         pileup varify_allele varify_codon nt_VARified  \n",
       "0      *   G  ...A...A..A..             A          TGA         Yes  \n",
       "\n",
       "[1 rows x 23 columns]"
      ]
     },
     "execution_count": 470,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 472,
   "id": "5741b160",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>varify_allele</th>\n",
       "      <th>nt_VARified</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>varify_codon</th>\n",
       "      <th>varify_aa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>Yes</td>\n",
       "      <td>TGG</td>\n",
       "      <td>TGA</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele varify_allele nt_VARified ref_codon  \\\n",
       "0   Chr1  1100023          G          A             A         Yes       TGG   \n",
       "\n",
       "  varify_codon varify_aa  \n",
       "0          TGA         *  "
      ]
     },
     "execution_count": 472,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.apply(axis=1, func=translate)\n",
    "#df = effect(df)\n",
    "\n",
    "reorg = ['chr_id', 'snp_pos', 'ref_allele', 'alt_allele', \n",
    "         'varify_allele', 'nt_VARified', 'ref_codon', 'varify_codon', 'varify_aa', #'effect'\n",
    "        ]\n",
    "\n",
    "df = df[reorg]\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85851f3a",
   "metadata": {},
   "source": [
    "# testing the case2 function here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 473,
   "id": "b003091f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>effect</th>\n",
       "      <th>codon1_genome_pos</th>\n",
       "      <th>codon2_genome_pos</th>\n",
       "      <th>codon3_genome_pos</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>alt_codon</th>\n",
       "      <th>ref_aa</th>\n",
       "      <th>alt_aa</th>\n",
       "      <th>varify_codon</th>\n",
       "      <th>varify_allele</th>\n",
       "      <th>varify_aa</th>\n",
       "      <th>nt_VARified</th>\n",
       "      <th>codon_VARified</th>\n",
       "      <th>aa_VARified</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>stop_gain</td>\n",
       "      <td>1100021</td>\n",
       "      <td>1100022</td>\n",
       "      <td>1100023</td>\n",
       "      <td>TGG</td>\n",
       "      <td>TGA</td>\n",
       "      <td>W</td>\n",
       "      <td>*</td>\n",
       "      <td>TGA</td>\n",
       "      <td>A</td>\n",
       "      <td>*</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele     effect  codon1_genome_pos  \\\n",
       "0   Chr1  1100023          G          A  stop_gain            1100021   \n",
       "\n",
       "   codon2_genome_pos  codon3_genome_pos ref_codon alt_codon ref_aa alt_aa  \\\n",
       "0            1100022            1100023       TGG       TGA      W      *   \n",
       "\n",
       "  varify_codon varify_allele varify_aa nt_VARified codon_VARified aa_VARified  \n",
       "0          TGA             A         *         Yes            Yes         Yes  "
      ]
     },
     "execution_count": 473,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# let's just do 1 line\n",
    "x = x.iloc[[0]]\n",
    "\n",
    "# dictionary of alternate codons from pileup\n",
    "alt_dict = altDict(mp1)\n",
    "\n",
    "# running the script\n",
    "snp_df = x.apply(axis=1, func=snpDict)\n",
    "snp_df = pd.concat(list(snp_df))\n",
    "df = pd.merge(x, snp_df, on='snp_pos')\n",
    "\n",
    "# Add the dictionary\n",
    "df = df.apply(axis=1, func=translate)\n",
    "df = df.apply(axis=1, func=altValidate)\n",
    "df = df.apply(axis=1, func=codonVarify)\n",
    "df = df.apply(axis=1, func=aaVarify)\n",
    "\n",
    "reorg = ['chr_id', 'snp_pos', 'ref_allele', 'alt_allele', \n",
    "         #'gene_id', 'mrna_id','prot_id', 'strand', \n",
    "         'effect', #'snp_cds_pos', \n",
    "         'codon1_genome_pos','codon2_genome_pos', 'codon3_genome_pos', \n",
    "         #'snp_aa_pos', \n",
    "         'ref_codon','alt_codon', 'ref_aa', 'alt_aa', 'varify_codon', 'varify_allele',\n",
    "         'varify_aa', 'nt_VARified', 'codon_VARified', 'aa_VARified']\n",
    "df = df.drop(columns='pileup', axis=1)\n",
    "df[reorg]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b137877e",
   "metadata": {},
   "source": [
    "# Case 1.c \n",
    "\n",
    "To determine if the codon is VARified, VARify will look at the two other positions in the same codon (the ‘non-focal positions’) and determine if ONLY the reference nucleotide is observed in RNA. \n",
    "\n",
    "For example, if both non-focal positions are monomorphic for the reference nucleotide in the RNA mpileup (i.e., the reference nucleotide from VCF in snpeff table), then the codon is verified and the “codon_VARified” column will be set to ‘yes’. The Case 1 example above would have the codon_VARified column set to ‘yes’.\n",
    "\n",
    "All other situations will have the “codon_VARified” column set to “no”. These cases could include cases in which one (or both) non-focal positions have two or more nucleotides in the RNA mpileup (note: this is NOT shown in Case 1 example above). As we discussed, this could occur frequently because of sequencing errors. We will case this “Case 1b”. \n",
    "\n",
    "Another situation might be if one (or both) non-focal positions are monomorphic for a non-reference nucleotide. This also would be inconsistent with the codon reported in alt_codon column and the “codon_VARified” column should be reported as “no”. We will call this “Case 1c”\n",
    "\n",
    "Case 1b:\n",
    "\n",
    "Given: VCF same as Case 1.\n",
    "RNA Mpileup:\n",
    "|chromosome_id|snp_pos|ref_allele|pos|pileup||\n",
    "|-----|-----|-----|-----|-----|-----|\n",
    "|Chr1|1100021|T|12|.......|other columns|\n",
    "|Chr1|1100022|G|13|..TT..|other columns|\n",
    "|Chr1|1100023|G|13|..A..A..A..|other columns|\n",
    "\n",
    "\n",
    "Case 1c:\n",
    "\n",
    "Given: VCF same as Case 1\n",
    "RNA Mpileup:\n",
    "|chromosome_id|snp_pos|ref_allele|pos|pileup||\n",
    "|-----|-----|-----|-----|-----|-----|\n",
    "|Chr1|1100021|T|12|CCCCCC|other columns|\n",
    "|Chr1|1100022|G|13|...|other columns|\n",
    "|Chr1|1100023|G|13|..A..A..A..|other columns|\n",
    "\n",
    "*note: in Case 1c, the mpileup does not have the reference nucleotide at non-focal position 1100021. Therefore the alt_codon (TGG) is not verified (even though both non-focal positions are monomorphic in RNA mpileup.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 335,
   "id": "78551977",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>3</th>\n",
       "      <th>pileup</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100021</td>\n",
       "      <td>T</td>\n",
       "      <td>12</td>\n",
       "      <td>CCCCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100022</td>\n",
       "      <td>G</td>\n",
       "      <td>13</td>\n",
       "      <td>.............</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>G</td>\n",
       "      <td>13</td>\n",
       "      <td>...A...A..A..</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0  snp_pos ref   3         pileup\n",
       "0  Chr1  1100021   T  12         CCCCCC\n",
       "1  Chr1  1100022   G  13  .............\n",
       "2  Chr1  1100023   G  13  ...A...A..A.."
      ]
     },
     "execution_count": 335,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mp_1c = readMP('../data/test/case1c_pileup.txt')\n",
    "\n",
    "mp_1c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 474,
   "id": "e861a16a",
   "metadata": {},
   "outputs": [],
   "source": [
    "x_1c = mergeTab(mp_1c, toy_snpeff)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 477,
   "id": "d99c8e8d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1100021: 'C', 1100023: 'A'}"
      ]
     },
     "execution_count": 477,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alt_dict = altDict(mp_1c)\n",
    "alt_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38f477f8",
   "metadata": {},
   "source": [
    "There are two issues\n",
    "1. When there are two snps in one codon found in both DNA & RNA\n",
    "    * Solution: output complex codon\n",
    "2. When there are two snps found in RNA only \n",
    "    * Solution: Output comment - POS2 in RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2950f7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def snpDict(x, alt_dict=alt_dict):\n",
    "    import itertools\n",
    "    # Takes the current position of SNP\n",
    "    p3 = int(x.codon3_genome_pos)\n",
    "    p2 = int(x.codon2_genome_pos)\n",
    "    p1 = int(x.codon1_genome_pos)\n",
    "    \n",
    "    # Empty Dict to fill if pos exists\n",
    "    subDict = {p1 : '.', p2 : '.', p3 : '.'}\n",
    "    res = {key: alt_dict.get(key, subDict[key]) for key in subDict}\n",
    "    res = ''.join(res.values())\n",
    "    \n",
    "    # set up to create varify_alt\n",
    "    ref = list(x['ref_codon']) # starts from reference\n",
    "    df = pd.DataFrame()\n",
    "    \n",
    "    # create a new dictionary\n",
    "    snp_dict = {}\n",
    "    multi_check = False\n",
    "\n",
    "    # MultiFlag for complex codons\n",
    "    if (list(res)).count(\".\") < 2:\n",
    "        # if < 2, then there are multi snps \n",
    "        flag = 'complex_codon'\n",
    "        multi_check = True\n",
    "    else: \n",
    "        flag = ''\n",
    "\n",
    "    if multi_check == True: \n",
    "        # if multi_check true, then complex_codon\n",
    "        snp_dict['varify_codon'] = flag\n",
    "        snp_dict['snp_pos'] = x['snp_pos']\n",
    "        current_pos = int(x['snp_pos'])\n",
    "        snp_dict['varify_allele'] = alt_dict[int(current_pos)] # fill with current pos snp\n",
    "                                      \n",
    "    else:\n",
    "        current_pos = int(x['snp_pos'])\n",
    "        idx = list(subDict.keys()).index(current_pos)\n",
    "        # get ref\n",
    "        ref[idx] = alt_dict[current_pos]\n",
    "        snp_dict['varify_codon'] = ''.join(ref)\n",
    "        snp_dict['snp_pos'] = x['snp_pos']\n",
    "        snp_dict['varify_allele'] = alt_dict[int(current_pos)] # fill with current pos snp\n",
    "\n",
    "        \n",
    "    snp_df = pd.DataFrame([snp_dict])\n",
    "\n",
    "\n",
    "    return snp_df\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfd7e130",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Case 2: Multi-substitution case\n",
    "\n",
    "In the case of multi substitution, we have alts on pos1 and pos2 of the codon. \n",
    "We want to capture this information, although pos3 has no alt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a87d112e",
   "metadata": {},
   "source": [
    "## Case2 Mpileup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 309,
   "id": "26b47f42",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>3</th>\n",
       "      <th>pileup</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312000</td>\n",
       "      <td>G</td>\n",
       "      <td>12</td>\n",
       "      <td>.A.A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312001</td>\n",
       "      <td>T</td>\n",
       "      <td>13</td>\n",
       "      <td>..C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312002</td>\n",
       "      <td>A</td>\n",
       "      <td>13</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0  snp_pos ref   3 pileup\n",
       "0  Chr2   312000   G  12   .A.A\n",
       "1  Chr2   312001   T  13    ..C\n",
       "2  Chr2   312002   A  13    ..."
      ]
     },
     "execution_count": 309,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mp2 = readMP('../data/toy/toy_mpileup.txt')\n",
    "mp2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be421564",
   "metadata": {},
   "source": [
    "## Case2 Snpeff-like table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 310,
   "id": "174b97ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mrna_id</th>\n",
       "      <th>prot_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>effect</th>\n",
       "      <th>snp_cds_pos</th>\n",
       "      <th>codon1_genome_pos</th>\n",
       "      <th>codon2_genome_pos</th>\n",
       "      <th>codon3_genome_pos</th>\n",
       "      <th>snp_aa_pos</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>alt_codon</th>\n",
       "      <th>ref_aa</th>\n",
       "      <th>alt_aa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gene1</td>\n",
       "      <td>mrna1</td>\n",
       "      <td>prot1</td>\n",
       "      <td>+</td>\n",
       "      <td>stop_gain</td>\n",
       "      <td>21</td>\n",
       "      <td>1100021</td>\n",
       "      <td>1100022</td>\n",
       "      <td>1100023</td>\n",
       "      <td>7</td>\n",
       "      <td>TGG</td>\n",
       "      <td>TGA</td>\n",
       "      <td>W</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312000</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gene2</td>\n",
       "      <td>mrna2</td>\n",
       "      <td>prot2</td>\n",
       "      <td>+</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>4</td>\n",
       "      <td>312000</td>\n",
       "      <td>312001</td>\n",
       "      <td>312002</td>\n",
       "      <td>2</td>\n",
       "      <td>GTA</td>\n",
       "      <td>ATA</td>\n",
       "      <td>V</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312001</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>gene2</td>\n",
       "      <td>mrna2</td>\n",
       "      <td>prot2</td>\n",
       "      <td>+</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>5</td>\n",
       "      <td>312000</td>\n",
       "      <td>312001</td>\n",
       "      <td>312002</td>\n",
       "      <td>2</td>\n",
       "      <td>GTA</td>\n",
       "      <td>GCA</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele gene_id mrna_id prot_id strand  \\\n",
       "0   Chr1  1100023          G          A   gene1   mrna1   prot1      +   \n",
       "1   Chr2   312000          G          A   gene2   mrna2   prot2      +   \n",
       "2   Chr2   312001          T          C   gene2   mrna2   prot2      +   \n",
       "\n",
       "          effect  snp_cds_pos  codon1_genome_pos  codon2_genome_pos  \\\n",
       "0      stop_gain           21            1100021            1100022   \n",
       "1  nonsynonymous            4             312000             312001   \n",
       "2  nonsynonymous            5             312000             312001   \n",
       "\n",
       "   codon3_genome_pos  snp_aa_pos ref_codon alt_codon ref_aa alt_aa  \n",
       "0            1100023           7       TGG       TGA      W      *  \n",
       "1             312002           2       GTA       ATA      V      I  \n",
       "2             312002           2       GTA       GCA      V      A  "
      ]
     },
     "execution_count": 310,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "toy_snpeff2 = pd.read_table('../data/toy/toy.snp_effects.tsv') # toy_snpeff\n",
    "toy_snpeff2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe13a88f",
   "metadata": {},
   "source": [
    "## Goal Example\n",
    "| chr_id|snp_pos     | Ref     | Pred_Alt |Varify_Alt|VARified|ref_codon|varify_codon|pred_codon|varify_aa|effect|alt_codons|flag|\n",
    "| ---| --- | --- | --- | --- | --- |--- |--- |--- |--- |---|--|--|\n",
    "|Chr2|312000| G      | A       |A|Yes| GTA | ATA |T|*|||\n",
    "|Chr2|312001| T      | C       |C|Yes| GTA | GCA |T|*|||"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f0bdfb5",
   "metadata": {},
   "source": [
    "## Merging mpileup and snpeff together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 311,
   "id": "bf819b82",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{312000: 'A', 312001: 'C'}"
      ]
     },
     "execution_count": 311,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merging the two tables\n",
    "x2 = mergeTab(mp2, toy_snpeff2)\n",
    "\n",
    "# dictionary of alternate codons from pileup\n",
    "mp_dict2 = altDict(x2)\n",
    "\n",
    "mp_dict2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 312,
   "id": "3ace3cc1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mrna_id</th>\n",
       "      <th>prot_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>effect</th>\n",
       "      <th>snp_cds_pos</th>\n",
       "      <th>codon1_genome_pos</th>\n",
       "      <th>codon2_genome_pos</th>\n",
       "      <th>codon3_genome_pos</th>\n",
       "      <th>snp_aa_pos</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>alt_codon</th>\n",
       "      <th>ref_aa</th>\n",
       "      <th>alt_aa</th>\n",
       "      <th>pileup</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312000</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gene2</td>\n",
       "      <td>mrna2</td>\n",
       "      <td>prot2</td>\n",
       "      <td>+</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>4</td>\n",
       "      <td>312000</td>\n",
       "      <td>312001</td>\n",
       "      <td>312002</td>\n",
       "      <td>2</td>\n",
       "      <td>GTA</td>\n",
       "      <td>ATA</td>\n",
       "      <td>V</td>\n",
       "      <td>I</td>\n",
       "      <td>.A.A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312001</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>gene2</td>\n",
       "      <td>mrna2</td>\n",
       "      <td>prot2</td>\n",
       "      <td>+</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>5</td>\n",
       "      <td>312000</td>\n",
       "      <td>312001</td>\n",
       "      <td>312002</td>\n",
       "      <td>2</td>\n",
       "      <td>GTA</td>\n",
       "      <td>GCA</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "      <td>..C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele gene_id mrna_id prot_id strand  \\\n",
       "0   Chr2   312000          G          A   gene2   mrna2   prot2      +   \n",
       "1   Chr2   312001          T          C   gene2   mrna2   prot2      +   \n",
       "\n",
       "          effect  snp_cds_pos  codon1_genome_pos  codon2_genome_pos  \\\n",
       "0  nonsynonymous            4             312000             312001   \n",
       "1  nonsynonymous            5             312000             312001   \n",
       "\n",
       "   codon3_genome_pos  snp_aa_pos ref_codon alt_codon ref_aa alt_aa pileup  \n",
       "0             312002           2       GTA       ATA      V      I   .A.A  \n",
       "1             312002           2       GTA       GCA      V      A    ..C  "
      ]
     },
     "execution_count": 312,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 313,
   "id": "7aaeb8de",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('A', 'T', 'A'), ('A', 'C', 'A'), ('G', 'T', 'A'), ('G', 'C', 'A')]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['ATA', 'ACA', 'GTA', 'GCA']"
      ]
     },
     "execution_count": 313,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import itertools\n",
    "\n",
    "permutations = list(itertools.product([\"A\",\"G\"], [\"T\",\"C\"], [\"A\"]))\n",
    "\n",
    "print(permutations)\n",
    "\n",
    "[\"\".join(x) for x in permutations]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 314,
   "id": "e944ea18",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{312000: 'A', 312001: 'C'}"
      ]
     },
     "execution_count": 314,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alt_dict = altDict(x2)\n",
    "alt_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "327cfdb8",
   "metadata": {},
   "source": [
    "## Running Case 2\n",
    "\n",
    "Things to Add: \n",
    "    * A control if nothing exists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 321,
   "id": "696b23c0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mrna_id</th>\n",
       "      <th>prot_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>effect</th>\n",
       "      <th>snp_cds_pos</th>\n",
       "      <th>...</th>\n",
       "      <th>snp_aa_pos</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>alt_codon</th>\n",
       "      <th>ref_aa</th>\n",
       "      <th>alt_aa</th>\n",
       "      <th>pileup</th>\n",
       "      <th>varify_codon</th>\n",
       "      <th>varify_allele</th>\n",
       "      <th>nt_VARified</th>\n",
       "      <th>varify_aa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312000</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gene2</td>\n",
       "      <td>mrna2</td>\n",
       "      <td>prot2</td>\n",
       "      <td>+</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>GTA</td>\n",
       "      <td>ATA</td>\n",
       "      <td>V</td>\n",
       "      <td>I</td>\n",
       "      <td>.A.A</td>\n",
       "      <td>complex_codon</td>\n",
       "      <td>A</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312001</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>gene2</td>\n",
       "      <td>mrna2</td>\n",
       "      <td>prot2</td>\n",
       "      <td>+</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>GTA</td>\n",
       "      <td>GCA</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "      <td>..C</td>\n",
       "      <td>complex_codon</td>\n",
       "      <td>C</td>\n",
       "      <td>Yes</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele gene_id mrna_id prot_id strand  \\\n",
       "0   Chr2   312000          G          A   gene2   mrna2   prot2      +   \n",
       "1   Chr2   312001          T          C   gene2   mrna2   prot2      +   \n",
       "\n",
       "          effect  snp_cds_pos  ...  snp_aa_pos  ref_codon  alt_codon  ref_aa  \\\n",
       "0  nonsynonymous            4  ...           2        GTA        ATA       V   \n",
       "1  nonsynonymous            5  ...           2        GTA        GCA       V   \n",
       "\n",
       "  alt_aa pileup   varify_codon varify_allele nt_VARified varify_aa  \n",
       "0      I   .A.A  complex_codon             A         Yes            \n",
       "1      A    ..C  complex_codon             C         Yes            \n",
       "\n",
       "[2 rows x 23 columns]"
      ]
     },
     "execution_count": 321,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_df = x2.apply(axis=1, func=snpDict)\n",
    "snp_df = pd.concat(list(snp_df))\n",
    "df = pd.merge(x2, snp_df, on='snp_pos')\n",
    "\n",
    "# Add the dictionary\n",
    "df = df.apply(axis=1, func=altValidate)\n",
    "df = df.apply(axis=1, func=translate)\n",
    "\n",
    "# reorg = ['chr_id', 'snp_pos', \n",
    "#          'ref_allele', 'alt_allele','varify_allele', 'VARified', \n",
    "#          'codon1_genome_pos', 'codon2_genome_pos', 'codon3_genome_pos',\n",
    "#          'ref_codon', 'alt_codon','varify_codon', \n",
    "#          'ref_aa', 'alt_aa', 'varify_aa', #\"alt_codons\",'flag'\n",
    "#          #'effect'\n",
    "#         ]\n",
    "\n",
    "df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19d574f4",
   "metadata": {},
   "source": [
    "# Case 2.5 - Both Case 1 and Case 2 on one "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 302,
   "id": "87b97300",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>3</th>\n",
       "      <th>pileup</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100021</td>\n",
       "      <td>T</td>\n",
       "      <td>12</td>\n",
       "      <td>.............</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100022</td>\n",
       "      <td>G</td>\n",
       "      <td>13</td>\n",
       "      <td>.............</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>G</td>\n",
       "      <td>13</td>\n",
       "      <td>...A...A..A..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312000</td>\n",
       "      <td>G</td>\n",
       "      <td>12</td>\n",
       "      <td>.A.A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312001</td>\n",
       "      <td>T</td>\n",
       "      <td>13</td>\n",
       "      <td>..C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312002</td>\n",
       "      <td>A</td>\n",
       "      <td>13</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0  snp_pos ref   3         pileup\n",
       "0  Chr1  1100021   T  12  .............\n",
       "1  Chr1  1100022   G  13  .............\n",
       "2  Chr1  1100023   G  13  ...A...A..A..\n",
       "0  Chr2   312000   G  12           .A.A\n",
       "1  Chr2   312001   T  13            ..C\n",
       "2  Chr2   312002   A  13            ..."
      ]
     },
     "execution_count": 302,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# need to merge the mpileups\n",
    "mp3 = pd.concat([mp1, mp2])\n",
    "\n",
    "# this will allow us to create one dictionary\n",
    "mp3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 303,
   "id": "8b38c7f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mrna_id</th>\n",
       "      <th>prot_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>effect</th>\n",
       "      <th>snp_cds_pos</th>\n",
       "      <th>codon1_genome_pos</th>\n",
       "      <th>codon2_genome_pos</th>\n",
       "      <th>codon3_genome_pos</th>\n",
       "      <th>snp_aa_pos</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>alt_codon</th>\n",
       "      <th>ref_aa</th>\n",
       "      <th>alt_aa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gene1</td>\n",
       "      <td>mrna1</td>\n",
       "      <td>prot1</td>\n",
       "      <td>+</td>\n",
       "      <td>stop_gain</td>\n",
       "      <td>21</td>\n",
       "      <td>1100021</td>\n",
       "      <td>1100022</td>\n",
       "      <td>1100023</td>\n",
       "      <td>7</td>\n",
       "      <td>TGG</td>\n",
       "      <td>TGA</td>\n",
       "      <td>W</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312000</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gene2</td>\n",
       "      <td>mrna2</td>\n",
       "      <td>prot2</td>\n",
       "      <td>+</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>4</td>\n",
       "      <td>312000</td>\n",
       "      <td>312001</td>\n",
       "      <td>312002</td>\n",
       "      <td>2</td>\n",
       "      <td>GTA</td>\n",
       "      <td>ATA</td>\n",
       "      <td>V</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312001</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>gene2</td>\n",
       "      <td>mrna2</td>\n",
       "      <td>prot2</td>\n",
       "      <td>+</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>5</td>\n",
       "      <td>312000</td>\n",
       "      <td>312001</td>\n",
       "      <td>312002</td>\n",
       "      <td>2</td>\n",
       "      <td>GTA</td>\n",
       "      <td>GCA</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele gene_id mrna_id prot_id strand  \\\n",
       "0   Chr1  1100023          G          A   gene1   mrna1   prot1      +   \n",
       "1   Chr2   312000          G          A   gene2   mrna2   prot2      +   \n",
       "2   Chr2   312001          T          C   gene2   mrna2   prot2      +   \n",
       "\n",
       "          effect  snp_cds_pos  codon1_genome_pos  codon2_genome_pos  \\\n",
       "0      stop_gain           21            1100021            1100022   \n",
       "1  nonsynonymous            4             312000             312001   \n",
       "2  nonsynonymous            5             312000             312001   \n",
       "\n",
       "   codon3_genome_pos  snp_aa_pos ref_codon alt_codon ref_aa alt_aa  \n",
       "0            1100023           7       TGG       TGA      W      *  \n",
       "1             312002           2       GTA       ATA      V      I  \n",
       "2             312002           2       GTA       GCA      V      A  "
      ]
     },
     "execution_count": 303,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# toysnpeff2 has the snptable info for everything\n",
    "toy_snpeff2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 328,
   "id": "05afa772",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mrna_id</th>\n",
       "      <th>prot_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>effect</th>\n",
       "      <th>snp_cds_pos</th>\n",
       "      <th>...</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>alt_codon</th>\n",
       "      <th>ref_aa</th>\n",
       "      <th>alt_aa</th>\n",
       "      <th>varify_codon</th>\n",
       "      <th>varify_allele</th>\n",
       "      <th>varify_aa</th>\n",
       "      <th>nt_VARified</th>\n",
       "      <th>codon_VARified</th>\n",
       "      <th>aa_VARified</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1100023</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gene1</td>\n",
       "      <td>mrna1</td>\n",
       "      <td>prot1</td>\n",
       "      <td>+</td>\n",
       "      <td>stop_gain</td>\n",
       "      <td>21</td>\n",
       "      <td>...</td>\n",
       "      <td>TGG</td>\n",
       "      <td>TGA</td>\n",
       "      <td>W</td>\n",
       "      <td>*</td>\n",
       "      <td>TGA</td>\n",
       "      <td>A</td>\n",
       "      <td>*</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312000</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gene2</td>\n",
       "      <td>mrna2</td>\n",
       "      <td>prot2</td>\n",
       "      <td>+</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>GTA</td>\n",
       "      <td>ATA</td>\n",
       "      <td>V</td>\n",
       "      <td>I</td>\n",
       "      <td>complex_codon</td>\n",
       "      <td>A</td>\n",
       "      <td></td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>312001</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>gene2</td>\n",
       "      <td>mrna2</td>\n",
       "      <td>prot2</td>\n",
       "      <td>+</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>GTA</td>\n",
       "      <td>GCA</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "      <td>complex_codon</td>\n",
       "      <td>C</td>\n",
       "      <td></td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele gene_id mrna_id prot_id strand  \\\n",
       "0   Chr1  1100023          G          A   gene1   mrna1   prot1      +   \n",
       "1   Chr2   312000          G          A   gene2   mrna2   prot2      +   \n",
       "2   Chr2   312001          T          C   gene2   mrna2   prot2      +   \n",
       "\n",
       "          effect  snp_cds_pos  ...  ref_codon  alt_codon  ref_aa  alt_aa  \\\n",
       "0      stop_gain           21  ...        TGG        TGA       W       *   \n",
       "1  nonsynonymous            4  ...        GTA        ATA       V       I   \n",
       "2  nonsynonymous            5  ...        GTA        GCA       V       A   \n",
       "\n",
       "    varify_codon varify_allele varify_aa nt_VARified codon_VARified  \\\n",
       "0            TGA             A         *         Yes            Yes   \n",
       "1  complex_codon             A                   Yes             No   \n",
       "2  complex_codon             C                   Yes             No   \n",
       "\n",
       "  aa_VARified  \n",
       "0         Yes  \n",
       "1          No  \n",
       "2          No  \n",
       "\n",
       "[3 rows x 24 columns]"
      ]
     },
     "execution_count": 328,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merging the two tables\n",
    "x = mergeTab(mp3, toy_snpeff2)\n",
    "\n",
    "alt_dict = altDict(x)\n",
    "\n",
    "# running the script\n",
    "snp_df = x.apply(axis=1, func=snpDict)\n",
    "snp_df = pd.concat(list(snp_df))\n",
    "df = pd.merge(x, snp_df, on='snp_pos')\n",
    "\n",
    "# Add the dictionary\n",
    "df = df.apply(axis=1, func=translate)\n",
    "df = df.apply(axis=1, func=altValidate)\n",
    "df = df.apply(axis=1, func=codonVarify)\n",
    "df = df.apply(axis=1, func=aaVarify)\n",
    "df = df.drop(columns='pileup', axis=1)\n",
    "\n",
    "df\n",
    "\n",
    "# reorg = ['chr_id', 'snp_pos', 'ref_allele', 'alt_allele', \n",
    "#          #'gene_id', 'mrna_id','prot_id', 'strand', \n",
    "#          'effect', #'snp_cds_pos', \n",
    "#          'codon1_genome_pos','codon2_genome_pos', 'codon3_genome_pos', \n",
    "#          #'snp_aa_pos', \n",
    "#          'ref_codon','alt_codon', 'ref_aa', 'alt_aa', 'varify_codon', 'varify_allele',\n",
    "#          'varify_aa', 'nt_VARified', 'codon_VARified', 'aa_VARified']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 331,
   "id": "1d050ee2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = varify(x)\n",
    "df.to_csv('../data/output/varify_output.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9131e49b",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Real World Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6a3b353",
   "metadata": {},
   "outputs": [],
   "source": [
    "import io\n",
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "def read_vcf(path):\n",
    "    with open(path, 'r') as f:\n",
    "        lines = [l for l in f if not l.startswith('##')]\n",
    "    return pd.read_csv(\n",
    "        io.StringIO(''.join(lines)),\n",
    "        dtype={'#CHROM': str, 'POS': int, 'ID': str, 'REF': str, 'ALT': str,\n",
    "               'QUAL': str, 'FILTER': str, 'INFO': str},\n",
    "        sep='\\t'\n",
    "    ).rename(columns={'#CHROM': 'CHROM'})\n",
    "\n",
    "vcf_file = '../data/ENCFF853AEO.vcf'\n",
    "\n",
    "vcf = read_vcf(vcf_file)\n",
    "\n",
    "vcf['POS'] == 822764"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf4babf8",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf[vcf['POS'] >= 876431]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82cf17f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcffile2 = '../data/ENCFF541HLI.vcf'\n",
    "vcf2 = read_vcf(vcffile2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c021232",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf2[vcf2['POS'] >= 876431] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ceaf96aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "mp3 = readMP('../data/out2.mpileup')\n",
    "\n",
    "mp3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d14b4608",
   "metadata": {},
   "outputs": [],
   "source": [
    "mp3[mp3['snp_pos'] >= 876887]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "303883ae",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
