{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1106b087",
   "metadata": {},
   "source": [
    "# VARify"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c87d4722",
   "metadata": {},
   "source": [
    "\n",
    "## Workflow\n",
    "\n",
    "\n",
    "<img src=\"wf.png\" width=\"600\">"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b3680f4",
   "metadata": {},
   "source": [
    "## Exploring Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "70705ac3",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "# mpileup \n",
    "mp = pd.read_table('../data/test/mpiletest-out.txt', header=None)\n",
    "mp = mp.rename(columns={1:\"snp_pos\", 4:\"pileup\"})\n",
    "\n",
    "# custom out file\n",
    "cf = pd.read_table('../data/test/snp_out.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ccd793ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>codon1_genome_pos</th>\n",
       "      <th>codon2_genome_pos</th>\n",
       "      <th>codon3_genome_pos</th>\n",
       "      <th>ref_codon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>521</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>519</td>\n",
       "      <td>520</td>\n",
       "      <td>521</td>\n",
       "      <td>AAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>523</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>521</td>\n",
       "      <td>522</td>\n",
       "      <td>523</td>\n",
       "      <td>GTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>524</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>522</td>\n",
       "      <td>523</td>\n",
       "      <td>524</td>\n",
       "      <td>TAC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele  codon1_genome_pos  codon2_genome_pos  \\\n",
       "0   chr1      521          G          C                519                520   \n",
       "1   chr1      523          A          T                521                522   \n",
       "2   chr1      524          C          G                522                523   \n",
       "\n",
       "   codon3_genome_pos ref_codon  \n",
       "0                521       AAG  \n",
       "1                523       GTA  \n",
       "2                524       TAC  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# column names\n",
    "# cols = ['chr_id', 'snp_pos', 'ref_allele', \n",
    "#         'alt_allele', 'gene_id', 'mrna_id',\n",
    "#         'prot_id', 'strand', 'effect', \n",
    "#         'snp_cds_pos', 'codon1_genome_pos',\n",
    "#         'codon2_genome_pos', 'codon3_genome_pos', \n",
    "#         'snp_aa_pos', 'ref_codon','alt_codon', \n",
    "#         'ref_aa', 'alt_aa']\n",
    "\n",
    "sub_cols = ['chr_id', 'snp_pos', 'ref_allele', \n",
    "            'alt_allele', 'codon1_genome_pos',\n",
    "            'codon2_genome_pos', 'codon3_genome_pos', \n",
    "            'ref_codon']\n",
    "\n",
    "# subset of custom file, cf\n",
    "sub_cf = cf[sub_cols]\n",
    "sub_cf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "509754eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>pileup</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>seq1</td>\n",
       "      <td>521</td>\n",
       "      <td>G</td>\n",
       "      <td>26</td>\n",
       "      <td>.,......A.a.A,..,..,.,.,,^].</td>\n",
       "      <td>DJJJJJJJJJJJJJJJJJJJJJJJA7</td>\n",
       "      <td>]]]]]]]]]]]]]]]]]]]]]]]]]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>seq1</td>\n",
       "      <td>522</td>\n",
       "      <td>T</td>\n",
       "      <td>25</td>\n",
       "      <td>,...........,..,..,.,.,,.</td>\n",
       "      <td>JJJJJJJJJJJJJJJJJJJJJJJA7</td>\n",
       "      <td>]]]]]]]]]]]]]]]]]]]]]]]]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>seq1</td>\n",
       "      <td>523</td>\n",
       "      <td>A</td>\n",
       "      <td>25</td>\n",
       "      <td>tTTTTTTTTTTTtTTtTTtTtTttT</td>\n",
       "      <td>JJJJJJJJJJJJJJJJJJJJJJJA7</td>\n",
       "      <td>]]]]]]]]]]]]]]]]]]]]]]]]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>seq1</td>\n",
       "      <td>524</td>\n",
       "      <td>C</td>\n",
       "      <td>27</td>\n",
       "      <td>,...G.......,.G,..,.,.,,G..</td>\n",
       "      <td>JJJJJJJJJJJJJJJJJJJJJJJJG88</td>\n",
       "      <td>]]]]]]]]]]]]]]]]]]]]]]]]]]]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0  snp_pos  2   3                        pileup  \\\n",
       "0  seq1      521  G  26  .,......A.a.A,..,..,.,.,,^].   \n",
       "1  seq1      522  T  25     ,...........,..,..,.,.,,.   \n",
       "2  seq1      523  A  25     tTTTTTTTTTTTtTTtTTtTtTttT   \n",
       "3  seq1      524  C  27   ,...G.......,.G,..,.,.,,G..   \n",
       "\n",
       "                             5                            6  \n",
       "0   DJJJJJJJJJJJJJJJJJJJJJJJA7   ]]]]]]]]]]]]]]]]]]]]]]]]]]  \n",
       "1    JJJJJJJJJJJJJJJJJJJJJJJA7    ]]]]]]]]]]]]]]]]]]]]]]]]]  \n",
       "2    JJJJJJJJJJJJJJJJJJJJJJJA7    ]]]]]]]]]]]]]]]]]]]]]]]]]  \n",
       "3  JJJJJJJJJJJJJJJJJJJJJJJJG88  ]]]]]]]]]]]]]]]]]]]]]]]]]]]  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mp"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26ef5f9d",
   "metadata": {},
   "source": [
    "## Merging the two dataframes together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "98f81b15",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mrna_id</th>\n",
       "      <th>prot_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>effect</th>\n",
       "      <th>snp_cds_pos</th>\n",
       "      <th>codon1_genome_pos</th>\n",
       "      <th>codon2_genome_pos</th>\n",
       "      <th>codon3_genome_pos</th>\n",
       "      <th>snp_aa_pos</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>alt_codon</th>\n",
       "      <th>ref_aa</th>\n",
       "      <th>alt_aa</th>\n",
       "      <th>pileup</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>521</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>seq1</td>\n",
       "      <td>seq1.1</td>\n",
       "      <td>seq1.1</td>\n",
       "      <td>+</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>265</td>\n",
       "      <td>519</td>\n",
       "      <td>520</td>\n",
       "      <td>521</td>\n",
       "      <td>89</td>\n",
       "      <td>AAG</td>\n",
       "      <td>AAC</td>\n",
       "      <td>K</td>\n",
       "      <td>N</td>\n",
       "      <td>.,......A.a.A,..,..,.,.,,^].</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>523</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>seq1</td>\n",
       "      <td>seq1.1</td>\n",
       "      <td>seq1.1</td>\n",
       "      <td>+</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>267</td>\n",
       "      <td>521</td>\n",
       "      <td>522</td>\n",
       "      <td>523</td>\n",
       "      <td>89</td>\n",
       "      <td>GTA</td>\n",
       "      <td>GTT</td>\n",
       "      <td>E</td>\n",
       "      <td>E</td>\n",
       "      <td>tTTTTTTTTTTTtTTtTTtTtTttT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>524</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>seq1</td>\n",
       "      <td>seq1.1</td>\n",
       "      <td>seq1.1</td>\n",
       "      <td>+</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>268</td>\n",
       "      <td>522</td>\n",
       "      <td>523</td>\n",
       "      <td>524</td>\n",
       "      <td>89</td>\n",
       "      <td>TAC</td>\n",
       "      <td>TAG</td>\n",
       "      <td>Y</td>\n",
       "      <td>X</td>\n",
       "      <td>,...G.......,.G,..,.,.,,G..</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele gene_id mrna_id prot_id strand  \\\n",
       "0   chr1      521          G          C    seq1  seq1.1  seq1.1      +   \n",
       "1   chr1      523          A          T    seq1  seq1.1  seq1.1      +   \n",
       "2   chr1      524          C          G    seq1  seq1.1  seq1.1      +   \n",
       "\n",
       "       effect  snp_cds_pos  codon1_genome_pos  codon2_genome_pos  \\\n",
       "0  synonymous          265                519                520   \n",
       "1  synonymous          267                521                522   \n",
       "2  synonymous          268                522                523   \n",
       "\n",
       "   codon3_genome_pos  snp_aa_pos ref_codon alt_codon ref_aa alt_aa  \\\n",
       "0                521          89       AAG       AAC      K      N   \n",
       "1                523          89       GTA       GTT      E      E   \n",
       "2                524          89       TAC       TAG      Y      X   \n",
       "\n",
       "                         pileup  \n",
       "0  .,......A.a.A,..,..,.,.,,^].  \n",
       "1     tTTTTTTTTTTTtTTtTTtTtTttT  \n",
       "2   ,...G.......,.G,..,.,.,,G..  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sub_mp = mp[['snp_pos','pileup']]\n",
    "merged_var = pd.merge(cf, sub_mp, on='snp_pos')\n",
    "merged_var"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13186127",
   "metadata": {},
   "source": [
    "## Split Dataframe\n",
    "\n",
    "Splitting the snpeff table by codons, get genome position and alt/ref codon\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "48e8d272",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Utility File\n",
    "\n",
    "# libraries\n",
    "import warnings\n",
    "import pandas as pd\n",
    "\n",
    "# For every row in here, we're going to apply this uniqueSNP function\n",
    "\n",
    "def mergeTab(mp, so):\n",
    "    sub_mp = mp[['snp_pos', 'pileup']]\n",
    "    mdf = pd.merge(so, sub_mp, on='snp_pos')\n",
    "    return mdf\n",
    "\n",
    "\n",
    "def varOnly(a):\n",
    "    '''\n",
    "    purpose: remove non-nt values in pileup\n",
    "    input: string of mpileup \n",
    "    output: nt string\n",
    "    '''\n",
    "    x = [i for i in a if i.upper() in ['A','C','T','G']]\n",
    "    \n",
    "    return x\n",
    "\n",
    "def isMono(ref, temp):\n",
    "    # Check if two snps are monomorphic\n",
    "    c=0\n",
    "    count = [c+1 for snp in temp if snp == ref]\n",
    "    if sum(count) >= 2: \n",
    "        warnings.warn(\"More than 2 monomorphic alleles at position.\")\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "    \n",
    "def multiError(x, pos):\n",
    "    if len(x) > 2: \n",
    "        warnings.warn(\"There are more than 2 SNPS at position\"); print(pos)\n",
    "    \n",
    "def uniqueSNP(x):\n",
    "    '''\n",
    "    Creates a combination of SNPs\n",
    "    Replaces Alt at each position\n",
    "    '''\n",
    "    l = str(x['pileup'])\n",
    "    # make upper\n",
    "    l = l.upper() # just to make simple\n",
    "    l = varOnly(l)\n",
    "\n",
    "    # Get list of unique\n",
    "    a = [i for i in l]\n",
    "    a_unique = list(set(a)) \n",
    "    \n",
    "    # create a dictionary\n",
    "    df = pd.DataFrame()\n",
    "    temp = [i for i in x['ref_codon']]\n",
    "    snp_dict = {}\n",
    "    \n",
    "#     flag = False\n",
    "    \n",
    "    multiError(a_unique, x['snp_pos'])\n",
    "\n",
    "    test = []\n",
    "    for snp in a_unique:\n",
    "        snp_dict['snp_pos'] = int(x['snp_pos'])\n",
    "        snp_dict['varify_alt'] = snp\n",
    "\n",
    "#         if flag:\n",
    "#             snp_dict['flag'] = '*'\n",
    "#         else:\n",
    "#             snp_dict['flag'] = ''\n",
    "        \n",
    "        for i in range(len(temp)):\n",
    "            temp[i] = snp\n",
    "            test.append(''.join(temp))\n",
    "            snp_dict[f\"varify_codon{i+1}\"] = ''.join(temp)\n",
    "            temp = [i for i in ref]\n",
    "            \n",
    "        flag = isMono(ref, test)  \n",
    "        \n",
    "        if flag: \n",
    "            snp_dict['flag'] = '*'\n",
    "        else:\n",
    "            snp_dict['flag'] = ''\n",
    "            \n",
    "        df = df.append(snp_dict, ignore_index=True)\n",
    "\n",
    "\n",
    "        snp_dict = {}\n",
    "        test = []\n",
    "\n",
    "    df = pd.merge(cf, df, on='snp_pos')\n",
    "    return df\n",
    "\n",
    "def altValidate(x):\n",
    "    # Varify alt == Alt? T or F\n",
    "    # Add flag to capture\n",
    "    \n",
    "    x['VARified'] = 'Yes' if x['alt_allele'] == x['varify_alt'] else 'No'\n",
    "    \n",
    "    return x\n",
    "\n",
    "def varify(x):\n",
    "    df = x.apply(axis=1, func=uniqueSNP)\n",
    "    df = pd.concat(list(df))\n",
    "    df = df.apply(axis=1, func=altValidate)\n",
    "    reorg = ['chr_id', 'snp_pos', 'ref_allele', 'alt_allele', \n",
    "             'varify_alt', 'VARified', \n",
    "             'codon1_genome_pos','codon2_genome_pos',\n",
    "             'codon3_genome_pos', 'ref_codon', 'varify_codon1', \n",
    "             'varify_codon2', 'varify_codon3','flag']\n",
    "    \n",
    "    return df[reorg]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "7eac0ac4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>varify_alt</th>\n",
       "      <th>VARified</th>\n",
       "      <th>codon1_genome_pos</th>\n",
       "      <th>codon2_genome_pos</th>\n",
       "      <th>codon3_genome_pos</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>varify_codon1</th>\n",
       "      <th>varify_codon2</th>\n",
       "      <th>varify_codon3</th>\n",
       "      <th>flag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>521</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>No</td>\n",
       "      <td>519</td>\n",
       "      <td>520</td>\n",
       "      <td>521</td>\n",
       "      <td>AAG</td>\n",
       "      <td>AAG</td>\n",
       "      <td>AAG</td>\n",
       "      <td>AAA</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>523</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>Yes</td>\n",
       "      <td>521</td>\n",
       "      <td>522</td>\n",
       "      <td>523</td>\n",
       "      <td>GTA</td>\n",
       "      <td>TTA</td>\n",
       "      <td>GTA</td>\n",
       "      <td>GTT</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>524</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>Yes</td>\n",
       "      <td>522</td>\n",
       "      <td>523</td>\n",
       "      <td>524</td>\n",
       "      <td>TAC</td>\n",
       "      <td>GAC</td>\n",
       "      <td>TGC</td>\n",
       "      <td>TAG</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele varify_alt VARified  \\\n",
       "0   chr1      521          G          C          A       No   \n",
       "0   chr1      523          A          T          T      Yes   \n",
       "0   chr1      524          C          G          G      Yes   \n",
       "\n",
       "   codon1_genome_pos  codon2_genome_pos  codon3_genome_pos ref_codon  \\\n",
       "0                519                520                521       AAG   \n",
       "0                521                522                523       GTA   \n",
       "0                522                523                524       TAC   \n",
       "\n",
       "  varify_codon1 varify_codon2 varify_codon3 flag  \n",
       "0           AAG           AAG           AAA    *  \n",
       "0           TTA           GTA           GTT       \n",
       "0           GAC           TGC           TAG       "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdf = mergeTab(mp, cf)\n",
    "varify(merged_var)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c37f09f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>pileup</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>seq1</td>\n",
       "      <td>521</td>\n",
       "      <td>G</td>\n",
       "      <td>26</td>\n",
       "      <td>.,......C.T.A,..,..,.,.,,^].</td>\n",
       "      <td>DJJJJJJJJJJJJJJJJJJJJJJJA7</td>\n",
       "      <td>]]]]]]]]]]]]]]]]]]]]]]]]]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>seq1</td>\n",
       "      <td>522</td>\n",
       "      <td>T</td>\n",
       "      <td>25</td>\n",
       "      <td>,...........,..,..,.,.,,.</td>\n",
       "      <td>JJJJJJJJJJJJJJJJJJJJJJJA7</td>\n",
       "      <td>]]]]]]]]]]]]]]]]]]]]]]]]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>seq1</td>\n",
       "      <td>523</td>\n",
       "      <td>A</td>\n",
       "      <td>25</td>\n",
       "      <td>tTTTTTTTTCTTtTTtTCtTtTttC</td>\n",
       "      <td>JJJJJJJJJJJJJJJJJJJJJJJA7</td>\n",
       "      <td>]]]]]]]]]]]]]]]]]]]]]]]]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>seq1</td>\n",
       "      <td>524</td>\n",
       "      <td>C</td>\n",
       "      <td>27</td>\n",
       "      <td>,...G.......,.G,..,.,.,,A..</td>\n",
       "      <td>JJJJJJJJJJJJJJJJJJJJJJJJG88</td>\n",
       "      <td>]]]]]]]]]]]]]]]]]]]]]]]]]]]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0  snp_pos  2   3                        pileup  \\\n",
       "0  seq1      521  G  26  .,......C.T.A,..,..,.,.,,^].   \n",
       "1  seq1      522  T  25     ,...........,..,..,.,.,,.   \n",
       "2  seq1      523  A  25     tTTTTTTTTCTTtTTtTCtTtTttC   \n",
       "3  seq1      524  C  27   ,...G.......,.G,..,.,.,,A..   \n",
       "\n",
       "                             5                            6  \n",
       "0   DJJJJJJJJJJJJJJJJJJJJJJJA7   ]]]]]]]]]]]]]]]]]]]]]]]]]]  \n",
       "1    JJJJJJJJJJJJJJJJJJJJJJJA7    ]]]]]]]]]]]]]]]]]]]]]]]]]  \n",
       "2    JJJJJJJJJJJJJJJJJJJJJJJA7    ]]]]]]]]]]]]]]]]]]]]]]]]]  \n",
       "3  JJJJJJJJJJJJJJJJJJJJJJJJG88  ]]]]]]]]]]]]]]]]]]]]]]]]]]]  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# pile up \n",
    "mp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "38ae94f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mrna_id</th>\n",
       "      <th>prot_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>effect</th>\n",
       "      <th>snp_cds_pos</th>\n",
       "      <th>codon1_genome_pos</th>\n",
       "      <th>codon2_genome_pos</th>\n",
       "      <th>codon3_genome_pos</th>\n",
       "      <th>snp_aa_pos</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>alt_codon</th>\n",
       "      <th>ref_aa</th>\n",
       "      <th>alt_aa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>521</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>seq1</td>\n",
       "      <td>seq1.1</td>\n",
       "      <td>seq1.1</td>\n",
       "      <td>+</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>265</td>\n",
       "      <td>519</td>\n",
       "      <td>520</td>\n",
       "      <td>521</td>\n",
       "      <td>89</td>\n",
       "      <td>AAG</td>\n",
       "      <td>AAC</td>\n",
       "      <td>K</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>523</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>seq1</td>\n",
       "      <td>seq1.1</td>\n",
       "      <td>seq1.1</td>\n",
       "      <td>+</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>267</td>\n",
       "      <td>521</td>\n",
       "      <td>522</td>\n",
       "      <td>523</td>\n",
       "      <td>89</td>\n",
       "      <td>GTA</td>\n",
       "      <td>GTT</td>\n",
       "      <td>E</td>\n",
       "      <td>E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>524</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>seq1</td>\n",
       "      <td>seq1.1</td>\n",
       "      <td>seq1.1</td>\n",
       "      <td>+</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>268</td>\n",
       "      <td>522</td>\n",
       "      <td>523</td>\n",
       "      <td>524</td>\n",
       "      <td>89</td>\n",
       "      <td>TAC</td>\n",
       "      <td>TAG</td>\n",
       "      <td>Y</td>\n",
       "      <td>X</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele gene_id mrna_id prot_id strand  \\\n",
       "0   chr1      521          G          C    seq1  seq1.1  seq1.1      +   \n",
       "1   chr1      523          A          T    seq1  seq1.1  seq1.1      +   \n",
       "2   chr1      524          C          G    seq1  seq1.1  seq1.1      +   \n",
       "\n",
       "       effect  snp_cds_pos  codon1_genome_pos  codon2_genome_pos  \\\n",
       "0  synonymous          265                519                520   \n",
       "1  synonymous          267                521                522   \n",
       "2  synonymous          268                522                523   \n",
       "\n",
       "   codon3_genome_pos  snp_aa_pos ref_codon alt_codon ref_aa alt_aa  \n",
       "0                521          89       AAG       AAC      K      N  \n",
       "1                523          89       GTA       GTT      E      E  \n",
       "2                524          89       TAC       TAG      Y      X  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# snp-eff like table\n",
    "cf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78742d03",
   "metadata": {},
   "source": [
    "### To Do\n",
    "* [x] output comparison\n",
    "* [ ] Type Effect\n",
    "* [x] Make into Function\n",
    "* [ ] Build larger test data\n",
    "* [ ] Nextflow with Assembly/Sam to mpileup\n",
    "\n",
    "\n",
    "### Notes\n",
    "**Test Data:**\n",
    "* Take processed mRNA transcript (FASTA) —> VCF w/ 2-3 SNPS\n",
    "\n",
    "**Process:** \n",
    "* RNA FASTA -align to genome using gmap->SAM/BAM—>BAM -variantcalling-> VCF —> Capture Variants\n",
    "* DNA FASTA -align to genome using gmap->SAM/BAM—>BAM -variantcalling-> VCF —> Capture Variants |—> Compare RV and GV"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32ede3ea",
   "metadata": {},
   "source": [
    "# VARify for same ALT as prediction\n",
    "\n",
    "| Pos     | Ref     | Alt |Varify_Alt|Varify|\n",
    "| ----------- | ----------- | ----------- | ----------- | ----------- |\n",
    "|202| A      | T       |T|Yes|\n",
    "|204| C   | G        |.|No|\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "efd4acb4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>varify_alt</th>\n",
       "      <th>VARified</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>521</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>523</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>524</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele varify_alt VARified\n",
       "0   chr1      521          G          C          A       No\n",
       "0   chr1      523          A          T          T      Yes\n",
       "0   chr1      524          C          G          G      Yes"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "varify_out = varify(merged_var)\n",
    "\n",
    "varify_out.iloc[:,:6]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "66f9acc4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>varify_alt</th>\n",
       "      <th>VARified</th>\n",
       "      <th>codon1_genome_pos</th>\n",
       "      <th>codon2_genome_pos</th>\n",
       "      <th>codon3_genome_pos</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>varify_codon1</th>\n",
       "      <th>varify_codon2</th>\n",
       "      <th>varify_codon3</th>\n",
       "      <th>flag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>523</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>Yes</td>\n",
       "      <td>521</td>\n",
       "      <td>522</td>\n",
       "      <td>523</td>\n",
       "      <td>GTA</td>\n",
       "      <td>TTA</td>\n",
       "      <td>GTA</td>\n",
       "      <td>GTT</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>524</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>Yes</td>\n",
       "      <td>522</td>\n",
       "      <td>523</td>\n",
       "      <td>524</td>\n",
       "      <td>TAC</td>\n",
       "      <td>GAC</td>\n",
       "      <td>TGC</td>\n",
       "      <td>TAG</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele varify_alt VARified  \\\n",
       "0   chr1      523          A          T          T      Yes   \n",
       "0   chr1      524          C          G          G      Yes   \n",
       "\n",
       "   codon1_genome_pos  codon2_genome_pos  codon3_genome_pos ref_codon  \\\n",
       "0                521                522                523       GTA   \n",
       "0                522                523                524       TAC   \n",
       "\n",
       "  varify_codon1 varify_codon2 varify_codon3 flag  \n",
       "0           TTA           GTA           GTT       \n",
       "0           GAC           TGC           TAG       "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "varify_out.loc[varify_out['VARified'] == 'Yes']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c7dcd34",
   "metadata": {},
   "source": [
    "# Capture codon combinations from different positions\n",
    "\n",
    "Logic: \n",
    "If position 1,2,3 and 1 and 3 have variances - we want to maintain these variances. \n",
    "\n",
    "We can place variances of 1 and 3 into dictionaries. \n",
    "If at position 3 and position 1 has variances --> combine and showcase AA\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "380eaf3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Codon Combos\n",
    "\n",
    "def altDict(x):\n",
    "    '''\n",
    "    \n",
    "    Creates a combination of SNPs\n",
    "    if snps have existing alts\n",
    "    \n",
    "    Outputs into a dictionary\n",
    "    \n",
    "    '''\n",
    "    # At each line, takes reference codon\n",
    "    sub_x = x.loc[x['VARified'] == 'Yes']\n",
    "    \n",
    "    # create a dictionary of Alts\n",
    "    alt_dict = dict(zip(sub_x['snp_pos'], sub_x['alt_allele']))\n",
    "    \n",
    "    return alt_dict\n",
    "\n",
    "def altCombo(x, alt_dict=alt_dict):\n",
    "    '''\n",
    "    Creates combos based on\n",
    "    Current position and alt dictionary\n",
    "    Can be used with apply\n",
    "    '''\n",
    "    comboDict = {}\n",
    "    p3 = int(x['snp_pos'])\n",
    "    p2 = p3-1\n",
    "    p1 = p2-1\n",
    "    \n",
    "    comboDict = {p1 : '.', p2 : '.', p3 : '.'}\n",
    "    \n",
    "    # create a dictionary\n",
    "    df = pd.DataFrame()\n",
    "    temp = [i for i in x['ref_codon']]\n",
    "    snp_dict = {}\n",
    "    \n",
    "    res = {key: alt_dict.get(key, comboDict[key]) for key in comboDict}\n",
    "    ''.join(res.values())\n",
    "    \n",
    "    # make combinations\n",
    "    for i in range(len(temp)):\n",
    "        temp[i] = snp\n",
    "        test.append(''.join(temp))\n",
    "        snp_dict[f\"varify_codon{i+1}\"] = ''.join(temp)\n",
    "        temp = [i for i in ref]\n",
    "\n",
    "    flag = isMono(ref, test)  \n",
    "\n",
    "    if flag: \n",
    "        snp_dict['flag'] = '*'\n",
    "    else:\n",
    "        snp_dict['flag'] = ''\n",
    "\n",
    "    df = df.append(snp_dict, ignore_index=True)\n",
    "\n",
    "\n",
    "    snp_dict = {}\n",
    "    test = []\n",
    "\n",
    "    df = pd.merge(cf, df, on='snp_pos')\n",
    "    \n",
    "    return comboDict\n",
    "\n",
    "\n",
    "def flatten(d, parent_key='', sep='_'):\n",
    "    import collections\n",
    "    items = []\n",
    "    for k, v in d.items():\n",
    "        new_key = parent_key + sep + k if parent_key else k\n",
    "        if isinstance(v, collections.MutableMapping):\n",
    "            items.extend(flatten(v, new_key, sep=sep).items())\n",
    "        else:\n",
    "            items.append((new_key, v))\n",
    "    return dict(items)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "6d6dfacc",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "\"None of [Int64Index([519, 520, 521], dtype='int64')] are in the [index]\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[111], line 12\u001b[0m\n\u001b[1;32m      9\u001b[0m combo_dict \u001b[38;5;241m=\u001b[39m varify_out\u001b[38;5;241m.\u001b[39mapply(axis \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m, func\u001b[38;5;241m=\u001b[39maltCombo)\n\u001b[1;32m     10\u001b[0m \u001b[38;5;66;03m# combo_dict = flatten(combo_dict)\u001b[39;00m\n\u001b[0;32m---> 12\u001b[0m res \u001b[38;5;241m=\u001b[39m {key: alt_dict\u001b[38;5;241m.\u001b[39mget(key, combo_dict[key]) \u001b[38;5;28;01mfor\u001b[39;00m key \u001b[38;5;129;01min\u001b[39;00m combo_dict}\n\u001b[1;32m     13\u001b[0m \u001b[38;5;66;03m#res = {key:value for (key, value) in res.items() if value is not None}\u001b[39;00m\n\u001b[1;32m     14\u001b[0m res\u001b[38;5;241m.\u001b[39mitems\n",
      "Cell \u001b[0;32mIn[111], line 12\u001b[0m, in \u001b[0;36m<dictcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m      9\u001b[0m combo_dict \u001b[38;5;241m=\u001b[39m varify_out\u001b[38;5;241m.\u001b[39mapply(axis \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m, func\u001b[38;5;241m=\u001b[39maltCombo)\n\u001b[1;32m     10\u001b[0m \u001b[38;5;66;03m# combo_dict = flatten(combo_dict)\u001b[39;00m\n\u001b[0;32m---> 12\u001b[0m res \u001b[38;5;241m=\u001b[39m {key: alt_dict\u001b[38;5;241m.\u001b[39mget(key, \u001b[43mcombo_dict\u001b[49m\u001b[43m[\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m]\u001b[49m) \u001b[38;5;28;01mfor\u001b[39;00m key \u001b[38;5;129;01min\u001b[39;00m combo_dict}\n\u001b[1;32m     13\u001b[0m \u001b[38;5;66;03m#res = {key:value for (key, value) in res.items() if value is not None}\u001b[39;00m\n\u001b[1;32m     14\u001b[0m res\u001b[38;5;241m.\u001b[39mitems\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/pandas/core/series.py:966\u001b[0m, in \u001b[0;36mSeries.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    963\u001b[0m     key \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39masarray(key, dtype\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mbool\u001b[39m)\n\u001b[1;32m    964\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_get_values(key)\n\u001b[0;32m--> 966\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_with\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/pandas/core/series.py:1001\u001b[0m, in \u001b[0;36mSeries._get_with\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    997\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m key_type \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124minteger\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m    998\u001b[0m     \u001b[38;5;66;03m# We need to decide whether to treat this as a positional indexer\u001b[39;00m\n\u001b[1;32m    999\u001b[0m     \u001b[38;5;66;03m#  (i.e. self.iloc) or label-based (i.e. self.loc)\u001b[39;00m\n\u001b[1;32m   1000\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mindex\u001b[38;5;241m.\u001b[39m_should_fallback_to_positional():\n\u001b[0;32m-> 1001\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mloc\u001b[49m\u001b[43m[\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m]\u001b[49m\n\u001b[1;32m   1002\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1003\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39miloc[key]\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/pandas/core/indexing.py:931\u001b[0m, in \u001b[0;36m_LocationIndexer.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    928\u001b[0m axis \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maxis \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;241m0\u001b[39m\n\u001b[1;32m    930\u001b[0m maybe_callable \u001b[38;5;241m=\u001b[39m com\u001b[38;5;241m.\u001b[39mapply_if_callable(key, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj)\n\u001b[0;32m--> 931\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_getitem_axis\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmaybe_callable\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/pandas/core/indexing.py:1153\u001b[0m, in \u001b[0;36m_LocIndexer._getitem_axis\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1150\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(key, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mndim\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;129;01mand\u001b[39;00m key\u001b[38;5;241m.\u001b[39mndim \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m   1151\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCannot index with multidimensional key\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m-> 1153\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_getitem_iterable\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1155\u001b[0m \u001b[38;5;66;03m# nested tuple slicing\u001b[39;00m\n\u001b[1;32m   1156\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_nested_tuple(key, labels):\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/pandas/core/indexing.py:1093\u001b[0m, in \u001b[0;36m_LocIndexer._getitem_iterable\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1090\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_validate_key(key, axis)\n\u001b[1;32m   1092\u001b[0m \u001b[38;5;66;03m# A collection of keys\u001b[39;00m\n\u001b[0;32m-> 1093\u001b[0m keyarr, indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_listlike_indexer\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1094\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj\u001b[38;5;241m.\u001b[39m_reindex_with_indexers(\n\u001b[1;32m   1095\u001b[0m     {axis: [keyarr, indexer]}, copy\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, allow_dups\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m   1096\u001b[0m )\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/pandas/core/indexing.py:1314\u001b[0m, in \u001b[0;36m_LocIndexer._get_listlike_indexer\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1311\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1312\u001b[0m     keyarr, indexer, new_indexer \u001b[38;5;241m=\u001b[39m ax\u001b[38;5;241m.\u001b[39m_reindex_non_unique(keyarr)\n\u001b[0;32m-> 1314\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_validate_read_indexer\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkeyarr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindexer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1316\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m needs_i8_conversion(ax\u001b[38;5;241m.\u001b[39mdtype) \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(\n\u001b[1;32m   1317\u001b[0m     ax, (IntervalIndex, CategoricalIndex)\n\u001b[1;32m   1318\u001b[0m ):\n\u001b[1;32m   1319\u001b[0m     \u001b[38;5;66;03m# For CategoricalIndex take instead of reindex to preserve dtype.\u001b[39;00m\n\u001b[1;32m   1320\u001b[0m     \u001b[38;5;66;03m#  For IntervalIndex this is to map integers to the Intervals they match to.\u001b[39;00m\n\u001b[1;32m   1321\u001b[0m     keyarr \u001b[38;5;241m=\u001b[39m ax\u001b[38;5;241m.\u001b[39mtake(indexer)\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/pandas/core/indexing.py:1374\u001b[0m, in \u001b[0;36m_LocIndexer._validate_read_indexer\u001b[0;34m(self, key, indexer, axis)\u001b[0m\n\u001b[1;32m   1372\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m use_interval_msg:\n\u001b[1;32m   1373\u001b[0m         key \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(key)\n\u001b[0;32m-> 1374\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNone of [\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mkey\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m] are in the [\u001b[39m\u001b[38;5;132;01m{\u001b[39;00maxis_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m]\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   1376\u001b[0m not_found \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(ensure_index(key)[missing_mask\u001b[38;5;241m.\u001b[39mnonzero()[\u001b[38;5;241m0\u001b[39m]]\u001b[38;5;241m.\u001b[39munique())\n\u001b[1;32m   1377\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mnot_found\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m not in index\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mKeyError\u001b[0m: \"None of [Int64Index([519, 520, 521], dtype='int64')] are in the [index]\""
     ]
    }
   ],
   "source": [
    "# At each line, takes reference codon\n",
    "sub_x = varify_out.loc[varify_out['VARified'] == 'Yes']\n",
    "\n",
    "# ref = sub_x['ref_codon']\n",
    "alt_dict = dict(zip(sub_x['snp_pos'], sub_x['alt_allele']))\n",
    "\n",
    "alt_dict\n",
    "\n",
    "combo_dict = varify_out.apply(axis = 1, func=altCombo)\n",
    "# combo_dict = flatten(combo_dict)\n",
    "\n",
    "res = {key: alt_dict.get(key, combo_dict[key]) for key in combo_dict}\n",
    "#res = {key:value for (key, value) in res.items() if value is not None}\n",
    "res.items"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "c32980b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'TG'"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3f708429",
   "metadata": {},
   "source": [
    "# Running Script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37bbd614",
   "metadata": {},
   "outputs": [],
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "%run '../script/main.py' --in_mp '../data/test/mpiletest-out.txt' --in_so '../data/test/snp_out.txt'"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
