{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1106b087",
   "metadata": {},
   "source": [
    "# VARify"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c87d4722",
   "metadata": {},
   "source": [
    "\n",
    "## Workflow\n",
    "\n",
    "\n",
    "<img src=\"wf.png\" width=\"600\">"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b3680f4",
   "metadata": {},
   "source": [
    "## Exploring Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "70705ac3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "# mpileup \n",
    "mp = pd.read_table('../data/test/mpiletest-out.txt', header=None)\n",
    "mp = mp.rename(columns={1:\"snp_pos\", 4:\"pileup\"})\n",
    "\n",
    "# custom out file\n",
    "cf = pd.read_table('../data/test/snp_out.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccd793ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# column names\n",
    "# cols = ['chr_id', 'snp_pos', 'ref_allele', \n",
    "#         'alt_allele', 'gene_id', 'mrna_id',\n",
    "#         'prot_id', 'strand', 'effect', \n",
    "#         'snp_cds_pos', 'codon1_genome_pos',\n",
    "#         'codon2_genome_pos', 'codon3_genome_pos', \n",
    "#         'snp_aa_pos', 'ref_codon','alt_codon', \n",
    "#         'ref_aa', 'alt_aa']\n",
    "\n",
    "sub_cols = ['chr_id', 'snp_pos', 'ref_allele', \n",
    "            'alt_allele', 'codon1_genome_pos',\n",
    "            'codon2_genome_pos', 'codon3_genome_pos', \n",
    "            'ref_codon']\n",
    "\n",
    "# subset of custom file, cf\n",
    "sub_cf = cf[sub_cols]\n",
    "sub_cf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26ef5f9d",
   "metadata": {},
   "source": [
    "## Merging the two dataframes together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "98f81b15",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mrna_id</th>\n",
       "      <th>prot_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>effect</th>\n",
       "      <th>snp_cds_pos</th>\n",
       "      <th>codon1_genome_pos</th>\n",
       "      <th>codon2_genome_pos</th>\n",
       "      <th>codon3_genome_pos</th>\n",
       "      <th>snp_aa_pos</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>alt_codon</th>\n",
       "      <th>ref_aa</th>\n",
       "      <th>alt_aa</th>\n",
       "      <th>pileup</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>521</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>seq1</td>\n",
       "      <td>seq1.1</td>\n",
       "      <td>seq1.1</td>\n",
       "      <td>+</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>265</td>\n",
       "      <td>519</td>\n",
       "      <td>520</td>\n",
       "      <td>521</td>\n",
       "      <td>89</td>\n",
       "      <td>AAG</td>\n",
       "      <td>AAC</td>\n",
       "      <td>K</td>\n",
       "      <td>N</td>\n",
       "      <td>.,......C.T.A,..,..,.,.,,^].</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>523</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>seq1</td>\n",
       "      <td>seq1.1</td>\n",
       "      <td>seq1.1</td>\n",
       "      <td>+</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>267</td>\n",
       "      <td>521</td>\n",
       "      <td>522</td>\n",
       "      <td>523</td>\n",
       "      <td>89</td>\n",
       "      <td>GTA</td>\n",
       "      <td>GTT</td>\n",
       "      <td>E</td>\n",
       "      <td>E</td>\n",
       "      <td>tTTTTTTTTCTTtTTtTCtTtTttC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>524</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>seq1</td>\n",
       "      <td>seq1.1</td>\n",
       "      <td>seq1.1</td>\n",
       "      <td>+</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>268</td>\n",
       "      <td>522</td>\n",
       "      <td>523</td>\n",
       "      <td>524</td>\n",
       "      <td>89</td>\n",
       "      <td>TAC</td>\n",
       "      <td>TAG</td>\n",
       "      <td>Y</td>\n",
       "      <td>X</td>\n",
       "      <td>,...G.......,.G,..,.,.,,A..</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele gene_id mrna_id prot_id strand  \\\n",
       "0   chr1      521          G          C    seq1  seq1.1  seq1.1      +   \n",
       "1   chr1      523          A          T    seq1  seq1.1  seq1.1      +   \n",
       "2   chr1      524          C          G    seq1  seq1.1  seq1.1      +   \n",
       "\n",
       "       effect  snp_cds_pos  codon1_genome_pos  codon2_genome_pos  \\\n",
       "0  synonymous          265                519                520   \n",
       "1  synonymous          267                521                522   \n",
       "2  synonymous          268                522                523   \n",
       "\n",
       "   codon3_genome_pos  snp_aa_pos ref_codon alt_codon ref_aa alt_aa  \\\n",
       "0                521          89       AAG       AAC      K      N   \n",
       "1                523          89       GTA       GTT      E      E   \n",
       "2                524          89       TAC       TAG      Y      X   \n",
       "\n",
       "                         pileup  \n",
       "0  .,......C.T.A,..,..,.,.,,^].  \n",
       "1     tTTTTTTTTCTTtTTtTCtTtTttC  \n",
       "2   ,...G.......,.G,..,.,.,,A..  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sub_mp = mp[['snp_pos','pileup']]\n",
    "merged_var = pd.merge(cf, sub_mp, on='snp_pos')\n",
    "merged_var"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13186127",
   "metadata": {},
   "source": [
    "## Split Dataframe\n",
    "\n",
    "Splitting the snpeff table by codons, get genome position and alt/ref codon\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "48e8d272",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Utility File\n",
    "\n",
    "# libraries\n",
    "import warnings\n",
    "import pandas as pd\n",
    "\n",
    "# For every row in here, we're going to apply this uniqueSNP function\n",
    "\n",
    "def mergeTab(mp, so):\n",
    "    sub_mp = mp[['snp_pos', 'pileup']]\n",
    "    mdf = pd.merge(so, sub_mp, on='snp_pos')\n",
    "    return mdf\n",
    "\n",
    "\n",
    "def varOnly(a):\n",
    "    '''\n",
    "    purpose: remove non-nt values in pileup\n",
    "    input: string of mpileup \n",
    "    output: nt string\n",
    "    '''\n",
    "    x = [i for i in a if i.upper() in ['A','C','T','G']]\n",
    "    \n",
    "    return x\n",
    "\n",
    "def uniqueSNP(x):\n",
    "    l = str(x['pileup'])\n",
    "    # make upper\n",
    "    l = l.upper() # just to make simple\n",
    "    l = varOnly(l)\n",
    "\n",
    "    # Get list of unique\n",
    "    a = [i for i in l]\n",
    "    a_unique = list(set(a)) \n",
    "    \n",
    "    # create a dictionary\n",
    "    df = pd.DataFrame()\n",
    "    temp = [i for i in x['ref_codon']]\n",
    "    snp_dict = {}\n",
    "    \n",
    "    flag = False\n",
    "    \n",
    "    if len(a_unique) > 2: \n",
    "        warnings.warn(\"There are more than 2 SNPS at position\"); print(x['snp_pos'])\n",
    "        flag = True\n",
    "\n",
    "    for snp in a_unique:\n",
    "        snp_dict['snp_pos'] = int(x['snp_pos'])\n",
    "        snp_dict['varify_alt'] = snp\n",
    "\n",
    "        if flag:\n",
    "            snp_dict['flag'] = '*'\n",
    "        else:\n",
    "            snp_dict['flag'] = ''\n",
    "        \n",
    "        for i in range(len(temp)):\n",
    "            temp[i] = snp     \n",
    "            snp_dict[f\"varify_codon{i+1}\"] = ''.join(temp)\n",
    "            temp = [i for i in x['ref_codon']]\n",
    "\n",
    "        df = df.append(snp_dict, ignore_index=True)\n",
    "        \n",
    "        snp_dict = {}\n",
    "\n",
    "    df = pd.merge(so, df, on='snp_pos')\n",
    "    return df\n",
    "\n",
    "\n",
    "def varify(x):\n",
    "    df = x.apply(axis=1, func=uniqueSNP)\n",
    "    df = pd.concat(list(df))\n",
    "    reorg = ['chr_id', 'snp_pos', 'ref_allele', 'alt_allele', \n",
    "             'varify_alt', 'codon1_genome_pos','codon2_genome_pos',\n",
    "             'codon3_genome_pos', 'ref_codon', 'varify_codon1', \n",
    "             'varify_codon2', 'varify_codon3','flag']\n",
    "\n",
    "    return df[reorg]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "7eac0ac4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "521\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-31-2363766a900e>:43: UserWarning: There are more than 2 SNPS at position\n",
      "  warnings.warn(\"There are more than 2 SNPS at position\"); print(x['snp_pos'])\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>varify_alt</th>\n",
       "      <th>codon1_genome_pos</th>\n",
       "      <th>codon2_genome_pos</th>\n",
       "      <th>codon3_genome_pos</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>varify_codon1</th>\n",
       "      <th>varify_codon2</th>\n",
       "      <th>varify_codon3</th>\n",
       "      <th>flag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>521</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>519</td>\n",
       "      <td>520</td>\n",
       "      <td>521</td>\n",
       "      <td>AAG</td>\n",
       "      <td>AAG</td>\n",
       "      <td>AAG</td>\n",
       "      <td>AAA</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>521</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>519</td>\n",
       "      <td>520</td>\n",
       "      <td>521</td>\n",
       "      <td>AAG</td>\n",
       "      <td>TAG</td>\n",
       "      <td>ATG</td>\n",
       "      <td>AAT</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>521</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>519</td>\n",
       "      <td>520</td>\n",
       "      <td>521</td>\n",
       "      <td>AAG</td>\n",
       "      <td>CAG</td>\n",
       "      <td>ACG</td>\n",
       "      <td>AAC</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>523</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>521</td>\n",
       "      <td>522</td>\n",
       "      <td>523</td>\n",
       "      <td>GTA</td>\n",
       "      <td>TTA</td>\n",
       "      <td>GTA</td>\n",
       "      <td>GTT</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>523</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>521</td>\n",
       "      <td>522</td>\n",
       "      <td>523</td>\n",
       "      <td>GTA</td>\n",
       "      <td>CTA</td>\n",
       "      <td>GCA</td>\n",
       "      <td>GTC</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>524</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>522</td>\n",
       "      <td>523</td>\n",
       "      <td>524</td>\n",
       "      <td>TAC</td>\n",
       "      <td>AAC</td>\n",
       "      <td>TAC</td>\n",
       "      <td>TAA</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>524</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>522</td>\n",
       "      <td>523</td>\n",
       "      <td>524</td>\n",
       "      <td>TAC</td>\n",
       "      <td>GAC</td>\n",
       "      <td>TGC</td>\n",
       "      <td>TAG</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_id  snp_pos ref_allele alt_allele varify_alt  codon1_genome_pos  \\\n",
       "0   chr1      521          G          C          A                519   \n",
       "1   chr1      521          G          C          T                519   \n",
       "2   chr1      521          G          C          C                519   \n",
       "0   chr1      523          A          T          T                521   \n",
       "1   chr1      523          A          T          C                521   \n",
       "0   chr1      524          C          G          A                522   \n",
       "1   chr1      524          C          G          G                522   \n",
       "\n",
       "   codon2_genome_pos  codon3_genome_pos ref_codon varify_codon1 varify_codon2  \\\n",
       "0                520                521       AAG           AAG           AAG   \n",
       "1                520                521       AAG           TAG           ATG   \n",
       "2                520                521       AAG           CAG           ACG   \n",
       "0                522                523       GTA           TTA           GTA   \n",
       "1                522                523       GTA           CTA           GCA   \n",
       "0                523                524       TAC           AAC           TAC   \n",
       "1                523                524       TAC           GAC           TGC   \n",
       "\n",
       "  varify_codon3 flag  \n",
       "0           AAA    *  \n",
       "1           AAT    *  \n",
       "2           AAC    *  \n",
       "0           GTT       \n",
       "1           GTC       \n",
       "0           TAA       \n",
       "1           TAG       "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdf = mergeTab(mp, so)\n",
    "varify(merged_var)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd189880",
   "metadata": {},
   "source": [
    "## Running script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "633cc6e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "521\n",
      "  chr_id  snp_pos ref_allele alt_allele varify_alt  codon1_genome_pos  \\\n",
      "0   chr1      521          G          C          A                519   \n",
      "1   chr1      521          G          C          T                519   \n",
      "2   chr1      521          G          C          C                519   \n",
      "0   chr1      523          A          T          T                521   \n",
      "1   chr1      523          A          T          C                521   \n",
      "0   chr1      524          C          G          A                522   \n",
      "1   chr1      524          C          G          G                522   \n",
      "\n",
      "   codon2_genome_pos  codon3_genome_pos ref_codon varify_codon1 varify_codon2  \\\n",
      "0                520                521       AAG           AAG           AAG   \n",
      "1                520                521       AAG           TAG           ATG   \n",
      "2                520                521       AAG           CAG           ACG   \n",
      "0                522                523       GTA           TTA           GTA   \n",
      "1                522                523       GTA           CTA           GCA   \n",
      "0                523                524       TAC           AAC           TAC   \n",
      "1                523                524       TAC           GAC           TGC   \n",
      "\n",
      "  varify_codon3 flag  \n",
      "0           AAA    *  \n",
      "1           AAT    *  \n",
      "2           AAC    *  \n",
      "0           GTT       \n",
      "1           GTC       \n",
      "0           TAA       \n",
      "1           TAG       \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/kkwock/Documents/GitHub/VARify/script/main.py:53: UserWarning: There are more than 2 SNPS at position\n",
      "  \n"
     ]
    }
   ],
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "%run '../script/main.py' --in_mp '../data/test/mpiletest-out.txt' --in_so '../data/test/snp_out.txt'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78742d03",
   "metadata": {},
   "source": [
    "### To Do\n",
    "* [x] output comparison\n",
    "* [ ] Type Effect\n",
    "* [x] Make into Function\n",
    "* [ ] Build larger test data\n",
    "* [ ] Nextflow with Assembly/Sam to mpileup\n",
    "\n",
    "\n",
    "### Notes\n",
    "**Test Data:**\n",
    "* Take processed mRNA transcript (FASTA) —> VCF w/ 2-3 SNPS\n",
    "\n",
    "**Process:** \n",
    "* RNA FASTA -align to genome using gmap->SAM/BAM—>BAM -variantcalling-> VCF —> Capture Variants\n",
    "* DNA FASTA -align to genome using gmap->SAM/BAM—>BAM -variantcalling-> VCF —> Capture Variants |—> Compare RV and GV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e34c8c5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "top_snp = majorSNP(merged_var)\n",
    "top_strand = strand(top_snp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68303b98",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_dict = {'confirmed_snp' : top_snp, 'confirmed_strand' : strand}\n",
    "merged_var['confirmed_snp'] = top_snp\n",
    "merged_var['confirmed_strand'] = top_strand\n",
    "\n",
    "# proposed final output\n",
    "merged_var.drop(columns='pileup')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "964a9242",
   "metadata": {},
   "outputs": [],
   "source": [
    "# majority snp\n",
    "def split(word):\n",
    "    return list(word)\n",
    "\n",
    "def majorSNP(merged_var):\n",
    "    n=0\n",
    "    arr = split(merged_var['pileup'][n])\n",
    "    \n",
    "    snps = pd.Series(arr).drop_duplicates().tolist()\n",
    "    \n",
    "    count = dict.fromkeys(snps, 0)\n",
    "\n",
    "    for snp_key in snps:\n",
    "        for arr_key in arr: \n",
    "            if snp_key == arr_key: \n",
    "                count[snp_key] = count.get(snp_key, 0) + 1\n",
    "                \n",
    "    top_snp = sorted(count)[0]\n",
    "    return top_snp\n",
    "\n",
    "# strand\n",
    "def strand(x):\n",
    "    pos = ['A','C','T','G', 'N']\n",
    "    rev = ['a','c','t','g', 'n']\n",
    "    if x in pos:\n",
    "        return '+'\n",
    "    if x in rev:\n",
    "        return '-'\n",
    "    \n",
    "# grab all snps\n",
    "#def uniqueSNP():\n",
    "    # get the list of SNPs\n",
    "    \n",
    "    # split into unique\n",
    "    \n",
    "\n",
    "def combine(x):\n",
    "    return print(*x, sep='')\n",
    "\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
