# NYU Bioinformatics Capstone Project
Repository for NYU Capstone Project with Dr. Flowers

# Contents
* `report.qmd`
* `scripts/utility.R`
* `scripts/temp.sh`

# Tool Method Plan
![image](https://user-images.githubusercontent.com/92037458/210402025-93ef1bd0-7528-4f35-9754-23ad3988c401.png)

# Background and Purpose
With genetic testing becoming an integral component in clinical diagnosis, accuracy in its predictions are critical for proper assessment and treatment for patients. Current predictive tools use genomic data via whole exome sequencing (WES) and whole genome sequencing (WGS) to predict disease risk, however the diagnostic accuracy is ~40%, leaving more than half of patients without a proper diagnosis (Kim et al., 2020). While comprehensive, both methods lack functional evidence to properly classify disease-causing variants. 

To bridge this gap, RNA-seq is being sought as a way to detect transcriptional defects and increase diagnostic yield (Curry et al., 2021; Docking et al., 2021; Yépez et al., 2022). One clinical study utilized multiple tools in order to compare the results of each method and improve patient stratification (Docking et al., 2021). Additionally, there is evidence of mutant transcripts escaping nonsense-mediated decay (NMD), with their translations not always leading to the expected loss of function (Anderson et al., 2017). This highlights the need for a predictive method for loss of functions in expressed mutations. 

The purpose of this tool is to compare and capture the predicted mutations in expressed transcripts. This would provide additional information to better predict and classify disease-causing variants. 


# References
* Anderson, J. L., Mulligan, T. S., Shen, M.-C., Wang, H., Scahill, C. M., Tan, F. J., Du, S. J., Busch-Nentwich, E. M., & Farber, S. A. (2017). mRNA processing in mutant zebrafish lines generated by chemical and CRISPR-mediated mutagenesis produces unexpected transcripts that escape nonsense-mediated decay. PLoS Genetics, 13(11), e1007105.

* Curry, P. D. K., Broda, K. L., & Carroll, C. J. (2021). The Role of RNA-Sequencing as a New Genetic Diagnosis Tool. Current Genetic Medicine Reports, 9(2), 13–21.

* Docking, T. R., Parker, J. D. K., Jädersten, M., Duns, G., Chang, L., Jiang, J., Pilsworth, J. A., Swanson, L. A., Chan, S. K., Chiu, R., Nip, K. M., Mar, S., Mo, A., Wang, X., Martinez-Høyer, S., Stubbins, R. J., Mungall, K. L., Mungall, A. J., Moore, R. A., … Karsan, A. (2021). A clinical transcriptome approach to patient stratification and therapy selection in acute myeloid leukemia. Nature Communications, 12(1), 2474.

* Kim, M.-J., Yum, M.-S., Seo, G. H., Lee, Y., Jang, H. N., Ko, T.-S., & Lee, B. H. (2020). Clinical Application of Whole Exome Sequencing to Identify Rare but Remediable Neurologic Disorders. Journal of Clinical Medicine Research, 9(11). https://doi.org/10.3390/jcm9113724

* Yépez, V. A., Gusic, M., Kopajtich, R., Mertes, C., Smith, N. H., Alston, C. L., Ban, R., Beblo, S., Berutti, R., Blessing, H., Ciara, E., Distelmaier, F., Freisinger, P., Häberle, J., Hayflick, S. J., Hempel, M., Itkis, Y. S., Kishita, Y., Klopstock, T., … Prokisch, H. (2022). Clinical implementation of RNA sequencing for Mendelian disease diagnostics. Genome Medicine, 14(1), 38.
